{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HuGS-DW: Human GWAS-SNP Cross-Reference Data Warehouse\n",
    "#### Este notebook e todos os ficheiros acessórios, podem ser directamente acedidos em:\n",
    "https://drive.google.com/open?id=1rTqXz1AE8mW-uZWK81W0BnNA2YmE11JZ\n",
    "### 1. Dimensions and facts tables of the data warehouse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. Define and model them in SQL\n",
    "##### O \"star schema\" actualizado do nosso DW, é o seguinte:\n",
    "\n",
    "\n",
    "![alt text](HuGs_Star_Schema.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As tabelas de dimensões e factos identificadas na fase 1 do projecto, foram actualizadas (com adição de novas tabelas de dimensões auxiliares, como a Allele_Dimension e mudanças de medidas associadas às tabelas de factos) e modeladas em SQL.\n",
    "\n",
    "### Tabelas de Dimensões\n",
    "\n",
    "* **Gender_Dimension**- Contém dados sobre esperança média de vida e rendimento nacional bruto (em dólares) para indivíduos de diferentes sexos, distribuídos pelos diferentes países nos quais se distribuem os indíviduos do estudo. Agrega uma primary key (gender_key) com outras variáveis como o country code, country e gender. Foram ainda adicionados indíces demográficos como Average Income e  Mortality rate, ambos de 2015 (data da publicação do estudo do 1000 genomes project). Esta dimensão vai permitir obter informação demográfica, diferenciada por género e país.\n",
    "\n",
    "* **Population_Dimension** - Nesta dimensão agrega-se informação acerca de cada uma das populações incluídas no estudo genético do 1000 genomes. Esta dimensão contém informação para: Population key (PK), Population Code, Population Description, Super Population Code, Super Population (East Asian, South Asian, African, European, American), Ethinicity, Country code, Country, Region (e.g. Utah) e City. Inclui também dados demográficos, como Life Expectancy e Gross Income (ambos relativos ao ano de 2015, ano de publicação do estudo). Esta informação demográfica foi também modelada em classes e adicionada à tabela de dimensão.\n",
    "\n",
    "* **Person_Dimension**- Dimensão modelada com: Person key(PK), Person  code, Population Code, Population Description, Super Population Code, Super Population, Country Code, Country e Gender. Esta dimensão agrega a informação sobre cada uma das pessoas estudadas no 1000 genomes project.\n",
    "\n",
    "* **Chromosome_Dimension**- Aqui, agrupa-se a informação sobre os cromossomas do genoma humano. Esta dimensão tem informação para Choromosome Key(PK), Chromosome Number, RefSeqID, Size(Mb), GC%, protein, rRNA, tRNA, other RNA, gene, pseudogene,  Number of genes per mb. \n",
    "\n",
    "* **Allele_Dimension**- Esta \"role-playing\" dimension, foi criada para ajudar na inteligibilidade do esquema de dados. Contém apenas duas linhas, quer permitem identificar os alelos como de referência ou alternativos. Visto que os dados de variantes genéticas do 1000 genomes estão codificados como 0, para um alelo de referência e 1, para um alelo alternativo, esta dimensão vai permitir identificar cada tipo de alelo dos indivíduos. Contém unicamente uma Allele Key e Allele Type.\n",
    "\n",
    "* **SNP_Dimension**- Esta dimensão contém informações acerca dos SNPs presentes no cromossoma 19, nomeadamente, a referência, dados acerca da sua localização e contexto genómico, informações acerca da frequência em diferentes populações e consequências moleculares associadas. A tabela de dimensão tem os seguintes atributos: SNP Key(PK), chromosome, snp position,  snp ref code, ReferenceAllele, AlternativeAllele, East Asian frequency, European frequency, African frequency, American frequency, South ASian frequency.\n",
    "\n",
    "* **Disease_Dimension**- Dimensão onde se agrega a informação acerca de doenças humanas (da “Human Disease Ontology” - DOID). Contêm  Disease key(PK), wikidata_id, Disease_name, Doid_ID, Mesh_ID, Type_of_Disease, Health_specialities, Symptoms, Genetic_Association, Organs affected, Caused, Alternative_names. \n",
    "\n",
    "* **Study_Dimension**- Esta dimensão inclui informação provenient de estudos de GWAS e agregados no site GWAS Catalog (https://www.ebi.ac.uk/gwas/). Agrupa informação sobre Study key (PK), Study,  Date, Disease, Region, Chromosome Position, Reported genes, Mapped_genes, Upstream Gene ID, Downstream Gene ID, SNP Gene IDS, Upstream Gene Distance, Dowstream Gene Distance, Strongest SNP Risk Allele , SNPs , Context, Intergenic, Risk Allele Frequency, p -value. Em suma, nesta dimensão existem vários parâmetros estatísticos para a associação de uma doença a uma determinada variante genética.\n",
    "\n",
    "### Tabelas de Factos\n",
    "* **SNP Person Fact**- Esta tabela de factos mudou relativamente ao nosso esquema inicial. É agora uma factless table, que contém informação acerca do genótipo em cada um dos SNPs de cada um dos 2503 indivíduos estudados. Inclui chaves estrangeiras que vão permitir ligar os factos, às diferentes dimensões: Gender Key, Population Key, Person Key, Chromossome Key, SNP Key, Allele 1 Key, Allele 2 Key (estas duas últimas colunas, ligam à mesma tabela de Allele_Dimension).\n",
    "\n",
    "* **Disease GWAS fact**- Esta tabela vai conter informação atómica acerca da associação genética entre uma doença e um SNP,  identificada num determinado estudo. Mudamos esta fact table, para incluir uma medida não aditiva, o p-value, que efectivamente mede a a validade probabilística da associação entre uma doença e um SNP. Inclui chaves estrangeiras que ligam às várias dimensões, como SNP Key, Disease Keye Study Key e ainda uma medida não aditiva, p-value."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. Identify hierarchies and fact granularity\n",
    "##### Granularidade das tabelas de facto\n",
    "- A granularidade dos factos não mudaram relativamente à primeira fase.\n",
    "- **SNP Person Fact**- Nesta tabela cada linha corresponde à variante de um dado SNP para um dado indivíduo. Inclui o genótipo para os dois alelos do indivíduo e referências para tabelas de dimensões contendo informações sobre esse indivíduo, população, género, cromossoma e informações sobre o SNP.\n",
    "- **Disease GWAS Fact**- Nesta tabela cada linha corresponde a uma referência num estudo que associa um SNP a uma doença e a respectiva medida probabilística dessa associação.\n",
    "\n",
    "##### Hierarquias do sistema de dados\n",
    "- Dentro das várias tabelas de dimensões, não há hierarquias óbvias que necessitem de uma modelação mais elaborada. \n",
    "- A única hierarquia que existe nos nossos dados, prende-se com informação geográfica na Population Dimension. Com o objectivo de manter a legibilidade da estrutura de dados, para o end-user, as hierarquias são mantidas na mesma tabela. \n",
    "- Com o crescimento do número de indíviduos estudados, poder-se-iam contemplar criar outtriggers para dados demográficos e organizar melhor a hierarquia geográfica. Presentemente, tendo em conta a dimensão do dataset, esta modelação não se justifica."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### c. Create the dimensions and facts tables in the DBMS (postgreSQL)\n",
    "- Em baixo colocamos o código usado para modelar, filtrar, inserir dados e executar instruções SQL, que são utilizados ao longo de todo o notebook. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import psycopg2 as pg\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def pandafy(rows, colnames):\n",
    "    N=len(colnames)\n",
    "    D={cn: [] for cn in colnames}\n",
    "    for row in rows:\n",
    "        for i in range(N): D[colnames[i]].append(row[i])\n",
    "    pdfy=pd.DataFrame(D)\n",
    "    return pdfy\n",
    "\n",
    "def excuteSingleSQLstatement(sql, host, database, user, password):\n",
    "    conn = pg.connect(host=host,database=database, user=user, password=password)\n",
    "    cur = conn.cursor()\n",
    "    cur.execute(sql)\n",
    "    cur.close()\n",
    "    conn.commit()\n",
    "    conn.close()    \n",
    "\n",
    "def getSQLfromQuery(sql, params, host, database, user, password):\n",
    "    conn = pg.connect(host=host, database=database, user=user, password=password)\n",
    "    cur  = conn.cursor()    \n",
    "    if len(params)==0:\n",
    "        cur.execute(sql)\n",
    "    else: \n",
    "        cur.execute(sql, params)\n",
    "    data = cur.fetchall()\n",
    "    colnames = [desc[0] for desc in cur.description]\n",
    "    cur.close()\n",
    "    conn.close()\n",
    "    df=pandafy(data, colnames)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Em baixo, estão os comandos SQL para criar as várias tabelas de dimensões e factos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ATTENTION!!!!\n",
    "# Run only if you want to DROP all tables from the database!!!\n",
    "\n",
    "SQL = '''\n",
    "drop table if exists chromosome_dimension;\n",
    "drop table if exists person_dimension;\n",
    "drop table if exists population_dimension;\n",
    "drop table if exists gender_dimension;\n",
    "drop table if exists allele_dimension;\n",
    "drop table if exists disease_dimension;\n",
    "drop table if exists snp_dim;\n",
    "drop table if exists study_dim;\n",
    "drop table if exists disease_gwas_fact;\n",
    "drop table if exists snp_person_fact\n",
    "\n",
    "'''\n",
    "\n",
    "# Remove # from the line below,  if you want to drop all tables.\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL = '''\n",
    "create table chromosome_dimension(\n",
    "chromosome_key int,\n",
    "chromosome varchar(2),\n",
    "refseq varchar(12), \n",
    "size_Mb float,\n",
    "GC_percentage float,\n",
    "proteins numeric,\n",
    "rRNA  numeric,\n",
    "tRNA numeric,\n",
    "other_RNA numeric,\n",
    "gene numeric,\n",
    "pseudogene numeric,\n",
    "genes_per_MB float,\n",
    "primary key (chromosome_key)\n",
    ");\n",
    "\n",
    "create table person_dimension(\n",
    "person_key int,\n",
    "person_code varchar(8),\n",
    "population_code char(3),\n",
    "population_description varchar(128), \n",
    "super_population_code char(3),\n",
    "super_population varchar(128),\n",
    "country_code char(3),\n",
    "country varchar(128),\n",
    "gender varchar(8),\n",
    "primary key (person_key)\n",
    ");\n",
    "\n",
    "create table population_dimension(\n",
    "population_key int,\n",
    "population_code char(3),\n",
    "population_description varchar(128), \n",
    "super_population_code char(3),\n",
    "super_population varchar(128),\n",
    "ethnicity varchar(32),\n",
    "country_code char(3),\n",
    "country varchar(128),\n",
    "region varchar(128),\n",
    "city varchar(128),\n",
    "gross_national_income int,\n",
    "gross_national_income_group varchar(24),\n",
    "life_expectancy_birth float,\n",
    "life_expectancy_birth_group varchar(24),\n",
    "primary key (population_key)\n",
    ");\n",
    "\n",
    "create table gender_dimension(\n",
    "gender_key int,\n",
    "country_code char(3),\n",
    "country varchar(128),\n",
    "gender varchar(8),\n",
    "literacy_rate float,\n",
    "unemployment_rate float,\n",
    "mortality_per_1000_habitants float,\n",
    "primary key (gender_key)\n",
    ");\n",
    "\n",
    "create table allele_dimension(\n",
    "allele_key int,\n",
    "allele_type_code char(3),\n",
    "allele_type varchar(12),\n",
    "primary key (allele_key)\n",
    ");\n",
    "\n",
    "create table disease_dimension(\n",
    "disease_key int,\n",
    "wikidata_id varchar(30),\n",
    "disease_name varchar(160),\n",
    "doid_id int, \n",
    "mesh_id varchar(20),\n",
    "type_of_disease varchar(1760),\n",
    "health_speciality  varchar(160),\n",
    "symptoms varchar(1860),\n",
    "genetic_associations varchar(60) ARRAY,\n",
    "affects varchar(280),\n",
    "caused_by varchar(480),\n",
    "alternative_names varchar(20000),\n",
    "primary key (disease_key)\n",
    ");\n",
    "\n",
    "create table snp_dim(\n",
    "snp_id serial primary key,\n",
    "snp_ref_code varchar(64),\n",
    "gene varchar(16),\n",
    "functional_consequence varchar(512), --arrays?\n",
    "clinical_significance varchar(512), -- arrays?\n",
    "chromosome float,\n",
    "snp_position varchar(32),\n",
    "reference char(128),\n",
    "alternative char(128),\n",
    "info_string varchar(1024),\n",
    "context varchar(64),\n",
    "easian_frequency float,\n",
    "european_frequency float,\n",
    "african_frequency float,\n",
    "american_frequency float,\n",
    "sasian_frequency float,\n",
    "variation_type varchar(34),\n",
    "consequence varchar(35),\n",
    "mapped_gene varchar(36),\n",
    "upstream_gene_id varchar(37), \n",
    "downstream_gene_id varchar(38),\n",
    "upstream_gene_distance varchar(39),\n",
    "dowstream_gene_distance varchar(40),\n",
    "region varchar(41),\n",
    "chr_pos varchar(42)\n",
    ");\n",
    "\n",
    "create table study_dim(\n",
    "study_id serial primary key,\n",
    "pumbmed_id int,\n",
    "study_date date,\n",
    "disease varchar(256),\n",
    "region varchar(32),\n",
    "chr_pos varchar(64),\n",
    "reported_genes varchar(120) array,\n",
    "mapped_genes varchar(121) array,\n",
    "upstream_gene_id varchar(122),\n",
    "downstream_gene_id varchar(123),\n",
    "snp_gene_ids varchar(512) array,\n",
    "upstream_gene_distance varchar(124),\n",
    "dowstream_gene_distance varchar(125),\n",
    "strongest_snp_risk_allele varchar(126),\n",
    "snps varchar(127),\n",
    "snp_id_current varchar(128),\n",
    "context varchar(129),\n",
    "intergenic varchar(130),\n",
    "risk_allele_frequency varchar(131),\n",
    "p_value float\n",
    ");\n",
    "\n",
    "--facts\n",
    "create table snp_person_fact(\n",
    "gender_id integer REFERENCES gender_dimension(gender_key),\n",
    "population_id integer REFERENCES population_dimension(population_key),\n",
    "snp_id integer REFERENCES snp_dim(snp_id),\n",
    "person_id integer REFERENCES person_dimension(person_key),\n",
    "chromossome_id integer REFERENCES chromosome_dimension(chromosome_key),\n",
    "allele_1 int REFERENCES allele_dimension(allele_key), \n",
    "allele_2 int REFERENCES allele_dimension(allele_key)\n",
    ");\n",
    "\n",
    "create table disease_gwas_fact(\n",
    "disease_id integer REFERENCES disease_dimension(disease_key),\n",
    "snp_id integer REFERENCES snp_dim(snp_id),\n",
    "study_id integer REFERENCES study_dim(study_id),\n",
    "p_value float\n",
    ")\n",
    "'''\n",
    "\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Define an ETL workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. Identify all data sources for all dimensions. Add URL links to all data that should be available. If not public data, point to dropbox files, Google drive, or whatever\n",
    "\n",
    "O preenchimento das várias dimensões deste DW, implicou um uso de um grande número de datasets distintos, que foram processados, filtrados e agregados em várias tabelas durante o processo de ETL.\n",
    "Assim, por dimensão, os datasets usados foram:\n",
    "\n",
    "#### Person_Dimension\n",
    "- Esta dimensão, contendo informação acerca de cada um dos indíviduos cujo genoma foi sequenceado, está contida nos FTPs do 1000 genomes project.\n",
    "- Dataset com informação acerca dos indivíduos analisados no estudo: ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel\n",
    "- Informação acerca das populações dos indivíduos estudados: ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/20131219.populations.tsv\n",
    "- Informação acerca das superpopulações dos indivíduos estudados: ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/20131219.superpopulations.tsv\n",
    "\n",
    "#### Population_Dimension\n",
    "- Esta dimensão contém informação acerca das populações analisadas no estudo dos 1000 genomes project. A tabela foi manualmente curada, visto ter uma dimensão reduzida.\n",
    "- Informação acerca das populações dos indivíduos estudados: ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/20131219.populations.tsv\n",
    "- Tabela modificada das populações: https://drive.google.com/open?id=1TQdAgXMhoDs3CBjnOIFuEcJDjat3U704bJtktkNSnOo\n",
    "- Para informação demográfica acerca dos países, usamos os dados do \"The World Bank\" (https://data.worldbank.org/)\n",
    "- Gross Income: https://data.worldbank.org/indicator/NY.GNP.PCAP.PP.CD\n",
    "- Gross Income (Google Drive direct download): https://drive.google.com/open?id=1u0pDvbPSGk0xGdrJksvlC6ia3F54yxTyviA98CT1chw\n",
    "- Life expectancy at birth: https://data.worldbank.org/indicator/sp.dyn.le00.in\n",
    "- Life expectancy at birth (Google Drive direct download): https://drive.google.com/open?id=1BNiG-yshJJgKSv9P5qXbp6nWHKnmS7S6\n",
    "\n",
    "#### Gender_Dimension\n",
    "- Esta dimensão agrega informação acerca do género e de indíces demográficos em cada um dos países em que o estudo foi realizado.\n",
    "- A informação acerca do género, é derivada das tabelas de indivíduos e populações descritas em cima. A estas adiciona-se dados do \"The World Bank\".\n",
    "- Mortality rate per 1,000 adults: for female- https://data.worldbank.org/indicator/SP.DYN.AMRT.FE for male https://data.worldbank.org/indicator/SP.DYN.AMRT.MA\n",
    "- Unemployment per 1,000 adults: for female -https://data.worldbank.org/indicator/SL.UEM.TOTL.FE.ZS for male https://data.worldbank.org/indicator/SL.UEM.TOTL.MA.ZS?view=chart\n",
    "- Literacy rate : for female-https://data.worldbank.org/indicator/SE.ADT.LITR.FE.ZS for male -https://data.worldbank.org/indicator/SE.ADT.LITR.MA.ZS \n",
    "\n",
    "#### Chromosome_Dimension\n",
    "- Dimensão que agrega informação acerca dos vários cromossomas nos quais se distribuem os SNPs identificados no estudo.\n",
    "- Apesar deste exercício académico só conter a informação genética do cromossoma 19 (devido à dimensão dos datasets), a informação de todos os cromossomas foi agregada nesta dimensão.\n",
    "- Tabela com informação acerca dos cromossomas do genoma humano: https://www.ncbi.nlm.nih.gov/genome/?term=human[organism]\n",
    "- Link para csv criada: https://drive.google.com/open?id=1hUzEmuz-3nwMRg-VHmmR8NJUvthkjvl7\n",
    "\n",
    "#### Allele_Dimension\n",
    "- Esta pequena dimensão role-playing, agrega informação acerca do tipo de variação encontrada no genótipo de cada indíviduo. Em suma, dá informação acerca da presença de uma alelo de referência ou alternativo em cada cromossoma.\n",
    "- A informação genética para cada indíviduo analizado no estudo, para o cromossoma 19, encontra-se em: ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr19.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz\n",
    "\n",
    "#### SNP_Dimension\n",
    "- Esta dimensão agrega vários tipos de informação acerca dos SNPs no cromossoma 19. Combina vários datasets:\n",
    "- Informação genética dos 1000 genomes project: ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr19.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz\n",
    "- Informação retirada do dbSNP: https://www.ncbi.nlm.nih.gov/snp/?term=9%5BChromosome%5D\n",
    "- Informação retirada do GWAS dataset: https://www.ebi.ac.uk/gwas/api/search/downloads/alternative\n",
    "\n",
    "#### Study_Dimension\n",
    "- Esta dimensão contém informação de estudos de Genome-Wide Association Studies.\n",
    "- Contém informação retirada do GWAS association dataset: https://www.ebi.ac.uk/gwas/api/search/downloads/alternative\n",
    "\n",
    "#### Disease_Dimension\n",
    "- Aqui guardam-se as doenças humanas para as quais existe um identificador Doid (Human Disease Ontology).\n",
    "- Esta informação foi particularmente difícil de agregar (por falta de uma ferramenta que agregue esta informação). No final, a informação foi descarregada do Wikidata, usando SPARQL.\n",
    "- Informação acerca de todas as doenças com um doidID: https://tinyurl.com/y9hu7nvb\n",
    "- Informação acerca de termos alternativos para as doenças: https://tinyurl.com/yd2zwg69"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. For each dimension show the code used for modeling, filtering and inserting data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Chromosome_Dimension"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Código para ler e filtrar informação acerca de cromossomas humanos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "def load_csv(csv_file):\n",
    "    '''This function reads a csv file with information about human chromosomes.\n",
    "    It compiles the information about each chromosome into a list of lists.\n",
    "    It corrects value types. It also calculates the gene density for each chromosome.\n",
    "    Requires: A csv file with information about each chromosome.\n",
    "    Ensures: A list of lists, with all modeled information about chromosomes.\n",
    "    '''\n",
    "    \n",
    "    with open(csv_file, newline='') as csv_file:\n",
    "        csv_reader = csv.reader(csv_file, delimiter=',')\n",
    "        data = list(csv_reader)\n",
    "\n",
    "        chromosomes = []\n",
    "        i = 1\n",
    "        \n",
    "        for chr in data[1:25]:\n",
    "            key = i\n",
    "            chrom = chr[2]\n",
    "            refSeq = chr[3]\n",
    "            size = float(chr[5])\n",
    "            GC = float(chr[6])\n",
    "            prot = int(chr[7].replace(',', ''))\n",
    "            if chr[8].isdigit():\n",
    "                rRNA = int(chr[8])\n",
    "            else:\n",
    "                rRNA = 0\n",
    "            if chr[9].isdigit():\n",
    "                tRNA = int(chr[9])\n",
    "            else:\n",
    "                tRNA = 0\n",
    "            if chr[10].isdigit():\n",
    "                otherRNA = int(chr[10].replace(',', ''))\n",
    "            else:\n",
    "                otherRNA = 0                \n",
    "            gene = int(chr[11].replace(',', ''))\n",
    "            if chr[12].isdigit():\n",
    "                pseudogene = int(chr[12].replace(',', ''))\n",
    "            else:\n",
    "                pseudogene = 0                \n",
    "            genesPerMb = round(gene/size, 2)\n",
    "            \n",
    "            i += 1\n",
    "            \n",
    "            chromosomes.append([key, chrom, refSeq, size, GC, prot, rRNA, tRNA, otherRNA, gene, pseudogene, genesPerMb])\n",
    "     \n",
    "    return chromosomes\n",
    "\n",
    "chromosomes = (load_csv(\"Chromosome_Dimension_DB.csv\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inserção e verificação dos dados na Chromosome_Dimension"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SQL = \"\"\"drop table chromosome_dimension\"\"\"\n",
    "SQL= \"\"\"create table chromosome_dimension(\n",
    "\tchromosome_key int,\n",
    "\tchromosome varchar(2),\n",
    "\trefseq varchar(12), \n",
    "\tsize_Mb float,\n",
    "\tGC_percentage float,\n",
    "    proteins numeric,\n",
    "    rRNA  numeric,\n",
    "    tRNA numeric,\n",
    "    other_RNA numeric,\n",
    "    gene numeric,\n",
    "    pseudogene numeric,\n",
    "    genes_per_MB float,\n",
    "\tprimary key (chromosome_key)\n",
    ")\"\"\"\n",
    "\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Este código assume que estamos a inserir os dados no servidor pgadmin do DI da FCUL.\n",
    "conn = pg.connect(host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "curs = conn.cursor()\n",
    "for chr in chromosomes:\n",
    "    key, chrom, refSeq, size, GC, prot, rRNA, tRNA, otherRNA, gene, pseudogene, genesPerMb = chr\n",
    "    curs.execute(\"insert into chromosome_dimension values(%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\", (key, chrom, refSeq, size, GC, prot, rRNA, tRNA, otherRNA, gene, pseudogene, genesPerMb))\n",
    "curs.close()\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_key</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>refseq</th>\n",
       "      <th>size_mb</th>\n",
       "      <th>gc_percentage</th>\n",
       "      <th>proteins</th>\n",
       "      <th>rrna</th>\n",
       "      <th>trna</th>\n",
       "      <th>other_rna</th>\n",
       "      <th>gene</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>genes_per_mb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>248.96</td>\n",
       "      <td>42.3</td>\n",
       "      <td>11096</td>\n",
       "      <td>17</td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "      <td>5101</td>\n",
       "      <td>0</td>\n",
       "      <td>20.49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>242.19</td>\n",
       "      <td>40.3</td>\n",
       "      <td>8455</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>3879</td>\n",
       "      <td>0</td>\n",
       "      <td>16.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>NC_000003.12</td>\n",
       "      <td>198.30</td>\n",
       "      <td>39.7</td>\n",
       "      <td>7242</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2988</td>\n",
       "      <td>909</td>\n",
       "      <td>15.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>NC_000004.12</td>\n",
       "      <td>190.22</td>\n",
       "      <td>38.3</td>\n",
       "      <td>4646</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2439</td>\n",
       "      <td>805</td>\n",
       "      <td>12.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>181.54</td>\n",
       "      <td>39.5</td>\n",
       "      <td>4803</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>2595</td>\n",
       "      <td>789</td>\n",
       "      <td>14.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>170.81</td>\n",
       "      <td>39.6</td>\n",
       "      <td>5630</td>\n",
       "      <td>0</td>\n",
       "      <td>138</td>\n",
       "      <td>0</td>\n",
       "      <td>3019</td>\n",
       "      <td>892</td>\n",
       "      <td>17.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>159.35</td>\n",
       "      <td>40.7</td>\n",
       "      <td>5220</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>2774</td>\n",
       "      <td>914</td>\n",
       "      <td>17.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>NC_000008.11</td>\n",
       "      <td>145.14</td>\n",
       "      <td>40.2</td>\n",
       "      <td>4087</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2174</td>\n",
       "      <td>680</td>\n",
       "      <td>14.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>NC_000009.12</td>\n",
       "      <td>138.40</td>\n",
       "      <td>42.3</td>\n",
       "      <td>4679</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2269</td>\n",
       "      <td>723</td>\n",
       "      <td>16.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>133.80</td>\n",
       "      <td>41.6</td>\n",
       "      <td>5449</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2178</td>\n",
       "      <td>643</td>\n",
       "      <td>16.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>135.09</td>\n",
       "      <td>41.6</td>\n",
       "      <td>6715</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>2923</td>\n",
       "      <td>834</td>\n",
       "      <td>21.64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>133.28</td>\n",
       "      <td>40.8</td>\n",
       "      <td>5954</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>2536</td>\n",
       "      <td>698</td>\n",
       "      <td>19.03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>NC_000013.11</td>\n",
       "      <td>114.36</td>\n",
       "      <td>40.2</td>\n",
       "      <td>2029</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1379</td>\n",
       "      <td>475</td>\n",
       "      <td>12.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>NC_000014.9</td>\n",
       "      <td>107.04</td>\n",
       "      <td>42.2</td>\n",
       "      <td>3497</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>2063</td>\n",
       "      <td>586</td>\n",
       "      <td>19.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>101.99</td>\n",
       "      <td>43.4</td>\n",
       "      <td>3559</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1827</td>\n",
       "      <td>564</td>\n",
       "      <td>17.91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>90.34</td>\n",
       "      <td>45.1</td>\n",
       "      <td>4564</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>0</td>\n",
       "      <td>1948</td>\n",
       "      <td>478</td>\n",
       "      <td>21.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>83.26</td>\n",
       "      <td>45.3</td>\n",
       "      <td>6115</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>2452</td>\n",
       "      <td>572</td>\n",
       "      <td>29.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>NC_000018.10</td>\n",
       "      <td>80.37</td>\n",
       "      <td>39.8</td>\n",
       "      <td>2035</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>986</td>\n",
       "      <td>298</td>\n",
       "      <td>12.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>58.62</td>\n",
       "      <td>47.9</td>\n",
       "      <td>6700</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>2490</td>\n",
       "      <td>523</td>\n",
       "      <td>42.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>NC_000020.11</td>\n",
       "      <td>64.44</td>\n",
       "      <td>43.9</td>\n",
       "      <td>2810</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1358</td>\n",
       "      <td>341</td>\n",
       "      <td>21.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>NC_000021.9</td>\n",
       "      <td>46.71</td>\n",
       "      <td>42.2</td>\n",
       "      <td>1283</td>\n",
       "      <td>12</td>\n",
       "      <td>1</td>\n",
       "      <td>708</td>\n",
       "      <td>778</td>\n",
       "      <td>209</td>\n",
       "      <td>16.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>50.82</td>\n",
       "      <td>47.7</td>\n",
       "      <td>2512</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1186</td>\n",
       "      <td>355</td>\n",
       "      <td>23.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>X</td>\n",
       "      <td>NC_000023.11</td>\n",
       "      <td>156.04</td>\n",
       "      <td>39.6</td>\n",
       "      <td>3811</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>2199</td>\n",
       "      <td>893</td>\n",
       "      <td>14.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>Y</td>\n",
       "      <td>NC_000024.10</td>\n",
       "      <td>57.23</td>\n",
       "      <td>45.4</td>\n",
       "      <td>321</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>318</td>\n",
       "      <td>582</td>\n",
       "      <td>396</td>\n",
       "      <td>10.17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chromosome_key chromosome        refseq  size_mb  gc_percentage proteins  \\\n",
       "0                1          1  NC_000001.11   248.96           42.3    11096   \n",
       "1                2          2  NC_000002.12   242.19           40.3     8455   \n",
       "2                3          3  NC_000003.12   198.30           39.7     7242   \n",
       "3                4          4  NC_000004.12   190.22           38.3     4646   \n",
       "4                5          5  NC_000005.10   181.54           39.5     4803   \n",
       "5                6          6  NC_000006.12   170.81           39.6     5630   \n",
       "6                7          7  NC_000007.14   159.35           40.7     5220   \n",
       "7                8          8  NC_000008.11   145.14           40.2     4087   \n",
       "8                9          9  NC_000009.12   138.40           42.3     4679   \n",
       "9               10         10  NC_000010.11   133.80           41.6     5449   \n",
       "10              11         11  NC_000011.10   135.09           41.6     6715   \n",
       "11              12         12  NC_000012.12   133.28           40.8     5954   \n",
       "12              13         13  NC_000013.11   114.36           40.2     2029   \n",
       "13              14         14   NC_000014.9   107.04           42.2     3497   \n",
       "14              15         15  NC_000015.10   101.99           43.4     3559   \n",
       "15              16         16  NC_000016.10    90.34           45.1     4564   \n",
       "16              17         17  NC_000017.11    83.26           45.3     6115   \n",
       "17              18         18  NC_000018.10    80.37           39.8     2035   \n",
       "18              19         19  NC_000019.10    58.62           47.9     6700   \n",
       "19              20         20  NC_000020.11    64.44           43.9     2810   \n",
       "20              21         21   NC_000021.9    46.71           42.2     1283   \n",
       "21              22         22  NC_000022.11    50.82           47.7     2512   \n",
       "22              23          X  NC_000023.11   156.04           39.6     3811   \n",
       "23              24          Y  NC_000024.10    57.23           45.4      321   \n",
       "\n",
       "   rrna trna other_rna  gene pseudogene  genes_per_mb  \n",
       "0    17   90         0  5101          0         20.49  \n",
       "1     0    7         0  3879          0         16.02  \n",
       "2     0    4         0  2988        909         15.07  \n",
       "3     0    1         0  2439        805         12.82  \n",
       "4     0   17         0  2595        789         14.29  \n",
       "5     0  138         0  3019        892         17.67  \n",
       "6     0   22         0  2774        914         17.41  \n",
       "7     0    4         0  2174        680         14.98  \n",
       "8     0    3         0  2269        723         16.39  \n",
       "9     0    3         0  2178        643         16.28  \n",
       "10    0   13         0  2923        834         21.64  \n",
       "11    0    9         0  2536        698         19.03  \n",
       "12    0    4         0  1379        475         12.06  \n",
       "13    0   18         0  2063        586         19.27  \n",
       "14    0    9         0  1827        564         17.91  \n",
       "15    0   27         0  1948        478         21.56  \n",
       "16    0   33         0  2452        572         29.45  \n",
       "17    0    1         0   986        298         12.27  \n",
       "18    0    6         0  2490        523         42.48  \n",
       "19    0    0         0  1358        341         21.07  \n",
       "20   12    1       708   778        209         16.66  \n",
       "21    0    0         0  1186        355         23.34  \n",
       "22    0    4         0  2199        893         14.09  \n",
       "23    0    0       318   582        396         10.17  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select * from chromosome_dimension\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Person_Dimension"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Código para ler, filtrar e modelar dados acerca de 2503 indivíduos do estudo 1000 genomes project."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "def load_csv(pop_csv_file, csv_file):\n",
    "    '''\n",
    "    This function loads the file with information about the populations and a file with information about individuals.\n",
    "    It compiles this information into a new list, ready to insert into the Person_Dimension table in pgAdmin.\n",
    "    Requires: A population table and an individuals table, following a specific format.\n",
    "    Ensures: A list of individuals studied, with information about population, superpopulation and gender.\n",
    "    '''\n",
    "\n",
    "    with open(pop_csv_file, newline='') as csv_file1:\n",
    "        csv_reader = csv.reader(csv_file1, delimiter=\"\\t\")\n",
    "        popData = list(csv_reader)\n",
    "\n",
    "        population = []\n",
    "        i = 1\n",
    "        \n",
    "        for pop in popData[1:27]:\n",
    "            key = i\n",
    "            popCode = pop[0]\n",
    "            popDescript = pop[1]\n",
    "            superPopCode = pop[2]\n",
    "            superPop = pop[3]\n",
    "            ethnicity = pop[4]\n",
    "            countryCode = pop[5]\n",
    "            country = pop[6]\n",
    "            \n",
    "            i += 1\n",
    "            \n",
    "            population.append([key, popCode, popDescript, superPopCode, superPop, ethnicity, countryCode, country])\n",
    "            \n",
    "    with open(csv_file, newline='') as csv_file2:\n",
    "        csv_reader = csv.reader(csv_file2, delimiter=\"\\t\")\n",
    "        data = list(csv_reader)\n",
    "\n",
    "        person = []\n",
    "        i = 1\n",
    "        \n",
    "        for ppl in data[1:]:\n",
    "            key = i\n",
    "            individual = ppl[0]\n",
    "            popCode = ppl[1]\n",
    "            for pop in population:\n",
    "                if popCode == pop[1]:\n",
    "                    popDescript = pop[2]\n",
    "            superPopCode = ppl[2]\n",
    "            for pop in population:\n",
    "                if superPopCode == pop[3]:\n",
    "                    superPop = pop[4]\n",
    "            for pop in population:\n",
    "                if popCode == pop[1]:\n",
    "                    countryCode = pop[6]\n",
    "            for pop in population:\n",
    "                if countryCode == pop[6]:\n",
    "                    country = pop[7]\n",
    "            gender = ppl[3]\n",
    "            \n",
    "            i += 1\n",
    "            \n",
    "            person.append([key, individual, popCode, popDescript, superPopCode, superPop, countryCode, country, gender])\n",
    "\n",
    "    return person\n",
    "\n",
    "people = load_csv('20131219.populations - 20131219.populations.tsv', 'integrated_call_samples_v3.20130502.ALL.panel')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inserção e verificação dos dados na Person_Dimension"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SQL = \"\"\"drop table person_dimension\"\"\"\n",
    "SQL= \"\"\"create table person_dimension(\n",
    "\tperson_key int,\n",
    "\tperson_code varchar(8),\n",
    "\tpopulation_code char(3),\n",
    "    population_description varchar(128), \n",
    "\tsuper_population_code char(3),\n",
    "\tsuper_population varchar(128),\n",
    "    country_code char(3),\n",
    "    country varchar(128),\n",
    "    gender varchar(8),\n",
    "    primary key (person_key)\n",
    ")\"\"\"\n",
    "\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = pg.connect(host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "curs = conn.cursor()\n",
    "for ppl in people:\n",
    "    key, individual, popCode, popDescript, superPopCode, superPop, countryCode, country, gender = ppl\n",
    "    curs.execute(\"insert into person_dimension values(%s, %s, %s, %s, %s, %s, %s, %s, %s)\", (key, individual, popCode, popDescript, superPopCode, superPop, countryCode, country, gender))\n",
    "curs.close()\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>person_key</th>\n",
       "      <th>person_code</th>\n",
       "      <th>population_code</th>\n",
       "      <th>population_description</th>\n",
       "      <th>super_population_code</th>\n",
       "      <th>super_population</th>\n",
       "      <th>country_code</th>\n",
       "      <th>country</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>GBR</td>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>HG00097</td>\n",
       "      <td>GBR</td>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>HG00099</td>\n",
       "      <td>GBR</td>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>HG00100</td>\n",
       "      <td>GBR</td>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>HG00101</td>\n",
       "      <td>GBR</td>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2499</th>\n",
       "      <td>2500</td>\n",
       "      <td>NA21137</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati Indian in Houston,TX</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2500</th>\n",
       "      <td>2501</td>\n",
       "      <td>NA21141</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati Indian in Houston,TX</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2501</th>\n",
       "      <td>2502</td>\n",
       "      <td>NA21142</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati Indian in Houston,TX</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2502</th>\n",
       "      <td>2503</td>\n",
       "      <td>NA21143</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati Indian in Houston,TX</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2503</th>\n",
       "      <td>2504</td>\n",
       "      <td>NA21144</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati Indian in Houston,TX</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2504 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      person_key person_code population_code           population_description  \\\n",
       "0              1     HG00096             GBR  British in England and Scotland   \n",
       "1              2     HG00097             GBR  British in England and Scotland   \n",
       "2              3     HG00099             GBR  British in England and Scotland   \n",
       "3              4     HG00100             GBR  British in England and Scotland   \n",
       "4              5     HG00101             GBR  British in England and Scotland   \n",
       "...          ...         ...             ...                              ...   \n",
       "2499        2500     NA21137             GIH    Gujarati Indian in Houston,TX   \n",
       "2500        2501     NA21141             GIH    Gujarati Indian in Houston,TX   \n",
       "2501        2502     NA21142             GIH    Gujarati Indian in Houston,TX   \n",
       "2502        2503     NA21143             GIH    Gujarati Indian in Houston,TX   \n",
       "2503        2504     NA21144             GIH    Gujarati Indian in Houston,TX   \n",
       "\n",
       "     super_population_code super_population country_code  \\\n",
       "0                      EUR         European          GBR   \n",
       "1                      EUR         European          GBR   \n",
       "2                      EUR         European          GBR   \n",
       "3                      EUR         European          GBR   \n",
       "4                      EUR         European          GBR   \n",
       "...                    ...              ...          ...   \n",
       "2499                   SAS      South Asian          USA   \n",
       "2500                   SAS      South Asian          USA   \n",
       "2501                   SAS      South Asian          USA   \n",
       "2502                   SAS      South Asian          USA   \n",
       "2503                   SAS      South Asian          USA   \n",
       "\n",
       "                       country  gender  \n",
       "0               United Kingdom    male  \n",
       "1               United Kingdom  female  \n",
       "2               United Kingdom  female  \n",
       "3               United Kingdom  female  \n",
       "4               United Kingdom    male  \n",
       "...                        ...     ...  \n",
       "2499  United States of America  female  \n",
       "2500  United States of America  female  \n",
       "2501  United States of America  female  \n",
       "2502  United States of America  female  \n",
       "2503  United States of America  female  \n",
       "\n",
       "[2504 rows x 9 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select * from person_dimension\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Population_Dimension"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Código para ler, filtrar e modelar dados acerca das populações."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exemplo de um dos ficheiros do \"The World Bank\" e sua visualização."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Country Name</th>\n",
       "      <th>Country Code</th>\n",
       "      <th>Indicator Name</th>\n",
       "      <th>Indicator Code</th>\n",
       "      <th>1960</th>\n",
       "      <th>1961</th>\n",
       "      <th>1962</th>\n",
       "      <th>1963</th>\n",
       "      <th>1964</th>\n",
       "      <th>1965</th>\n",
       "      <th>...</th>\n",
       "      <th>2011</th>\n",
       "      <th>2012</th>\n",
       "      <th>2013</th>\n",
       "      <th>2014</th>\n",
       "      <th>2015</th>\n",
       "      <th>2016</th>\n",
       "      <th>2017</th>\n",
       "      <th>2018</th>\n",
       "      <th>2019</th>\n",
       "      <th>Unnamed: 64</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aruba</td>\n",
       "      <td>ABW</td>\n",
       "      <td>GNI per capita, Atlas method (current US$)</td>\n",
       "      <td>NY.GNP.PCAP.CD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>22450.0</td>\n",
       "      <td>23520.0</td>\n",
       "      <td>23970.0</td>\n",
       "      <td>24500.0</td>\n",
       "      <td>23780.0</td>\n",
       "      <td>23550.0</td>\n",
       "      <td>23630.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Afghanistan</td>\n",
       "      <td>AFG</td>\n",
       "      <td>GNI per capita, Atlas method (current US$)</td>\n",
       "      <td>NY.GNP.PCAP.CD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>530.0</td>\n",
       "      <td>630.0</td>\n",
       "      <td>660.0</td>\n",
       "      <td>630.0</td>\n",
       "      <td>600.0</td>\n",
       "      <td>570.0</td>\n",
       "      <td>550.0</td>\n",
       "      <td>550.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Angola</td>\n",
       "      <td>AGO</td>\n",
       "      <td>GNI per capita, Atlas method (current US$)</td>\n",
       "      <td>NY.GNP.PCAP.CD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3420.0</td>\n",
       "      <td>4170.0</td>\n",
       "      <td>4780.0</td>\n",
       "      <td>5010.0</td>\n",
       "      <td>4520.0</td>\n",
       "      <td>3770.0</td>\n",
       "      <td>3560.0</td>\n",
       "      <td>3370.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Albania</td>\n",
       "      <td>ALB</td>\n",
       "      <td>GNI per capita, Atlas method (current US$)</td>\n",
       "      <td>NY.GNP.PCAP.CD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>4410.0</td>\n",
       "      <td>4360.0</td>\n",
       "      <td>4540.0</td>\n",
       "      <td>4540.0</td>\n",
       "      <td>4390.0</td>\n",
       "      <td>4320.0</td>\n",
       "      <td>4290.0</td>\n",
       "      <td>4860.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Andorra</td>\n",
       "      <td>AND</td>\n",
       "      <td>GNI per capita, Atlas method (current US$)</td>\n",
       "      <td>NY.GNP.PCAP.CD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Country Name Country Code                              Indicator Name  \\\n",
       "0        Aruba          ABW  GNI per capita, Atlas method (current US$)   \n",
       "1  Afghanistan          AFG  GNI per capita, Atlas method (current US$)   \n",
       "2       Angola          AGO  GNI per capita, Atlas method (current US$)   \n",
       "3      Albania          ALB  GNI per capita, Atlas method (current US$)   \n",
       "4      Andorra          AND  GNI per capita, Atlas method (current US$)   \n",
       "\n",
       "   Indicator Code  1960  1961  1962  1963  1964  1965  ...     2011     2012  \\\n",
       "0  NY.GNP.PCAP.CD   NaN   NaN   NaN   NaN   NaN   NaN  ...  22450.0  23520.0   \n",
       "1  NY.GNP.PCAP.CD   NaN   NaN   NaN   NaN   NaN   NaN  ...    530.0    630.0   \n",
       "2  NY.GNP.PCAP.CD   NaN   NaN   NaN   NaN   NaN   NaN  ...   3420.0   4170.0   \n",
       "3  NY.GNP.PCAP.CD   NaN   NaN   NaN   NaN   NaN   NaN  ...   4410.0   4360.0   \n",
       "4  NY.GNP.PCAP.CD   NaN   NaN   NaN   NaN   NaN   NaN  ...      NaN      NaN   \n",
       "\n",
       "      2013     2014     2015     2016     2017    2018  2019  Unnamed: 64  \n",
       "0  23970.0  24500.0  23780.0  23550.0  23630.0     NaN   NaN          NaN  \n",
       "1    660.0    630.0    600.0    570.0    550.0   550.0   NaN          NaN  \n",
       "2   4780.0   5010.0   4520.0   3770.0   3560.0  3370.0   NaN          NaN  \n",
       "3   4540.0   4540.0   4390.0   4320.0   4290.0  4860.0   NaN          NaN  \n",
       "4      NaN      NaN      NaN      NaN      NaN     NaN   NaN          NaN  \n",
       "\n",
       "[5 rows x 65 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#step 1. read base data\n",
    "lines=open(\"API_NY.GNP.PCAP.CD_DS2_en_csv_v2_936138.csv\", \"rt\").readlines()\n",
    "\n",
    "#step 2. remove the first 4 lines as it only contain metadata and write all over again\n",
    "#        so that we may benefit from Pandas read_csv function\n",
    "fil=open(\"country_data.csv\", \"wt\")\n",
    "for lin in lines[4:]: fil.write(lin)\n",
    "fil.close()\n",
    "\n",
    "#step 3. read the data for pandas queries\n",
    "df=pd.read_csv(\"country_data.csv\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load the \"Growth National Income\" (in dolars) and the \"Life Expectancy at Birth\" into memory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Load the \"Growth National Income\" (in dolars) and the \"Life Expectancy at Birth\" into memory.\n",
    "\n",
    "import csv\n",
    "\n",
    "def load_csv(csv_file):\n",
    "    '''This function reads a csv file from \"The World Bank\", and extracts the information for a certain year\n",
    "    for all the countries.\n",
    "    Requires: csv file with demographic indexes for all countries.\n",
    "    Ensures: List with countries and respective index for the year of 2015, the year the 1000 genomes project\n",
    "    was published.\n",
    "    '''\n",
    "    \n",
    "    with open(csv_file, newline='') as csv_file:\n",
    "        csv_reader = csv.reader(csv_file, delimiter=\",\")\n",
    "        data = list(csv_reader)\n",
    "\n",
    "        countryList = []\n",
    "        i = 1\n",
    "        \n",
    "        for country in data[5:]:\n",
    "            key = i\n",
    "            cntry = country[0]\n",
    "            countryCode = country[1]\n",
    "            GNI2015 = country[59]\n",
    "                        \n",
    "            i +=1\n",
    "            \n",
    "            countryList.append([key, cntry, countryCode, GNI2015])\n",
    "   \n",
    "    return countryList\n",
    "\n",
    "GNI = load_csv('API_NY.GNP.PCAP.CD_DS2_en_csv_v2_936138.csv')\n",
    "LEB2015 = load_csv('API_SP.DYN.LE00.IN_DS2_en_csv_v2_935946.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "# Define age groups fo life expectancy\n",
    "def ExpectancyClass(expec):\n",
    "    if expec <=60  : return \"Below 60yo\"\n",
    "    if expec <=75 : return \"60yo - 75yo\"\n",
    "    return \"Above 75\"\n",
    "    \n",
    "# Define gross national income per capita per GNI brackets \n",
    "def IncomeClass(gni_pc):\n",
    "    if gni_pc <=1000 : return \"Below 1k\"\n",
    "    if gni_pc <=4000  : return \"1k - 4k\"\n",
    "    if gni_pc <=14000  : return \"4k - 14k\"\n",
    "    return \"Above 14k\"\n",
    "\n",
    "def load_csv(csv_file):\n",
    "    \n",
    "    with open(csv_file, newline='') as csv_file:\n",
    "        csv_reader = csv.reader(csv_file, delimiter=\"\\t\")\n",
    "        data = list(csv_reader)\n",
    "\n",
    "        population = []\n",
    "        i = 1\n",
    "        \n",
    "        for pop in data[1:27]:\n",
    "            key = i\n",
    "            popCode = pop[0]\n",
    "            popDescript = pop[1]\n",
    "            superPopCode = pop[2]\n",
    "            superPop = pop[3]\n",
    "            ethnicity = pop[4]\n",
    "            countryCode = pop[5]\n",
    "            country = pop[6]\n",
    "            region = pop[7]\n",
    "            city = pop[8]\n",
    "            for ctry in GNI:\n",
    "                if countryCode == ctry[2]:\n",
    "                    grossNationalIncome = float(ctry[3])\n",
    "            for ctry in LEB2015:\n",
    "                if countryCode == ctry[2]:\n",
    "                    lifeExpectancyBirth = round(float(ctry[3]), 2)\n",
    "            i += 1\n",
    "            \n",
    "            population.append([key, popCode, popDescript, superPopCode, superPop, ethnicity, countryCode, country, region, city, grossNationalIncome, IncomeClass(grossNationalIncome), lifeExpectancyBirth, ExpectancyClass(lifeExpectancyBirth)\n",
    "            ])\n",
    "         \n",
    "    return population\n",
    "\n",
    "population = (load_csv('20131219.populations - 20131219.populations.tsv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inserção e verificação dos dados na Population_Dimension"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SQL = \"\"\"drop table population_dimension\"\"\"\n",
    "SQL= \"\"\"create table population_dimension(\n",
    "\tpopulation_key int,\n",
    "\tpopulation_code char(3),\n",
    "\tpopulation_description varchar(128), \n",
    "\tsuper_population_code char(3),\n",
    "\tsuper_population varchar(128),\n",
    "    ethnicity varchar(32),\n",
    "    country_code char(3),\n",
    "    country varchar(128),\n",
    "    region varchar(128),\n",
    "    city varchar(128),\n",
    "    gross_national_income int,\n",
    "    gross_national_income_group varchar(24),\n",
    "    life_expectancy_birth float,\n",
    "    life_expectancy_birth_group varchar(24),\n",
    "    primary key (population_key)\n",
    ")\"\"\"\n",
    "\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = pg.connect(host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "curs = conn.cursor()\n",
    "for pop in population:\n",
    "    key, popCode, popDescript, superPopCode, superPop, ethnicity, countryCode, country, region, city, grossNationalIncome, IncomeClass, lifeExpectancyBirth, ExpectancyClass = pop\n",
    "    curs.execute(\"insert into population_dimension values(%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\", (key, popCode, popDescript, superPopCode, superPop, ethnicity, countryCode, country, region, city, grossNationalIncome, IncomeClass, lifeExpectancyBirth, ExpectancyClass))\n",
    "curs.close()\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population_key</th>\n",
       "      <th>population_code</th>\n",
       "      <th>population_description</th>\n",
       "      <th>super_population_code</th>\n",
       "      <th>super_population</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>country_code</th>\n",
       "      <th>country</th>\n",
       "      <th>region</th>\n",
       "      <th>city</th>\n",
       "      <th>gross_national_income</th>\n",
       "      <th>gross_national_income_group</th>\n",
       "      <th>life_expectancy_birth</th>\n",
       "      <th>life_expectancy_birth_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>CDX</td>\n",
       "      <td>Chinese Dai in Xishuangbanna, China</td>\n",
       "      <td>EAS</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>Dai</td>\n",
       "      <td>CHN</td>\n",
       "      <td>China</td>\n",
       "      <td>Xishuangbanna</td>\n",
       "      <td></td>\n",
       "      <td>7910</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>75.93</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>CHB</td>\n",
       "      <td>Han Chinese in Beijing, China</td>\n",
       "      <td>EAS</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>Han</td>\n",
       "      <td>CHN</td>\n",
       "      <td>China</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>Beijing</td>\n",
       "      <td>7910</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>75.93</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>JPT</td>\n",
       "      <td>Japanese in Tokyo, Japan</td>\n",
       "      <td>EAS</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>Japanese</td>\n",
       "      <td>JPN</td>\n",
       "      <td>Japan</td>\n",
       "      <td>Kantō</td>\n",
       "      <td>Tokyo</td>\n",
       "      <td>38840</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>83.79</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>KHV</td>\n",
       "      <td>Kinh in Ho Chi Minh City, Vietnam</td>\n",
       "      <td>EAS</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>Kinh</td>\n",
       "      <td>VNM</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Ho Chi Minh</td>\n",
       "      <td>Ho Chi Minh</td>\n",
       "      <td>1970</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>75.11</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>CHS</td>\n",
       "      <td>Southern Han Chinese, China</td>\n",
       "      <td>EAS</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>Han</td>\n",
       "      <td>CHN</td>\n",
       "      <td>China</td>\n",
       "      <td>Southern China</td>\n",
       "      <td></td>\n",
       "      <td>7910</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>75.93</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>BEB</td>\n",
       "      <td>Bengali in Bangladesh</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>Bengali</td>\n",
       "      <td>BGD</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td></td>\n",
       "      <td>1220</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>71.51</td>\n",
       "      <td>60yo - 75yo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>GIH</td>\n",
       "      <td>Gujarati Indian in Houston,TX</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>Gujarati</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Houston, Texas</td>\n",
       "      <td>Houston</td>\n",
       "      <td>56720</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>78.69</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>ITU</td>\n",
       "      <td>Indian Telugu in the UK</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>Telugu</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td></td>\n",
       "      <td>44280</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>80.96</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>PJL</td>\n",
       "      <td>Punjabi in Lahore,Pakistan</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>Punjabi</td>\n",
       "      <td>PAK</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>Punjab</td>\n",
       "      <td>Lahore</td>\n",
       "      <td>1360</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>66.58</td>\n",
       "      <td>60yo - 75yo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>STU</td>\n",
       "      <td>Sri Lankan Tamil in the UK</td>\n",
       "      <td>SAS</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>Tamil</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td></td>\n",
       "      <td>44280</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>80.96</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>ASW</td>\n",
       "      <td>African Ancestry in Southwest US</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African</td>\n",
       "      <td>African</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Southwest United States of America</td>\n",
       "      <td></td>\n",
       "      <td>56720</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>78.69</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>ACB</td>\n",
       "      <td>African Caribbean in Barbados</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African</td>\n",
       "      <td>Caribbean</td>\n",
       "      <td>BRB</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>Barbados</td>\n",
       "      <td></td>\n",
       "      <td>15270</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>78.80</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>ESN</td>\n",
       "      <td>Esan in Nigeria</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African</td>\n",
       "      <td>Esan</td>\n",
       "      <td>NGA</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td></td>\n",
       "      <td>2880</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>53.11</td>\n",
       "      <td>Below 60yo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>GWD</td>\n",
       "      <td>Gambian in Western Division, The Gambia</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African</td>\n",
       "      <td>West Gambien</td>\n",
       "      <td>GMB</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>Western Division</td>\n",
       "      <td></td>\n",
       "      <td>620</td>\n",
       "      <td>Below 1k</td>\n",
       "      <td>60.91</td>\n",
       "      <td>60yo - 75yo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>LWK</td>\n",
       "      <td>Luhya in Webuye, Kenya</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African</td>\n",
       "      <td>Luhya</td>\n",
       "      <td>KEN</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>Bungoma County</td>\n",
       "      <td>Webuye</td>\n",
       "      <td>1290</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>64.80</td>\n",
       "      <td>60yo - 75yo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>MSL</td>\n",
       "      <td>Mende in Sierra Leone</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African</td>\n",
       "      <td>Mende</td>\n",
       "      <td>SLE</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td></td>\n",
       "      <td>550</td>\n",
       "      <td>Below 1k</td>\n",
       "      <td>52.94</td>\n",
       "      <td>Below 60yo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>YRI</td>\n",
       "      <td>Yoruba in Ibadan, Nigeria</td>\n",
       "      <td>AFR</td>\n",
       "      <td>African</td>\n",
       "      <td>Yoruba</td>\n",
       "      <td>NGA</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>Oyo</td>\n",
       "      <td>Ibadan</td>\n",
       "      <td>2880</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>53.11</td>\n",
       "      <td>Below 60yo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>GBR</td>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>British</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td></td>\n",
       "      <td>44280</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>80.96</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finnish in Finland</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>Finnish</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finland</td>\n",
       "      <td>Finland</td>\n",
       "      <td></td>\n",
       "      <td>47160</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>81.48</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>IBS</td>\n",
       "      <td>Iberian populations in Spain</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>Iberian</td>\n",
       "      <td>ESP</td>\n",
       "      <td>Spain</td>\n",
       "      <td>Spain</td>\n",
       "      <td></td>\n",
       "      <td>28430</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>82.83</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>TSI</td>\n",
       "      <td>Toscani in Italy</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>Tuscan</td>\n",
       "      <td>ITA</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Toscany</td>\n",
       "      <td></td>\n",
       "      <td>32980</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>82.54</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>CEU</td>\n",
       "      <td>Utah residents with Northern and Western Europ...</td>\n",
       "      <td>EUR</td>\n",
       "      <td>European</td>\n",
       "      <td>Northern and Western European</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Utah</td>\n",
       "      <td></td>\n",
       "      <td>56720</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>78.69</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>CLM</td>\n",
       "      <td>Colombian in Medellin, Colombia</td>\n",
       "      <td>AMR</td>\n",
       "      <td>American</td>\n",
       "      <td>Colombian</td>\n",
       "      <td>COL</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>Aburrá Valley</td>\n",
       "      <td>Medellin</td>\n",
       "      <td>7330</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>76.53</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>MXL</td>\n",
       "      <td>Mexican Ancestry in Los Angeles, California</td>\n",
       "      <td>AMR</td>\n",
       "      <td>American</td>\n",
       "      <td>Mexican</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>California</td>\n",
       "      <td>Los Angeles</td>\n",
       "      <td>56720</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>78.69</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>PEL</td>\n",
       "      <td>Peruvian in Lima, Peru</td>\n",
       "      <td>AMR</td>\n",
       "      <td>American</td>\n",
       "      <td>Peruvian</td>\n",
       "      <td>PER</td>\n",
       "      <td>Peru</td>\n",
       "      <td>Lima</td>\n",
       "      <td>Lima</td>\n",
       "      <td>6340</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>75.79</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26</td>\n",
       "      <td>PUR</td>\n",
       "      <td>Puerto Rican in Puerto Rico</td>\n",
       "      <td>AMR</td>\n",
       "      <td>American</td>\n",
       "      <td>Puerto Rican</td>\n",
       "      <td>PRI</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td></td>\n",
       "      <td>19420</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>79.35</td>\n",
       "      <td>Above 75</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    population_key population_code  \\\n",
       "0                1             CDX   \n",
       "1                2             CHB   \n",
       "2                3             JPT   \n",
       "3                4             KHV   \n",
       "4                5             CHS   \n",
       "5                6             BEB   \n",
       "6                7             GIH   \n",
       "7                8             ITU   \n",
       "8                9             PJL   \n",
       "9               10             STU   \n",
       "10              11             ASW   \n",
       "11              12             ACB   \n",
       "12              13             ESN   \n",
       "13              14             GWD   \n",
       "14              15             LWK   \n",
       "15              16             MSL   \n",
       "16              17             YRI   \n",
       "17              18             GBR   \n",
       "18              19             FIN   \n",
       "19              20             IBS   \n",
       "20              21             TSI   \n",
       "21              22             CEU   \n",
       "22              23             CLM   \n",
       "23              24             MXL   \n",
       "24              25             PEL   \n",
       "25              26             PUR   \n",
       "\n",
       "                               population_description super_population_code  \\\n",
       "0                 Chinese Dai in Xishuangbanna, China                   EAS   \n",
       "1                       Han Chinese in Beijing, China                   EAS   \n",
       "2                            Japanese in Tokyo, Japan                   EAS   \n",
       "3                   Kinh in Ho Chi Minh City, Vietnam                   EAS   \n",
       "4                         Southern Han Chinese, China                   EAS   \n",
       "5                               Bengali in Bangladesh                   SAS   \n",
       "6                       Gujarati Indian in Houston,TX                   SAS   \n",
       "7                             Indian Telugu in the UK                   SAS   \n",
       "8                          Punjabi in Lahore,Pakistan                   SAS   \n",
       "9                          Sri Lankan Tamil in the UK                   SAS   \n",
       "10                   African Ancestry in Southwest US                   AFR   \n",
       "11                      African Caribbean in Barbados                   AFR   \n",
       "12                                    Esan in Nigeria                   AFR   \n",
       "13            Gambian in Western Division, The Gambia                   AFR   \n",
       "14                             Luhya in Webuye, Kenya                   AFR   \n",
       "15                              Mende in Sierra Leone                   AFR   \n",
       "16                          Yoruba in Ibadan, Nigeria                   AFR   \n",
       "17                    British in England and Scotland                   EUR   \n",
       "18                                 Finnish in Finland                   EUR   \n",
       "19                       Iberian populations in Spain                   EUR   \n",
       "20                                   Toscani in Italy                   EUR   \n",
       "21  Utah residents with Northern and Western Europ...                   EUR   \n",
       "22                    Colombian in Medellin, Colombia                   AMR   \n",
       "23        Mexican Ancestry in Los Angeles, California                   AMR   \n",
       "24                             Peruvian in Lima, Peru                   AMR   \n",
       "25                        Puerto Rican in Puerto Rico                   AMR   \n",
       "\n",
       "   super_population                      ethnicity country_code  \\\n",
       "0        East Asian                            Dai          CHN   \n",
       "1        East Asian                            Han          CHN   \n",
       "2        East Asian                       Japanese          JPN   \n",
       "3        East Asian                           Kinh          VNM   \n",
       "4        East Asian                            Han          CHN   \n",
       "5       South Asian                        Bengali          BGD   \n",
       "6       South Asian                       Gujarati          USA   \n",
       "7       South Asian                         Telugu          GBR   \n",
       "8       South Asian                        Punjabi          PAK   \n",
       "9       South Asian                          Tamil          GBR   \n",
       "10          African                        African          USA   \n",
       "11          African                      Caribbean          BRB   \n",
       "12          African                           Esan          NGA   \n",
       "13          African                   West Gambien          GMB   \n",
       "14          African                         Luhya           KEN   \n",
       "15          African                          Mende          SLE   \n",
       "16          African                         Yoruba          NGA   \n",
       "17         European                        British          GBR   \n",
       "18         European                        Finnish          FIN   \n",
       "19         European                        Iberian          ESP   \n",
       "20         European                         Tuscan          ITA   \n",
       "21         European  Northern and Western European          USA   \n",
       "22         American                      Colombian          COL   \n",
       "23         American                        Mexican          USA   \n",
       "24         American                       Peruvian          PER   \n",
       "25         American                   Puerto Rican          PRI   \n",
       "\n",
       "                     country                              region         city  \\\n",
       "0                      China                       Xishuangbanna                \n",
       "1                      China                             Beijing      Beijing   \n",
       "2                      Japan                               Kantō        Tokyo   \n",
       "3                    Vietnam                         Ho Chi Minh  Ho Chi Minh   \n",
       "4                      China                      Southern China                \n",
       "5                 Bangladesh                          Bangladesh                \n",
       "6   United States of America                      Houston, Texas      Houston   \n",
       "7             United Kingdom                      United Kingdom                \n",
       "8                   Pakistan                              Punjab       Lahore   \n",
       "9             United Kingdom                      United Kingdom                \n",
       "10  United States of America  Southwest United States of America                \n",
       "11                  Barbados                            Barbados                \n",
       "12                   Nigeria                             Nigeria                \n",
       "13                The Gambia                    Western Division                \n",
       "14                     Kenya                      Bungoma County       Webuye   \n",
       "15              Sierra Leone                        Sierra Leone                \n",
       "16                   Nigeria                                 Oyo       Ibadan   \n",
       "17            United Kingdom                      United Kingdom                \n",
       "18                   Finland                             Finland                \n",
       "19                     Spain                               Spain                \n",
       "20                     Italy                             Toscany                \n",
       "21  United States of America                                Utah                \n",
       "22                  Colombia                       Aburrá Valley     Medellin   \n",
       "23  United States of America                          California  Los Angeles   \n",
       "24                      Peru                                Lima         Lima   \n",
       "25               Puerto Rico                         Puerto Rico                \n",
       "\n",
       "    gross_national_income gross_national_income_group  life_expectancy_birth  \\\n",
       "0                    7910                    4k - 14k                  75.93   \n",
       "1                    7910                    4k - 14k                  75.93   \n",
       "2                   38840                   Above 14k                  83.79   \n",
       "3                    1970                     1k - 4k                  75.11   \n",
       "4                    7910                    4k - 14k                  75.93   \n",
       "5                    1220                     1k - 4k                  71.51   \n",
       "6                   56720                   Above 14k                  78.69   \n",
       "7                   44280                   Above 14k                  80.96   \n",
       "8                    1360                     1k - 4k                  66.58   \n",
       "9                   44280                   Above 14k                  80.96   \n",
       "10                  56720                   Above 14k                  78.69   \n",
       "11                  15270                   Above 14k                  78.80   \n",
       "12                   2880                     1k - 4k                  53.11   \n",
       "13                    620                    Below 1k                  60.91   \n",
       "14                   1290                     1k - 4k                  64.80   \n",
       "15                    550                    Below 1k                  52.94   \n",
       "16                   2880                     1k - 4k                  53.11   \n",
       "17                  44280                   Above 14k                  80.96   \n",
       "18                  47160                   Above 14k                  81.48   \n",
       "19                  28430                   Above 14k                  82.83   \n",
       "20                  32980                   Above 14k                  82.54   \n",
       "21                  56720                   Above 14k                  78.69   \n",
       "22                   7330                    4k - 14k                  76.53   \n",
       "23                  56720                   Above 14k                  78.69   \n",
       "24                   6340                    4k - 14k                  75.79   \n",
       "25                  19420                   Above 14k                  79.35   \n",
       "\n",
       "   life_expectancy_birth_group  \n",
       "0                     Above 75  \n",
       "1                     Above 75  \n",
       "2                     Above 75  \n",
       "3                     Above 75  \n",
       "4                     Above 75  \n",
       "5                  60yo - 75yo  \n",
       "6                     Above 75  \n",
       "7                     Above 75  \n",
       "8                  60yo - 75yo  \n",
       "9                     Above 75  \n",
       "10                    Above 75  \n",
       "11                    Above 75  \n",
       "12                  Below 60yo  \n",
       "13                 60yo - 75yo  \n",
       "14                 60yo - 75yo  \n",
       "15                  Below 60yo  \n",
       "16                  Below 60yo  \n",
       "17                    Above 75  \n",
       "18                    Above 75  \n",
       "19                    Above 75  \n",
       "20                    Above 75  \n",
       "21                    Above 75  \n",
       "22                    Above 75  \n",
       "23                    Above 75  \n",
       "24                    Above 75  \n",
       "25                    Above 75  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select * from population_dimension\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gender_Dimension"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Código para ler, filtrar e modelar dados demográficos relativos ao género ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "def load_csv(csv_file):\n",
    "    '''This function reads a csv file from \"The World Bank\", and extracts the information for a certain year\n",
    "    for all the countries.\n",
    "    Requires: csv file with demographic indexes for all countries.\n",
    "    Ensures: List with countries and respective index for the year of 2015, the year the 1000 genomes project\n",
    "    was published.\n",
    "    '''\n",
    "    \n",
    "    with open(csv_file , newline='') as csv_file:\n",
    "        csv_reader = csv.reader(csv_file, delimiter=\",\")\n",
    "        data = list(csv_reader)\n",
    "\n",
    "        countryList = []\n",
    "        i = 1        \n",
    "        for country in data[5:]:\n",
    "            key = i\n",
    "            cntry = country[0]\n",
    "            countryCode = country[1]\n",
    "            LT2015 = country[59]\n",
    "                        \n",
    "            i +=1\n",
    "            \n",
    "            countryList.append([key, cntry, countryCode, LT2015])\n",
    "\n",
    "    return countryList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Literacy rate for male and females per country (% of male and females, ages 15-24, for the year of 2015).\n",
    "LTM = load_csv(\"API_SE.ADT.LT.MA.ZS_DS2_en_csv_v2_1000301.csv\")\n",
    "LTF = load_csv(\"API_SE.ADT.1524.LT.FE.ZS_DS2_en_csv_v2_989948.csv\")\n",
    "\n",
    "# Unemployement rate for male and female per country s for the year of 2015.\n",
    "UNEMM = load_csv(\"API_SL.UNEM.TOTL.MA.ZS_DS2_en_csv_v2_995787.csv\")\n",
    "UNEMF = load_csv(\"API_SL.UNEM.TOTL.FE.ZS_DS2_en_csv_v2_1002674.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_csv(csv_file):\n",
    "    '''This function reads a csv file from \"The World Bank\" with mortality indexes.It extracts the information for a certain year\n",
    "    for all the countries.\n",
    "    Requires: csv file with demographic indexes for all countries.\n",
    "    Ensures: List with countries and respective index for the year of 2015, the year the 1000 genomes project\n",
    "    was published.\n",
    "    '''\n",
    "    \n",
    "    with open(csv_file , newline='') as csv_file:\n",
    "        csv_reader = csv.reader(csv_file, delimiter=\",\")\n",
    "        data = list(csv_reader)\n",
    "        #print(data)\n",
    "\n",
    "        countryList = []\n",
    "        i = 1        \n",
    "        for country in data[5:]:\n",
    "            key = i\n",
    "            cntry = country[2]\n",
    "            countryCode = country[3]\n",
    "            MOR = country[11]\n",
    "                        \n",
    "            i +=1\n",
    "            \n",
    "            countryList.append([key, cntry, countryCode, MOR])\n",
    "        \n",
    "    return countryList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#MORTALITY  rate female/male per 1000 habitants, for the year of 2015.\n",
    "MORTM=load_csv(\"Mort_MA-1f90-45b0-992b-3231bcd2c05b_Data.csv\")\n",
    "MORTF=load_csv(\"Mort_FE-020e-4949-8810-d261d40a218a_Data.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "# This function identifies the countries found in our population dataset and creates a list with the countries found.\n",
    "# This list will be used to create the information to upload to the Gender_Dimension.\n",
    "def load_csv(csv_file):\n",
    "    '''This function identifies the countries found in our population dataset and creates a list with the countries found.\n",
    "    This list contains information, per country, of several indexes for male and females.\n",
    "    This list will be used to create the information to upload to the Gender_Dimension.\n",
    "    Requires: A csv file with information about population and lists (previously loaded), with information about several\n",
    "    demographic indexes.\n",
    "    Ensures: List of demographic indexes per sex, per country.\n",
    "    '''\n",
    "    \n",
    "    with open(csv_file, newline='') as csv_file:\n",
    "        csv_reader = csv.reader(csv_file, delimiter=\"\\t\")\n",
    "        data = list(csv_reader)\n",
    "\n",
    "        countries = []\n",
    "        \n",
    "        for pop in data[1:27]:\n",
    "            countryCode = pop[5]\n",
    "            country = pop[6]\n",
    "            \n",
    "            if [countryCode, country] not in countries:\n",
    "                countries.append([countryCode, country])\n",
    "        \n",
    "        countries.sort()\n",
    "    \n",
    "    gender = []\n",
    "    \n",
    "    i = 1\n",
    "    for country in countries:\n",
    "        keyMale = i\n",
    "        keyFemale = i+1\n",
    "        countryCode = country[0]\n",
    "        country = country[1]\n",
    "        for lit in LTM:\n",
    "            if countryCode == lit[2]:\n",
    "                if lit[3] != '':\n",
    "                    maleLiteracy = float(lit[3])\n",
    "                else:\n",
    "                    maleLiteracy = 0                    \n",
    "        for lit in LTF:\n",
    "            if countryCode == lit[2]:\n",
    "                if lit[3] != '':\n",
    "                    femaleLiteracy = float(lit[3])\n",
    "                else:\n",
    "                    femaleLiteracy = 0\n",
    "        for unemp in UNEMM:\n",
    "            if countryCode == unemp[2]:\n",
    "                if unemp[3] != '':\n",
    "                    maleUnemployement = float(unemp[3])\n",
    "                else:\n",
    "                    maleUnemployement = 0                \n",
    "        for unemp in UNEMF:\n",
    "            if countryCode == unemp[2]:\n",
    "                if unemp[3] != '':\n",
    "                    femaleUnemployement = float(unemp[3])\n",
    "                else:\n",
    "                    femaleUnemployement = 0 \n",
    "        for mort in MORTM:\n",
    "            if countryCode == mort[2]:\n",
    "                if mort[3] != '' and mort[3] != '..':\n",
    "                    maleMortality = float(mort[3])\n",
    "                else:\n",
    "                    maleMortality = 0 \n",
    "        for mort in MORTF:\n",
    "            if countryCode == mort[2]:\n",
    "                if mort[3] != '' and mort[3] != '..':\n",
    "                    femaleMortality = float(mort[3])\n",
    "                else:\n",
    "                    femaleMortality = 0\n",
    "                    \n",
    "        i += 2\n",
    "\n",
    "        gender.append([keyMale, countryCode, country, 'male', round(maleLiteracy, 2), round(maleUnemployement, 2), round(maleMortality, 2)])\n",
    "        gender.append([keyFemale, countryCode, country, 'female', round(femaleLiteracy, 2), round(femaleUnemployement, 2), round(femaleMortality, 2)])\n",
    "        \n",
    "    return gender\n",
    "\n",
    "gender = load_csv('20131219.populations - 20131219.populations.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inserção e verificação dos dados na Gender_Dimension"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SQL = \"\"\"drop table gender_dimension\"\"\"\n",
    "SQL= \"\"\"create table gender_dimension(\n",
    "\tgender_key int,\n",
    "    country_code char(3),\n",
    "    country varchar(128),\n",
    "    gender varchar(6),\n",
    "    literacy_rate float,\n",
    "    unemployment_rate float,\n",
    "    mortality_per_1000_habitants float,\n",
    "    primary key (gender_key)\n",
    ")\"\"\"\n",
    "\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = pg.connect(host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "curs = conn.cursor()\n",
    "for gen in gender:\n",
    "    key, gender, countryCode, country, literacy, unemployment, mortality_rate_1000_habi = gen\n",
    "    curs.execute(\"insert into gender_dimension values(%s, %s, %s, %s, %s, %s, %s)\", (key, gender, countryCode, country, literacy, unemployment, mortality_rate_1000_habi))\n",
    "curs.close()\n",
    "conn.commit()\n",
    "conn.close()\n",
    "\n",
    "# As there are lots of values for literacy_rate missing from \"The World Bank\" data, we will transform these values, that were inserted as 0.00, into NULL values.\n",
    "conn = pg.connect(host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "curs = conn.cursor()\n",
    "curs.execute(\"UPDATE gender_dimension SET literacy_rate = NULL WHERE literacy_rate = 0.00\")\n",
    "curs.close()\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gender_key</th>\n",
       "      <th>country_code</th>\n",
       "      <th>country</th>\n",
       "      <th>gender</th>\n",
       "      <th>literacy_rate</th>\n",
       "      <th>unemployment_rate</th>\n",
       "      <th>mortality_per_1000_habitants</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>BGD</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>male</td>\n",
       "      <td>68.06</td>\n",
       "      <td>3.16</td>\n",
       "      <td>150.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>BGD</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>female</td>\n",
       "      <td>89.54</td>\n",
       "      <td>7.47</td>\n",
       "      <td>109.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>BRB</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.35</td>\n",
       "      <td>124.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>BRB</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.32</td>\n",
       "      <td>74.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>CHN</td>\n",
       "      <td>China</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.11</td>\n",
       "      <td>100.07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>CHN</td>\n",
       "      <td>China</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.01</td>\n",
       "      <td>64.48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>COL</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>male</td>\n",
       "      <td>94.06</td>\n",
       "      <td>6.36</td>\n",
       "      <td>190.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>COL</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>female</td>\n",
       "      <td>98.92</td>\n",
       "      <td>10.84</td>\n",
       "      <td>89.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>ESP</td>\n",
       "      <td>Spain</td>\n",
       "      <td>male</td>\n",
       "      <td>98.77</td>\n",
       "      <td>20.78</td>\n",
       "      <td>77.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>ESP</td>\n",
       "      <td>Spain</td>\n",
       "      <td>female</td>\n",
       "      <td>99.69</td>\n",
       "      <td>23.55</td>\n",
       "      <td>38.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finland</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.88</td>\n",
       "      <td>94.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>FIN</td>\n",
       "      <td>Finland</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.84</td>\n",
       "      <td>43.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.45</td>\n",
       "      <td>84.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.12</td>\n",
       "      <td>54.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>GMB</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>male</td>\n",
       "      <td>61.77</td>\n",
       "      <td>6.76</td>\n",
       "      <td>289.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>GMB</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>female</td>\n",
       "      <td>64.45</td>\n",
       "      <td>12.58</td>\n",
       "      <td>234.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>ITA</td>\n",
       "      <td>Italy</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.31</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>ITA</td>\n",
       "      <td>Italy</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.69</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>JPN</td>\n",
       "      <td>Japan</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.62</td>\n",
       "      <td>69.98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>JPN</td>\n",
       "      <td>Japan</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.11</td>\n",
       "      <td>37.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>KEN</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.60</td>\n",
       "      <td>264.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>KEN</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.00</td>\n",
       "      <td>183.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>NGA</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.66</td>\n",
       "      <td>374.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>NGA</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.11</td>\n",
       "      <td>337.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>PAK</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.83</td>\n",
       "      <td>178.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26</td>\n",
       "      <td>PAK</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.18</td>\n",
       "      <td>141.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>27</td>\n",
       "      <td>PER</td>\n",
       "      <td>Peru</td>\n",
       "      <td>male</td>\n",
       "      <td>97.11</td>\n",
       "      <td>3.03</td>\n",
       "      <td>153.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>28</td>\n",
       "      <td>PER</td>\n",
       "      <td>Peru</td>\n",
       "      <td>female</td>\n",
       "      <td>98.77</td>\n",
       "      <td>2.96</td>\n",
       "      <td>94.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>29</td>\n",
       "      <td>PRI</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13.72</td>\n",
       "      <td>151.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>30</td>\n",
       "      <td>PRI</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.65</td>\n",
       "      <td>65.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>31</td>\n",
       "      <td>SLE</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.45</td>\n",
       "      <td>402.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>32</td>\n",
       "      <td>SLE</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.81</td>\n",
       "      <td>392.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>33</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.37</td>\n",
       "      <td>134.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>34</td>\n",
       "      <td>USA</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.18</td>\n",
       "      <td>80.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>35</td>\n",
       "      <td>VNM</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>male</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.24</td>\n",
       "      <td>182.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>36</td>\n",
       "      <td>VNM</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.99</td>\n",
       "      <td>66.44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gender_key country_code                   country  gender  literacy_rate  \\\n",
       "0            1          BGD                Bangladesh    male          68.06   \n",
       "1            2          BGD                Bangladesh  female          89.54   \n",
       "2            3          BRB                  Barbados    male            NaN   \n",
       "3            4          BRB                  Barbados  female            NaN   \n",
       "4            5          CHN                     China    male            NaN   \n",
       "5            6          CHN                     China  female            NaN   \n",
       "6            7          COL                  Colombia    male          94.06   \n",
       "7            8          COL                  Colombia  female          98.92   \n",
       "8            9          ESP                     Spain    male          98.77   \n",
       "9           10          ESP                     Spain  female          99.69   \n",
       "10          11          FIN                   Finland    male            NaN   \n",
       "11          12          FIN                   Finland  female            NaN   \n",
       "12          13          GBR            United Kingdom    male            NaN   \n",
       "13          14          GBR            United Kingdom  female            NaN   \n",
       "14          15          GMB                The Gambia    male          61.77   \n",
       "15          16          GMB                The Gambia  female          64.45   \n",
       "16          17          ITA                     Italy    male            NaN   \n",
       "17          18          ITA                     Italy  female            NaN   \n",
       "18          19          JPN                     Japan    male            NaN   \n",
       "19          20          JPN                     Japan  female            NaN   \n",
       "20          21          KEN                     Kenya    male            NaN   \n",
       "21          22          KEN                     Kenya  female            NaN   \n",
       "22          23          NGA                   Nigeria    male            NaN   \n",
       "23          24          NGA                   Nigeria  female            NaN   \n",
       "24          25          PAK                  Pakistan    male            NaN   \n",
       "25          26          PAK                  Pakistan  female            NaN   \n",
       "26          27          PER                      Peru    male          97.11   \n",
       "27          28          PER                      Peru  female          98.77   \n",
       "28          29          PRI               Puerto Rico    male            NaN   \n",
       "29          30          PRI               Puerto Rico  female            NaN   \n",
       "30          31          SLE              Sierra Leone    male            NaN   \n",
       "31          32          SLE              Sierra Leone  female            NaN   \n",
       "32          33          USA  United States of America    male            NaN   \n",
       "33          34          USA  United States of America  female            NaN   \n",
       "34          35          VNM                   Vietnam    male            NaN   \n",
       "35          36          VNM                   Vietnam  female            NaN   \n",
       "\n",
       "    unemployment_rate  mortality_per_1000_habitants  \n",
       "0                3.16                        150.04  \n",
       "1                7.47                        109.47  \n",
       "2               12.35                        124.02  \n",
       "3               10.32                         74.17  \n",
       "4                5.11                        100.07  \n",
       "5                4.01                         64.48  \n",
       "6                6.36                        190.84  \n",
       "7               10.84                         89.30  \n",
       "8               20.78                         77.00  \n",
       "9               23.55                         38.81  \n",
       "10               9.88                         94.25  \n",
       "11               8.84                         43.46  \n",
       "12               5.45                         84.74  \n",
       "13               5.12                         54.15  \n",
       "14               6.76                        289.18  \n",
       "15              12.58                        234.24  \n",
       "16              11.31                          0.00  \n",
       "17              12.69                          0.00  \n",
       "18               3.62                         69.98  \n",
       "19               3.11                         37.10  \n",
       "20               2.60                        264.09  \n",
       "21               3.00                        183.12  \n",
       "22               3.66                        374.77  \n",
       "23               5.11                        337.66  \n",
       "24               2.83                        178.54  \n",
       "25               6.18                        141.74  \n",
       "26               3.03                        153.40  \n",
       "27               2.96                         94.74  \n",
       "28              13.72                        151.23  \n",
       "29               9.65                         65.04  \n",
       "30               5.45                        402.80  \n",
       "31               3.81                        392.95  \n",
       "32               5.37                        134.26  \n",
       "33               5.18                         80.39  \n",
       "34               2.24                        182.66  \n",
       "35               1.99                         66.44  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select * from gender_dimension order by gender_key\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Allele_Dimension"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Esta pequena dimensão \"role-playing\", vai servir para interpretar o valor dos genótipos de cada indivíduo, em cada SNP, em cada linha de facto.\n",
    "- Após inserção dos factos na tabela SNP Person Fact, os valores de genotipagem para os diferentes alelos de 0 ou 1, serão interpretados como \"allele_is_reference\" para valores de 0, ou \"allele_is_alternative\", para valores de 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SQL = \"\"\"drop table allele_dimension\"\"\"\n",
    "SQL= \"\"\"create table allele_dimension(\n",
    "\tallele_key int,\n",
    "    allele_type_code char(3),\n",
    "    allele_type varchar(12),\n",
    "    primary key (allele_key)\n",
    ")\"\"\"\n",
    "\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inserção e verificação dos dados na Allele_Dimension\n",
    "- Esta dimensão só tem duas linhas, que identificam o alelo como de referência, ou alternativo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = pg.connect(host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "curs = conn.cursor()\n",
    "curs.execute(\"insert into allele_dimension values (0, 'Ref', 'Reference')\")\n",
    "curs.execute(\"insert into allele_dimension values (1, 'Alt', 'Alternative')\")             \n",
    "curs.close()\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_key</th>\n",
       "      <th>allele_type_code</th>\n",
       "      <th>allele_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Ref</td>\n",
       "      <td>Reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Alt</td>\n",
       "      <td>Alternative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   allele_key allele_type_code  allele_type\n",
       "0           0              Ref    Reference\n",
       "1           1              Alt  Alternative"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select * from allele_dimension\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Disease_Dimension"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Começamos por extrair a informação do Wikidata Query Service. Dada a dimensão dos dados e da query, tivemos que dividir a query em duas sub-queries e depois juntá-las.\n",
    "- Usamos os seguintes comandos SPARQL:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ATENÇÃO: Este comando é para introduzir no Wikidata Query service, encontrado em: https://query.wikidata.org/\n",
    "# Dado a dimensão da Query, tivemos que deixar alguns campos de fora, e dividir a query \n",
    "# Este comando dá-nos todas as doenças com um DOID identifier, mas não conseguimos obter nomes alternativos,  pois o query resultava num timeout.\n",
    "\n",
    "PREFIX dbo: <http://dbpedia.org/ontology/>\n",
    "# disesases\n",
    "SELECT \n",
    "?item  \n",
    "?itemName\n",
    "?DoidID\n",
    "?MeSHID\n",
    "(GROUP_CONCAT(DISTINCT ?subClassLabel ; SEPARATOR = \",\") AS  ?subClass) \n",
    "(GROUP_CONCAT(DISTINCT ?healthspecialtyLabel ; SEPARATOR = \",\") AS  ?healthspecialties) \n",
    "(GROUP_CONCAT(DISTINCT ?symptomLabel ; SEPARATOR = \",\") AS  ?symptoms) \n",
    "(GROUP_CONCAT(DISTINCT ?geneticassociationLabel ; SEPARATOR = \",\") AS  ?geneticassociations) \n",
    "(GROUP_CONCAT(DISTINCT ?afflictLabel ; SEPARATOR = \",\") AS  ?afflicts) \n",
    "(GROUP_CONCAT(DISTINCT ?hasCauseLabel ; SEPARATOR = \",\") AS  ?hasCauses) \n",
    "#(GROUP_CONCAT(DISTINCT(?altLabelLabel); separator = \", \") AS ?altLabel_list) \n",
    "{\n",
    "  ?item wdt:P31 wd:Q12136;\n",
    "        wdt:P699 ?doid;\n",
    "        wdt:P279 ?subClass.\n",
    "  OPTIONAL {?item wdt:P2293 ?geneticassociation.}\n",
    "  OPTIONAL {?item wdt:P486 ?MeSH.}\n",
    "  OPTIONAL {?item wdt:P1995 ?healthspecialty.}\n",
    "  OPTIONAL {?item wdt:P780 ?symptom.}\n",
    "  OPTIONAL {?item wdt:P828 ?hasCause.}\n",
    "  OPTIONAL {?item wdt:P689 ?afflict.}\n",
    "  #OPTIONAL{?item skos:altLabel ?altLabel . FILTER (lang(?altLabel) = \"en\") .}\n",
    "\n",
    "  SERVICE wikibase:label { bd:serviceParam wikibase:language \"[AUTO_LANGUAGE],en\". \n",
    "                         ?MeSH rdfs:label ?MeSHID.\n",
    "                         ?doid rdfs:label ?DoidID.\n",
    "                         ?subClass rdfs:label ?subClassLabel.\n",
    "                         ?healthspecialty rdfs:label ?healthspecialtyLabel.\n",
    "                         ?item rdfs:label ?itemName.\n",
    "                         ?symptom rdfs:label ?symptomLabel.\n",
    "                         ?geneticassociation rdfs:label ?geneticassociationLabel.\n",
    "                         ?afflict rdfs:label ?afflictLabel.\n",
    "                         ?hasCause rdfs:label ?hasCauseLabel.\n",
    "                         #?altLabel rdfs:label ?altLabelLabel.\n",
    "                         }\n",
    "}GROUP BY ?item ?itemName ?MeSHID ?DoidID  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ATENÇÃO: Este comando é para introduzir no Wikidata Query service, encontrado em: https://query.wikidata.org/\n",
    "# Esta query dá todas as doenças com um doid identifier, e nomes alternativos pelos quais podem ser conhecidas.\n",
    "\n",
    "PREFIX dbo: <http://dbpedia.org/ontology/>\n",
    "# disesases\n",
    "SELECT \n",
    "?item \n",
    "?doid\n",
    "?mesh\n",
    "\n",
    "(GROUP_CONCAT(DISTINCT(?altLabelLabel); separator = \", \") AS ?altLabel_list) \n",
    "{\n",
    "  ?item wdt:P31 wd:Q12136;\n",
    "        wdt:P699 ?doid;\n",
    "        wdt:P279 ?subClass.\n",
    "  OPTIONAL{?item wdt:P486 ?mesh.}\n",
    "  OPTIONAL{?item wdt:P4970 ?also.}\n",
    "  OPTIONAL{?item skos:altLabel ?altLabel . FILTER (lang(?altLabel) = \"en\") }\n",
    "  \n",
    "\n",
    "\n",
    "  SERVICE wikibase:label { bd:serviceParam wikibase:language \"[AUTO_LANGUAGE],en\".\n",
    "                         ?altLabel rdfs:label ?altLabelLabel.}\n",
    "         \n",
    "}GROUP BY ?item ?doid ?mesh  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Com os queries em cima, criámos dois ficheiros csv, com informação acerca de doenças humanas. De seguida, processamos estes ficheiros como dataframes, para fundir e filtrar informação.\n",
    "## Atenção, o ficheiro csv final já está criado, pelo que não é necessário fazer este passo de novo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "#Import csv files as DataFrames and reset index to be able to fuse dataframes by index.\n",
    "allDiseases = pd.read_csv('sparQL_query_all_diseases.csv')\n",
    "itemIndexAllDiseases = allDiseases.set_index('item')\n",
    "# itemIndexAllDiseases.head()\n",
    "\n",
    "allDiseasesAltLabel = pd.read_csv('sparQL_query_altLabel.csv')\n",
    "itemIndexAltLabels = allDiseasesAltLabel.set_index('item')\n",
    "# itemIndexAltLabels.head()\n",
    "\n",
    "# Merge the two dataframes by index and reset index.\n",
    "mergedDf = itemIndexAllDiseases.merge(itemIndexAltLabels, left_index=True, right_index=True)\n",
    "mergedAllDiseases = mergedDf.reset_index()\n",
    "# mergedAllDiseases.head()\n",
    "\n",
    "# Filter to remove unwanted columns and duplicates.\n",
    "mergedAllDiseases = mergedAllDiseases.drop(columns=['doid', 'mesh'])\n",
    "mergedAllDiseases = mergedAllDiseases.drop_duplicates()\n",
    "# mergedAllDiseases.head()\n",
    "\n",
    "# Create a cvs file with the filtered information:\n",
    "mergedAllDiseases.to_csv(r'allDiseases.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Código para ler, filtrar e modelar dados acerca de doenças."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "\n",
    "def load_csv(csv_file):\n",
    "    '''This function reads a csv file with disease information compiled from Wikidata and it creates a list of lists with disease data.\n",
    "        Requires: A csv file with a pre-determined structure.\n",
    "        ENsures: Returns a list of list with all the disease data, ready to insert into a dimension table.\n",
    "    '''\n",
    "    \n",
    "    with open(csv_file, newline='', encoding=\"utf8\") as csv_file:\n",
    "        csv_reader = csv.reader(csv_file, delimiter=',')\n",
    "        data = list(csv_reader)\n",
    "        \n",
    "        diseases = []\n",
    "        \n",
    "        id = 0\n",
    "        \n",
    "        new_header = ['Disease_Key','Wikidata', 'Disease_Name', 'Doid_ID', 'MeSH_ID', 'Type_of_Disease', 'Health_Specialities', 'Symptoms', 'Genetic_Association', 'Affects', 'Caused', 'Alternative_Names']\n",
    "        diseases.append(new_header)\n",
    "        \n",
    "        for item in data[1:]:\n",
    "            id += 1\n",
    "            wikidata = item[0].replace('http://www.wikidata.org/entity/','')\n",
    "            name = item[1]\n",
    "            doid = item[2].replace('DOID:','')\n",
    "            mesh = item[3]\n",
    "            subclass = item[4]\n",
    "            healthspecialtie = item[5]\n",
    "            symptoms = item[6]\n",
    "            geneticassociations = '{' + item[7] + '}'\n",
    "            affects = item[8]\n",
    "            hascauses = item[9]\n",
    "            alternativeName = item[10]\n",
    "            \n",
    "            diseases.append([id,wikidata,name,doid,mesh,subclass,healthspecialtie,symptoms,geneticassociations,affects,hascauses,alternativeName])\n",
    "        \n",
    "        return diseases\n",
    "            \n",
    "           \n",
    "diseases = (load_csv(\"allDiseases.csv\"))   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Disease_Key</th>\n",
       "      <th>Wikidata</th>\n",
       "      <th>Disease_Name</th>\n",
       "      <th>Doid_ID</th>\n",
       "      <th>MeSH_ID</th>\n",
       "      <th>Type_of_Disease</th>\n",
       "      <th>Health_Specialities</th>\n",
       "      <th>Symptoms</th>\n",
       "      <th>Genetic_Association</th>\n",
       "      <th>Affects</th>\n",
       "      <th>Caused</th>\n",
       "      <th>Alternative_Names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Q1002195</td>\n",
       "      <td>autosomal recessive limb-girdle muscular dystr...</td>\n",
       "      <td>0110297</td>\n",
       "      <td></td>\n",
       "      <td>autosomal recessive limb-girdle muscular dystr...</td>\n",
       "      <td>neurology</td>\n",
       "      <td></td>\n",
       "      <td>{POMT1}</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LGMD2K, limb-girdle muscular dystrophy-intelle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Q1004647</td>\n",
       "      <td>bullous pemphigoid</td>\n",
       "      <td>8506</td>\n",
       "      <td>D010391</td>\n",
       "      <td>pemphigoid,autoimmune disease of skin and conn...</td>\n",
       "      <td>dermatology</td>\n",
       "      <td></td>\n",
       "      <td>{}</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Bullous pemphigoid, Bullous pemphigoid (disord...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Q1016605</td>\n",
       "      <td>Burkitt lymphoma</td>\n",
       "      <td>8584</td>\n",
       "      <td>D002051</td>\n",
       "      <td>mature B-cell neoplasm</td>\n",
       "      <td>hematology</td>\n",
       "      <td></td>\n",
       "      <td>{MYC}</td>\n",
       "      <td></td>\n",
       "      <td>Epstein Barr virus infection</td>\n",
       "      <td>Burkitt lymphoma/leukaemia, Burkitt's Lymphoma...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Q1017169</td>\n",
       "      <td>Buruli ulcer disease</td>\n",
       "      <td>0050456</td>\n",
       "      <td>D054312</td>\n",
       "      <td>neglected tropical disease,mycobacterium infec...</td>\n",
       "      <td>infectious disease</td>\n",
       "      <td></td>\n",
       "      <td>{}</td>\n",
       "      <td></td>\n",
       "      <td>infection</td>\n",
       "      <td>Buruli ulcer, Bairnsdale ulcer, Daintree ulcer...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Q1018534</td>\n",
       "      <td>progressive familial intrahepatic cholestasis</td>\n",
       "      <td>0070221</td>\n",
       "      <td></td>\n",
       "      <td>liver disease,intrahepatic cholestasis,familia...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>{ABCB11,ABCB4,ATP8B1}</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PFIC; Byler disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Disease_Key  Wikidata                                       Disease_Name  \\\n",
       "0            1  Q1002195  autosomal recessive limb-girdle muscular dystr...   \n",
       "1            2  Q1004647                                 bullous pemphigoid   \n",
       "2            3  Q1016605                                   Burkitt lymphoma   \n",
       "3            4  Q1017169                               Buruli ulcer disease   \n",
       "4            5  Q1018534      progressive familial intrahepatic cholestasis   \n",
       "\n",
       "   Doid_ID  MeSH_ID                                    Type_of_Disease  \\\n",
       "0  0110297           autosomal recessive limb-girdle muscular dystr...   \n",
       "1     8506  D010391  pemphigoid,autoimmune disease of skin and conn...   \n",
       "2     8584  D002051                             mature B-cell neoplasm   \n",
       "3  0050456  D054312  neglected tropical disease,mycobacterium infec...   \n",
       "4  0070221           liver disease,intrahepatic cholestasis,familia...   \n",
       "\n",
       "  Health_Specialities Symptoms    Genetic_Association Affects  \\\n",
       "0           neurology                         {POMT1}           \n",
       "1         dermatology                              {}           \n",
       "2          hematology                           {MYC}           \n",
       "3  infectious disease                              {}           \n",
       "4                               {ABCB11,ABCB4,ATP8B1}           \n",
       "\n",
       "                         Caused  \\\n",
       "0                                 \n",
       "1                                 \n",
       "2  Epstein Barr virus infection   \n",
       "3                     infection   \n",
       "4                                 \n",
       "\n",
       "                                   Alternative_Names  \n",
       "0  LGMD2K, limb-girdle muscular dystrophy-intelle...  \n",
       "1  Bullous pemphigoid, Bullous pemphigoid (disord...  \n",
       "2  Burkitt lymphoma/leukaemia, Burkitt's Lymphoma...  \n",
       "3  Buruli ulcer, Bairnsdale ulcer, Daintree ulcer...  \n",
       "4                                PFIC; Byler disease  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualização da informação que será inserida da Disease_Dimension\n",
    "diseases_DF = pd.DataFrame(diseases[1:],columns=diseases[0])\n",
    "diseases_DF.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Inserção e verificação dos dados na Disease_Dimension."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SQL = \"\"\"drop table disease_dimension\"\"\"\n",
    "SQL= \"\"\"create table disease_dimension(\n",
    "disease_key int,\n",
    "wikidata_id varchar(30),\n",
    "disease_name varchar(160),\n",
    "doid_id int, \n",
    "mesh_id varchar(20),\n",
    "type_of_disease varchar(1760),\n",
    "health_speciality  varchar(160),\n",
    "symptoms varchar(1860),\n",
    "genetic_associations varchar(60) ARRAY,\n",
    "affects varchar(280),\n",
    "caused_by varchar(480),\n",
    "alternative_names varchar(20000),\n",
    "primary key (disease_key)\n",
    ")\"\"\"\n",
    "\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = pg.connect(host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "curs = conn.cursor()\n",
    "for dis in diseases[1:]:\n",
    "    id, wikidata, name, doid, mesh, subclass, healthspecialtie, symptoms, geneticassociations, affects, hascauses, alternativeName = dis\n",
    "    curs.execute(\"insert into disease_dimension values(%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)\", (id, wikidata, name, doid, mesh, subclass, healthspecialtie, symptoms, geneticassociations, affects, hascauses, alternativeName))    \n",
    "curs.close()\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_key</th>\n",
       "      <th>wikidata_id</th>\n",
       "      <th>disease_name</th>\n",
       "      <th>doid_id</th>\n",
       "      <th>mesh_id</th>\n",
       "      <th>type_of_disease</th>\n",
       "      <th>health_speciality</th>\n",
       "      <th>symptoms</th>\n",
       "      <th>genetic_associations</th>\n",
       "      <th>affects</th>\n",
       "      <th>caused_by</th>\n",
       "      <th>alternative_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>Q1002195</td>\n",
       "      <td>autosomal recessive limb-girdle muscular dystr...</td>\n",
       "      <td>110297</td>\n",
       "      <td></td>\n",
       "      <td>autosomal recessive limb-girdle muscular dystr...</td>\n",
       "      <td>neurology</td>\n",
       "      <td></td>\n",
       "      <td>[POMT1]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LGMD2K, limb-girdle muscular dystrophy-intelle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>Q1004647</td>\n",
       "      <td>bullous pemphigoid</td>\n",
       "      <td>8506</td>\n",
       "      <td>D010391</td>\n",
       "      <td>pemphigoid,autoimmune disease of skin and conn...</td>\n",
       "      <td>dermatology</td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Bullous pemphigoid, Bullous pemphigoid (disord...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>Q1016605</td>\n",
       "      <td>Burkitt lymphoma</td>\n",
       "      <td>8584</td>\n",
       "      <td>D002051</td>\n",
       "      <td>mature B-cell neoplasm</td>\n",
       "      <td>hematology</td>\n",
       "      <td></td>\n",
       "      <td>[MYC]</td>\n",
       "      <td></td>\n",
       "      <td>Epstein Barr virus infection</td>\n",
       "      <td>Burkitt lymphoma/leukaemia, Burkitt's Lymphoma...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>Q1017169</td>\n",
       "      <td>Buruli ulcer disease</td>\n",
       "      <td>50456</td>\n",
       "      <td>D054312</td>\n",
       "      <td>neglected tropical disease,mycobacterium infec...</td>\n",
       "      <td>infectious disease</td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>infection</td>\n",
       "      <td>Buruli ulcer, Bairnsdale ulcer, Daintree ulcer...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>Q1018534</td>\n",
       "      <td>progressive familial intrahepatic cholestasis</td>\n",
       "      <td>70221</td>\n",
       "      <td></td>\n",
       "      <td>liver disease,intrahepatic cholestasis,familia...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>[ABCB11, ABCB4, ATP8B1]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>PFIC; Byler disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9468</th>\n",
       "      <td>9469</td>\n",
       "      <td>Q992139</td>\n",
       "      <td>human granulocytic anaplasmosis</td>\n",
       "      <td>50025</td>\n",
       "      <td></td>\n",
       "      <td>tick-borne disease,anaplasmosis,ehrlichiosis,I...</td>\n",
       "      <td>infectious disease</td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td>infection</td>\n",
       "      <td>HGE, human granulocytic ehrlichiosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9469</th>\n",
       "      <td>9470</td>\n",
       "      <td>Q994794</td>\n",
       "      <td>noma</td>\n",
       "      <td>9672</td>\n",
       "      <td>D009625</td>\n",
       "      <td>aphthous stomatitis,bacterial infectious disea...</td>\n",
       "      <td>gastroenterology</td>\n",
       "      <td></td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cancrum oris, gangrenous stomatitis, necotizin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9470</th>\n",
       "      <td>9471</td>\n",
       "      <td>Q994859</td>\n",
       "      <td>homocystinuria</td>\n",
       "      <td>9263</td>\n",
       "      <td>D006712</td>\n",
       "      <td>amino acid metabolic disorder,sulfuraminoacidemia</td>\n",
       "      <td>endocrinology</td>\n",
       "      <td></td>\n",
       "      <td>[CBS]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>CBS deficiency, cystathionine beta synthase de...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9471</th>\n",
       "      <td>9472</td>\n",
       "      <td>Q994942</td>\n",
       "      <td>bruxism</td>\n",
       "      <td>2846</td>\n",
       "      <td>D002012</td>\n",
       "      <td>sleep disorder,parafunctional habit</td>\n",
       "      <td>dentistry</td>\n",
       "      <td>dental attrition</td>\n",
       "      <td>[]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>bruxism, Bruxism - teeth grinding, Grinding te...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9472</th>\n",
       "      <td>9473</td>\n",
       "      <td>Q998273</td>\n",
       "      <td>Smith-Lemli-Opitz syndrome</td>\n",
       "      <td>14692</td>\n",
       "      <td>D019082</td>\n",
       "      <td>syndrome,ptosis,lipid metabolism disorder,mult...</td>\n",
       "      <td>medical genetics</td>\n",
       "      <td></td>\n",
       "      <td>[DHCR7]</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Smith-Opitz-Inborn syndrome, Smith–Lemli–Opitz...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9473 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      disease_key wikidata_id  \\\n",
       "0               1    Q1002195   \n",
       "1               2    Q1004647   \n",
       "2               3    Q1016605   \n",
       "3               4    Q1017169   \n",
       "4               5    Q1018534   \n",
       "...           ...         ...   \n",
       "9468         9469     Q992139   \n",
       "9469         9470     Q994794   \n",
       "9470         9471     Q994859   \n",
       "9471         9472     Q994942   \n",
       "9472         9473     Q998273   \n",
       "\n",
       "                                           disease_name  doid_id  mesh_id  \\\n",
       "0     autosomal recessive limb-girdle muscular dystr...   110297            \n",
       "1                                    bullous pemphigoid     8506  D010391   \n",
       "2                                      Burkitt lymphoma     8584  D002051   \n",
       "3                                  Buruli ulcer disease    50456  D054312   \n",
       "4         progressive familial intrahepatic cholestasis    70221            \n",
       "...                                                 ...      ...      ...   \n",
       "9468                    human granulocytic anaplasmosis    50025            \n",
       "9469                                               noma     9672  D009625   \n",
       "9470                                     homocystinuria     9263  D006712   \n",
       "9471                                            bruxism     2846  D002012   \n",
       "9472                         Smith-Lemli-Opitz syndrome    14692  D019082   \n",
       "\n",
       "                                        type_of_disease   health_speciality  \\\n",
       "0     autosomal recessive limb-girdle muscular dystr...           neurology   \n",
       "1     pemphigoid,autoimmune disease of skin and conn...         dermatology   \n",
       "2                                mature B-cell neoplasm          hematology   \n",
       "3     neglected tropical disease,mycobacterium infec...  infectious disease   \n",
       "4     liver disease,intrahepatic cholestasis,familia...                       \n",
       "...                                                 ...                 ...   \n",
       "9468  tick-borne disease,anaplasmosis,ehrlichiosis,I...  infectious disease   \n",
       "9469  aphthous stomatitis,bacterial infectious disea...    gastroenterology   \n",
       "9470  amino acid metabolic disorder,sulfuraminoacidemia       endocrinology   \n",
       "9471                sleep disorder,parafunctional habit           dentistry   \n",
       "9472  syndrome,ptosis,lipid metabolism disorder,mult...    medical genetics   \n",
       "\n",
       "              symptoms     genetic_associations affects  \\\n",
       "0                                       [POMT1]           \n",
       "1                                            []           \n",
       "2                                         [MYC]           \n",
       "3                                            []           \n",
       "4                       [ABCB11, ABCB4, ATP8B1]           \n",
       "...                ...                      ...     ...   \n",
       "9468                                         []           \n",
       "9469                                         []           \n",
       "9470                                      [CBS]           \n",
       "9471  dental attrition                       []           \n",
       "9472                                    [DHCR7]           \n",
       "\n",
       "                         caused_by  \\\n",
       "0                                    \n",
       "1                                    \n",
       "2     Epstein Barr virus infection   \n",
       "3                        infection   \n",
       "4                                    \n",
       "...                            ...   \n",
       "9468                     infection   \n",
       "9469                                 \n",
       "9470                                 \n",
       "9471                                 \n",
       "9472                                 \n",
       "\n",
       "                                      alternative_names  \n",
       "0     LGMD2K, limb-girdle muscular dystrophy-intelle...  \n",
       "1     Bullous pemphigoid, Bullous pemphigoid (disord...  \n",
       "2     Burkitt lymphoma/leukaemia, Burkitt's Lymphoma...  \n",
       "3     Buruli ulcer, Bairnsdale ulcer, Daintree ulcer...  \n",
       "4                                   PFIC; Byler disease  \n",
       "...                                                 ...  \n",
       "9468               HGE, human granulocytic ehrlichiosis  \n",
       "9469  cancrum oris, gangrenous stomatitis, necotizin...  \n",
       "9470  CBS deficiency, cystathionine beta synthase de...  \n",
       "9471  bruxism, Bruxism - teeth grinding, Grinding te...  \n",
       "9472  Smith-Opitz-Inborn syndrome, Smith–Lemli–Opitz...  \n",
       "\n",
       "[9473 rows x 12 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select * from disease_dimension\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNP_Dimension and Study_Dimension"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- As duas dimensões SNP e Study, são obtidas conjuntamente, pois a dimensão SNP depende dos dados que produzem a Study Dimension.\n",
    "- A modelação destas dimensões foi muito complicada e o código ainda não se encontra optimizado."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = ''\n",
    "def connectToServer():\n",
    "    global conn\n",
    "    conn = pg.connect(host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Vamos primeiro buscar a informação dos SNPs ao NCBI dbSNP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>gene</th>\n",
       "      <th>functional_consequence</th>\n",
       "      <th>clinical_significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs688</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>intron_variant,synonymous_variant,coding_seque...</td>\n",
       "      <td>benign,benign-likely-benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs2962</td>\n",
       "      <td>INSR</td>\n",
       "      <td>missense_variant,synonymous_variant,coding_seq...</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs2963</td>\n",
       "      <td>INSR</td>\n",
       "      <td>synonymous_variant,coding_sequence_variant</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs3008</td>\n",
       "      <td>JAK3</td>\n",
       "      <td>3_prime_UTR_variant,genic_downstream_transcrip...</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs3556</td>\n",
       "      <td>PEPD</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2450175</th>\n",
       "      <td>rs576173908</td>\n",
       "      <td>FBN3</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2450176</th>\n",
       "      <td>rs576173926</td>\n",
       "      <td>TPM4</td>\n",
       "      <td>downstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2450177</th>\n",
       "      <td>rs576173983</td>\n",
       "      <td>SIPA1L3</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2450178</th>\n",
       "      <td>rs576173984</td>\n",
       "      <td>ADGRE1</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2450179</th>\n",
       "      <td>rs576174028</td>\n",
       "      <td>ZNF536</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2450180 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        snp_ref_code     gene  \\\n",
       "0              rs688     LDLR   \n",
       "1             rs2962     INSR   \n",
       "2             rs2963     INSR   \n",
       "3             rs3008     JAK3   \n",
       "4             rs3556     PEPD   \n",
       "...              ...      ...   \n",
       "2450175  rs576173908     FBN3   \n",
       "2450176  rs576173926     TPM4   \n",
       "2450177  rs576173983  SIPA1L3   \n",
       "2450178  rs576173984   ADGRE1   \n",
       "2450179  rs576174028   ZNF536   \n",
       "\n",
       "                                    functional_consequence  \\\n",
       "0        intron_variant,synonymous_variant,coding_seque...   \n",
       "1        missense_variant,synonymous_variant,coding_seq...   \n",
       "2               synonymous_variant,coding_sequence_variant   \n",
       "3        3_prime_UTR_variant,genic_downstream_transcrip...   \n",
       "4                                      3_prime_UTR_variant   \n",
       "...                                                    ...   \n",
       "2450175                                     intron_variant   \n",
       "2450176                      downstream_transcript_variant   \n",
       "2450177                                     intron_variant   \n",
       "2450178                                     intron_variant   \n",
       "2450179   intron_variant,genic_upstream_transcript_variant   \n",
       "\n",
       "               clinical_significance  \n",
       "0        benign,benign-likely-benign  \n",
       "1                             benign  \n",
       "2                             benign  \n",
       "3                             benign  \n",
       "4                             benign  \n",
       "...                              ...  \n",
       "2450175                         None  \n",
       "2450176                         None  \n",
       "2450177                         None  \n",
       "2450178                         None  \n",
       "2450179                         None  \n",
       "\n",
       "[2450180 rows x 4 columns]"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd \n",
    "import re\n",
    "\n",
    "i = 0\n",
    "snpInfoList = []\n",
    "snpInfo = []\n",
    "with open('SNPS_Chr19.txt') as f:\n",
    "    for l in f:\n",
    "        #print(l)\n",
    "        l = l.strip()\n",
    "        if re.match(\"^\\d+\\.\", l):\n",
    "            if snpInfo != []:\n",
    "                #print(snpInfoList)\n",
    "                snpInfoList.append(snpInfo)\n",
    "            snpInfo = []\n",
    "            snp = l.split(' ')[1]\n",
    "            snpInfo.append(snp)\n",
    "        #elif l.startswith(\"Chromosome\"):\n",
    "        #    chromosome,snp_position = l.split(\" \")[1].split(':')\n",
    "        #    snpInfo.append(chromosome)\n",
    "        #    snpInfo.append(snp_position)\n",
    "        elif l.startswith(\"Gene\"):\n",
    "            gene = l.split(\":\")[1]\n",
    "            snpInfo.append(gene)\n",
    "        elif l.startswith(\"Functional Consequence\"):\n",
    "            functional_consequence = l.split(\" \")[2]\n",
    "            snpInfo.append(functional_consequence)\n",
    "        elif l.startswith(\"Clinical significance\"):\n",
    "            clinical_significance = l.split(\" \")[2]\n",
    "            snpInfo.append(clinical_significance)\n",
    "        i += 1\n",
    "        #if i > 100:\n",
    "        #   break\n",
    "#print(snpLists)\n",
    "\n",
    "snps_chr19 = pd.DataFrame(snpInfoList,columns=['snp_ref_code','gene','functional_consequence','clinical_significance'])\n",
    "#snps_chr19 = pd.DataFrame(snpInfoList,columns=['snp_ref_code','chromosome','snp_position','gene','functional_consequence','clinical_significance'])\n",
    "snps_chr19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>gene</th>\n",
       "      <th>functional_consequence</th>\n",
       "      <th>clinical_significance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs688</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>intron_variant,synonymous_variant,coding_seque...</td>\n",
       "      <td>benign,benign-likely-benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs2962</td>\n",
       "      <td>INSR</td>\n",
       "      <td>missense_variant,synonymous_variant,coding_seq...</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs2963</td>\n",
       "      <td>INSR</td>\n",
       "      <td>synonymous_variant,coding_sequence_variant</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs3008</td>\n",
       "      <td>JAK3</td>\n",
       "      <td>3_prime_UTR_variant,genic_downstream_transcrip...</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs3556</td>\n",
       "      <td>PEPD</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>benign</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  snp_ref_code  gene                             functional_consequence  \\\n",
       "0        rs688  LDLR  intron_variant,synonymous_variant,coding_seque...   \n",
       "1       rs2962  INSR  missense_variant,synonymous_variant,coding_seq...   \n",
       "2       rs2963  INSR         synonymous_variant,coding_sequence_variant   \n",
       "3       rs3008  JAK3  3_prime_UTR_variant,genic_downstream_transcrip...   \n",
       "4       rs3556  PEPD                                3_prime_UTR_variant   \n",
       "\n",
       "         clinical_significance  \n",
       "0  benign,benign-likely-benign  \n",
       "1                       benign  \n",
       "2                       benign  \n",
       "3                       benign  \n",
       "4                       benign  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_chr19.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- De seguida, vamos buscar a informação dos SNPs ao ficheiro dos 1000 genomes project.\n",
    "- Estes dados foram previamente filtrados, para eliminar a informação genética de cada individuo e facilitar a introdução dos dados na tabela de dimensão SNP."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_filtered = pd.read_csv(\"SNP_filtered.csv\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CHROM', 'POS', 'ID', 'REF', 'ALT_1', 'EAS_AF_1', 'EUR_AF_1',\n",
       "       'AFR_AF_1', 'AMR_AF_1', 'SAS_AF_1'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_filtered.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_filtered.columns = ['chromosome', 'snp_position', 'snp_ref_code', 'reference', 'alternative', 'easian_frequency', 'european_frequency',\n",
    "       'african_frequency', 'american_frequency', 'sasian_frequency']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>reference</th>\n",
       "      <th>alternative</th>\n",
       "      <th>easian_frequency</th>\n",
       "      <th>european_frequency</th>\n",
       "      <th>african_frequency</th>\n",
       "      <th>american_frequency</th>\n",
       "      <th>sasian_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>60842</td>\n",
       "      <td>rs541392352</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>64705</td>\n",
       "      <td>rs559839262</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>0.0091</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>69984</td>\n",
       "      <td>rs372156287</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0079</td>\n",
       "      <td>0.0159</td>\n",
       "      <td>0.0287</td>\n",
       "      <td>0.0231</td>\n",
       "      <td>0.0092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>80844</td>\n",
       "      <td>rs546022921</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0020</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>80849</td>\n",
       "      <td>rs564694216</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.0109</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751873</th>\n",
       "      <td>19</td>\n",
       "      <td>59118740</td>\n",
       "      <td>rs546075210</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751874</th>\n",
       "      <td>19</td>\n",
       "      <td>59118764</td>\n",
       "      <td>rs556756243</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.0020</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751875</th>\n",
       "      <td>19</td>\n",
       "      <td>59118779</td>\n",
       "      <td>rs141816674</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.6498</td>\n",
       "      <td>0.3072</td>\n",
       "      <td>0.8835</td>\n",
       "      <td>0.4553</td>\n",
       "      <td>0.4315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751876</th>\n",
       "      <td>19</td>\n",
       "      <td>59118783</td>\n",
       "      <td>rs150801216</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.6498</td>\n",
       "      <td>0.3072</td>\n",
       "      <td>0.8835</td>\n",
       "      <td>0.4553</td>\n",
       "      <td>0.4315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751877</th>\n",
       "      <td>19</td>\n",
       "      <td>59118812</td>\n",
       "      <td>rs369599189</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1751878 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chromosome  snp_position snp_ref_code reference alternative  \\\n",
       "0                19         60842  rs541392352         A           G   \n",
       "1                19         64705  rs559839262         C           T   \n",
       "2                19         69984  rs372156287         G           A   \n",
       "3                19         80844  rs546022921         T           C   \n",
       "4                19         80849  rs564694216         G           C   \n",
       "...             ...           ...          ...       ...         ...   \n",
       "1751873          19      59118740  rs546075210         T           A   \n",
       "1751874          19      59118764  rs556756243         G           A   \n",
       "1751875          19      59118779  rs141816674         G           C   \n",
       "1751876          19      59118783  rs150801216         C           G   \n",
       "1751877          19      59118812  rs369599189         G           T   \n",
       "\n",
       "         easian_frequency  european_frequency  african_frequency  \\\n",
       "0                  0.0000              0.0000             0.0000   \n",
       "1                  0.0000              0.0030             0.0091   \n",
       "2                  0.0079              0.0159             0.0287   \n",
       "3                  0.0000              0.0020             0.0000   \n",
       "4                  0.0109              0.0000             0.0015   \n",
       "...                   ...                 ...                ...   \n",
       "1751873            0.0000              0.0000             0.0008   \n",
       "1751874            0.0020              0.0000             0.0000   \n",
       "1751875            0.6498              0.3072             0.8835   \n",
       "1751876            0.6498              0.3072             0.8835   \n",
       "1751877            0.0000              0.0010             0.0000   \n",
       "\n",
       "         american_frequency  sasian_frequency  \n",
       "0                    0.0014            0.0010  \n",
       "1                    0.0000            0.0000  \n",
       "2                    0.0231            0.0092  \n",
       "3                    0.0000            0.0000  \n",
       "4                    0.0000            0.0000  \n",
       "...                     ...               ...  \n",
       "1751873              0.0000            0.0000  \n",
       "1751874              0.0000            0.0000  \n",
       "1751875              0.4553            0.4315  \n",
       "1751876              0.4553            0.4315  \n",
       "1751877              0.0000            0.0000  \n",
       "\n",
       "[1751878 rows x 10 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Juntamos os dois dataframes num só."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>functional_consequence</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>reference</th>\n",
       "      <th>alternative</th>\n",
       "      <th>easian_frequency</th>\n",
       "      <th>european_frequency</th>\n",
       "      <th>african_frequency</th>\n",
       "      <th>american_frequency</th>\n",
       "      <th>sasian_frequency</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs10001</th>\n",
       "      <td>STXBP2</td>\n",
       "      <td>coding_sequence_variant,genic_downstream_trans...</td>\n",
       "      <td>benign</td>\n",
       "      <td>19</td>\n",
       "      <td>7711221</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.7034</td>\n",
       "      <td>0.4036</td>\n",
       "      <td>0.5356</td>\n",
       "      <td>0.4669</td>\n",
       "      <td>0.3865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1000237</th>\n",
       "      <td>GATAD2A</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19518316</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.3185</td>\n",
       "      <td>0.3370</td>\n",
       "      <td>0.6619</td>\n",
       "      <td>0.4813</td>\n",
       "      <td>0.5348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1000244</th>\n",
       "      <td>COPE</td>\n",
       "      <td>upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19030467</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1000245</th>\n",
       "      <td>COPE</td>\n",
       "      <td>intron_variant,upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19031087</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1000262</th>\n",
       "      <td>RPL18A</td>\n",
       "      <td>upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>17970627</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2470</td>\n",
       "      <td>0.2286</td>\n",
       "      <td>0.0628</td>\n",
       "      <td>0.3040</td>\n",
       "      <td>0.2280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs999814</th>\n",
       "      <td>IQCN</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>18368969</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.2371</td>\n",
       "      <td>0.2724</td>\n",
       "      <td>0.0189</td>\n",
       "      <td>0.3314</td>\n",
       "      <td>0.1646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs999815</th>\n",
       "      <td>IQCN</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>18368988</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.2381</td>\n",
       "      <td>0.2704</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.2017</td>\n",
       "      <td>0.1626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs999850</th>\n",
       "      <td>DPY19L3</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>32901045</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.6399</td>\n",
       "      <td>0.6402</td>\n",
       "      <td>0.4342</td>\n",
       "      <td>0.4265</td>\n",
       "      <td>0.5920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs999854</th>\n",
       "      <td>CACNA1A</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>13616206</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.2569</td>\n",
       "      <td>0.0109</td>\n",
       "      <td>0.0076</td>\n",
       "      <td>0.2320</td>\n",
       "      <td>0.0706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs999967</th>\n",
       "      <td>ADGRE2</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>14863522</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.5169</td>\n",
       "      <td>0.2654</td>\n",
       "      <td>0.6551</td>\n",
       "      <td>0.3833</td>\n",
       "      <td>0.5706</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1706564 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 gene                             functional_consequence  \\\n",
       "snp_ref_code                                                               \n",
       "rs10001        STXBP2  coding_sequence_variant,genic_downstream_trans...   \n",
       "rs1000237     GATAD2A   intron_variant,genic_upstream_transcript_variant   \n",
       "rs1000244        COPE                        upstream_transcript_variant   \n",
       "rs1000245        COPE         intron_variant,upstream_transcript_variant   \n",
       "rs1000262      RPL18A                        upstream_transcript_variant   \n",
       "...               ...                                                ...   \n",
       "rs999814         IQCN                                     intron_variant   \n",
       "rs999815         IQCN                                     intron_variant   \n",
       "rs999850      DPY19L3   intron_variant,genic_upstream_transcript_variant   \n",
       "rs999854      CACNA1A                                     intron_variant   \n",
       "rs999967       ADGRE2                                     intron_variant   \n",
       "\n",
       "             clinical_significance  chromosome  snp_position reference  \\\n",
       "snp_ref_code                                                             \n",
       "rs10001                     benign          19       7711221         T   \n",
       "rs1000237                     None          19      19518316         T   \n",
       "rs1000244                     None          19      19030467         A   \n",
       "rs1000245                     None          19      19031087         C   \n",
       "rs1000262                     None          19      17970627         C   \n",
       "...                            ...         ...           ...       ...   \n",
       "rs999814                      None          19      18368969         G   \n",
       "rs999815                      None          19      18368988         G   \n",
       "rs999850                      None          19      32901045         T   \n",
       "rs999854                      None          19      13616206         G   \n",
       "rs999967                      None          19      14863522         A   \n",
       "\n",
       "             alternative  easian_frequency  european_frequency  \\\n",
       "snp_ref_code                                                     \n",
       "rs10001                C            0.7034              0.4036   \n",
       "rs1000237              A            0.3185              0.3370   \n",
       "rs1000244              G            0.0000              0.0000   \n",
       "rs1000245              G            0.0000              0.0000   \n",
       "rs1000262              G            0.2470              0.2286   \n",
       "...                  ...               ...                 ...   \n",
       "rs999814               A            0.2371              0.2724   \n",
       "rs999815               A            0.2381              0.2704   \n",
       "rs999850               A            0.6399              0.6402   \n",
       "rs999854               A            0.2569              0.0109   \n",
       "rs999967               G            0.5169              0.2654   \n",
       "\n",
       "              african_frequency  american_frequency  sasian_frequency  \n",
       "snp_ref_code                                                           \n",
       "rs10001                  0.5356              0.4669            0.3865  \n",
       "rs1000237                0.6619              0.4813            0.5348  \n",
       "rs1000244                0.0174              0.0014            0.0000  \n",
       "rs1000245                0.0174              0.0014            0.0000  \n",
       "rs1000262                0.0628              0.3040            0.2280  \n",
       "...                         ...                 ...               ...  \n",
       "rs999814                 0.0189              0.3314            0.1646  \n",
       "rs999815                 0.0204              0.2017            0.1626  \n",
       "rs999850                 0.4342              0.4265            0.5920  \n",
       "rs999854                 0.0076              0.2320            0.0706  \n",
       "rs999967                 0.6551              0.3833            0.5706  \n",
       "\n",
       "[1706564 rows x 12 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = snps_chr19.set_index('snp_ref_code').join(snp_filtered.set_index('snp_ref_code'),how=\"inner\")\n",
    "df1\n",
    "\n",
    "#df1 = pd.merge(snps_chr19, snp_filtered, on='snp_ref_code', how='right')\n",
    "#df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>gene</th>\n",
       "      <th>functional_consequence</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>reference</th>\n",
       "      <th>alternative</th>\n",
       "      <th>easian_frequency</th>\n",
       "      <th>european_frequency</th>\n",
       "      <th>african_frequency</th>\n",
       "      <th>american_frequency</th>\n",
       "      <th>sasian_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs10001</td>\n",
       "      <td>STXBP2</td>\n",
       "      <td>coding_sequence_variant,genic_downstream_trans...</td>\n",
       "      <td>benign</td>\n",
       "      <td>19</td>\n",
       "      <td>7711221</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.7034</td>\n",
       "      <td>0.4036</td>\n",
       "      <td>0.5356</td>\n",
       "      <td>0.4669</td>\n",
       "      <td>0.3865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1000237</td>\n",
       "      <td>GATAD2A</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19518316</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.3185</td>\n",
       "      <td>0.3370</td>\n",
       "      <td>0.6619</td>\n",
       "      <td>0.4813</td>\n",
       "      <td>0.5348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1000244</td>\n",
       "      <td>COPE</td>\n",
       "      <td>upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19030467</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1000245</td>\n",
       "      <td>COPE</td>\n",
       "      <td>intron_variant,upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19031087</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1000262</td>\n",
       "      <td>RPL18A</td>\n",
       "      <td>upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>17970627</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2470</td>\n",
       "      <td>0.2286</td>\n",
       "      <td>0.0628</td>\n",
       "      <td>0.3040</td>\n",
       "      <td>0.2280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1706559</th>\n",
       "      <td>rs999814</td>\n",
       "      <td>IQCN</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>18368969</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.2371</td>\n",
       "      <td>0.2724</td>\n",
       "      <td>0.0189</td>\n",
       "      <td>0.3314</td>\n",
       "      <td>0.1646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1706560</th>\n",
       "      <td>rs999815</td>\n",
       "      <td>IQCN</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>18368988</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.2381</td>\n",
       "      <td>0.2704</td>\n",
       "      <td>0.0204</td>\n",
       "      <td>0.2017</td>\n",
       "      <td>0.1626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1706561</th>\n",
       "      <td>rs999850</td>\n",
       "      <td>DPY19L3</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>32901045</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.6399</td>\n",
       "      <td>0.6402</td>\n",
       "      <td>0.4342</td>\n",
       "      <td>0.4265</td>\n",
       "      <td>0.5920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1706562</th>\n",
       "      <td>rs999854</td>\n",
       "      <td>CACNA1A</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>13616206</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.2569</td>\n",
       "      <td>0.0109</td>\n",
       "      <td>0.0076</td>\n",
       "      <td>0.2320</td>\n",
       "      <td>0.0706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1706563</th>\n",
       "      <td>rs999967</td>\n",
       "      <td>ADGRE2</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>14863522</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.5169</td>\n",
       "      <td>0.2654</td>\n",
       "      <td>0.6551</td>\n",
       "      <td>0.3833</td>\n",
       "      <td>0.5706</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1706564 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        snp_ref_code     gene  \\\n",
       "0            rs10001   STXBP2   \n",
       "1          rs1000237  GATAD2A   \n",
       "2          rs1000244     COPE   \n",
       "3          rs1000245     COPE   \n",
       "4          rs1000262   RPL18A   \n",
       "...              ...      ...   \n",
       "1706559     rs999814     IQCN   \n",
       "1706560     rs999815     IQCN   \n",
       "1706561     rs999850  DPY19L3   \n",
       "1706562     rs999854  CACNA1A   \n",
       "1706563     rs999967   ADGRE2   \n",
       "\n",
       "                                    functional_consequence  \\\n",
       "0        coding_sequence_variant,genic_downstream_trans...   \n",
       "1         intron_variant,genic_upstream_transcript_variant   \n",
       "2                              upstream_transcript_variant   \n",
       "3               intron_variant,upstream_transcript_variant   \n",
       "4                              upstream_transcript_variant   \n",
       "...                                                    ...   \n",
       "1706559                                     intron_variant   \n",
       "1706560                                     intron_variant   \n",
       "1706561   intron_variant,genic_upstream_transcript_variant   \n",
       "1706562                                     intron_variant   \n",
       "1706563                                     intron_variant   \n",
       "\n",
       "        clinical_significance  chromosome  snp_position reference alternative  \\\n",
       "0                      benign          19       7711221         T           C   \n",
       "1                        None          19      19518316         T           A   \n",
       "2                        None          19      19030467         A           G   \n",
       "3                        None          19      19031087         C           G   \n",
       "4                        None          19      17970627         C           G   \n",
       "...                       ...         ...           ...       ...         ...   \n",
       "1706559                  None          19      18368969         G           A   \n",
       "1706560                  None          19      18368988         G           A   \n",
       "1706561                  None          19      32901045         T           A   \n",
       "1706562                  None          19      13616206         G           A   \n",
       "1706563                  None          19      14863522         A           G   \n",
       "\n",
       "         easian_frequency  european_frequency  african_frequency  \\\n",
       "0                  0.7034              0.4036             0.5356   \n",
       "1                  0.3185              0.3370             0.6619   \n",
       "2                  0.0000              0.0000             0.0174   \n",
       "3                  0.0000              0.0000             0.0174   \n",
       "4                  0.2470              0.2286             0.0628   \n",
       "...                   ...                 ...                ...   \n",
       "1706559            0.2371              0.2724             0.0189   \n",
       "1706560            0.2381              0.2704             0.0204   \n",
       "1706561            0.6399              0.6402             0.4342   \n",
       "1706562            0.2569              0.0109             0.0076   \n",
       "1706563            0.5169              0.2654             0.6551   \n",
       "\n",
       "         american_frequency  sasian_frequency  \n",
       "0                    0.4669            0.3865  \n",
       "1                    0.4813            0.5348  \n",
       "2                    0.0014            0.0000  \n",
       "3                    0.0014            0.0000  \n",
       "4                    0.3040            0.2280  \n",
       "...                     ...               ...  \n",
       "1706559              0.3314            0.1646  \n",
       "1706560              0.2017            0.1626  \n",
       "1706561              0.4265            0.5920  \n",
       "1706562              0.2320            0.0706  \n",
       "1706563              0.3833            0.5706  \n",
       "\n",
       "[1706564 rows x 13 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = df1.reset_index()\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- De seguida, vamos ler o dataset do GWAS catalogue e extrair a informação acerca dos SNPs identificados em cada estudo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['DATE ADDED TO CATALOG', 'PUBMEDID', 'FIRST AUTHOR', 'DATE', 'JOURNAL',\n",
       "       'LINK', 'STUDY', 'DISEASE/TRAIT', 'INITIAL SAMPLE SIZE',\n",
       "       'REPLICATION SAMPLE SIZE', 'REGION', 'CHR_ID', 'CHR_POS',\n",
       "       'REPORTED GENE(S)', 'MAPPED_GENE', 'UPSTREAM_GENE_ID',\n",
       "       'DOWNSTREAM_GENE_ID', 'SNP_GENE_IDS', 'UPSTREAM_GENE_DISTANCE',\n",
       "       'DOWNSTREAM_GENE_DISTANCE', 'STRONGEST SNP-RISK ALLELE', 'SNPS',\n",
       "       'MERGED', 'SNP_ID_CURRENT', 'CONTEXT', 'INTERGENIC',\n",
       "       'RISK ALLELE FREQUENCY', 'P-VALUE', 'PVALUE_MLOG', 'P-VALUE (TEXT)',\n",
       "       'OR or BETA', '95% CI (TEXT)', 'PLATFORM [SNPS PASSING QC]', 'CNV',\n",
       "       'MAPPED_TRAIT', 'MAPPED_TRAIT_URI', 'STUDY ACCESSION',\n",
       "       'GENOTYPING TECHNOLOGY'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_catalog = pd.read_csv(\"gwas_catalog_v1.0.2-associations_e98_r2020-03-08.tsv\",sep='\\t')\n",
    "gwas_catalog.columns\n",
    "\n",
    "#gwas_catalog.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(179364, 38)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_catalog.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- O dataset é processado para remover as colunas não necessárias. \n",
    "- Para a simplificação do processo, removemos também casos em que os estudos declaravam mais que um SNP nos estudos de associação. Além disso, limitámos a nossa análise aos estudos de associação genética com SNPs localizados no cromossoma 19 (seguindo as decisões feitas na elaboração do projecto).\n",
    "- Procedemos a mais filtragens dos dados, para garantir que a terminologia referente aos SNPs estava na forma rs*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['PUBMEDID', 'DATE', 'DISEASE/TRAIT', 'REGION', 'CHR_POS',\n",
       "       'REPORTED GENE(S)', 'MAPPED_GENE', 'UPSTREAM_GENE_ID',\n",
       "       'DOWNSTREAM_GENE_ID', 'SNP_GENE_IDS', 'UPSTREAM_GENE_DISTANCE',\n",
       "       'DOWNSTREAM_GENE_DISTANCE', 'STRONGEST SNP-RISK ALLELE', 'SNPS',\n",
       "       'CONTEXT', 'INTERGENIC', 'RISK ALLELE FREQUENCY', 'P-VALUE'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "del gwas_catalog[\"DATE ADDED TO CATALOG\"]\n",
    "del gwas_catalog[\"FIRST AUTHOR\"]\n",
    "del gwas_catalog[\"JOURNAL\"]\n",
    "del gwas_catalog[\"LINK\"]\n",
    "del gwas_catalog[\"STUDY\"]\n",
    "del gwas_catalog[\"INITIAL SAMPLE SIZE\"]\n",
    "del gwas_catalog[\"REPLICATION SAMPLE SIZE\"]\n",
    "del gwas_catalog[\"CHR_ID\"]\n",
    "del gwas_catalog[\"MERGED\"]\n",
    "del gwas_catalog[\"SNP_ID_CURRENT\"]\n",
    "del gwas_catalog[\"PVALUE_MLOG\"]\n",
    "del gwas_catalog[\"P-VALUE (TEXT)\"]\n",
    "del gwas_catalog[\"OR or BETA\"]\n",
    "del gwas_catalog[\"95% CI (TEXT)\"]\n",
    "del gwas_catalog[\"PLATFORM [SNPS PASSING QC]\"]\n",
    "del gwas_catalog[\"CNV\"]\n",
    "del gwas_catalog[\"MAPPED_TRAIT\"]\n",
    "del gwas_catalog[\"MAPPED_TRAIT_URI\"]\n",
    "del gwas_catalog[\"STUDY ACCESSION\"]\n",
    "del gwas_catalog[\"GENOTYPING TECHNOLOGY\"]\n",
    "\n",
    "gwas_catalog.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_catalog.columns = ['pumbmed_id', 'study_date', 'disease', 'region', 'chr_pos', 'reported_genes', 'mapped_genes',\n",
    "       'upstream_gene_id', 'downstream_gene_id', 'snp_gene_ids', 'upstream_gene_distance', 'dowstream_gene_distance',\n",
    "        'strongest_snp_risk_allele','snps','context','intergenic','risk_allele_frequency','p_value']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pumbmed_id</th>\n",
       "      <th>study_date</th>\n",
       "      <th>disease</th>\n",
       "      <th>region</th>\n",
       "      <th>chr_pos</th>\n",
       "      <th>reported_genes</th>\n",
       "      <th>mapped_genes</th>\n",
       "      <th>upstream_gene_id</th>\n",
       "      <th>downstream_gene_id</th>\n",
       "      <th>snp_gene_ids</th>\n",
       "      <th>upstream_gene_distance</th>\n",
       "      <th>dowstream_gene_distance</th>\n",
       "      <th>strongest_snp_risk_allele</th>\n",
       "      <th>snps</th>\n",
       "      <th>context</th>\n",
       "      <th>intergenic</th>\n",
       "      <th>risk_allele_frequency</th>\n",
       "      <th>p_value</th>\n",
       "      <th>multiple_snp_semicolon</th>\n",
       "      <th>multiple_snp_colon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>21102463</td>\n",
       "      <td>2010-11-21</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>1124836</td>\n",
       "      <td>GPX4, SBNO2</td>\n",
       "      <td>SBNO2, SBNO2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000064932, ENSG00000278788</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs740495-G</td>\n",
       "      <td>rs740495</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.247</td>\n",
       "      <td>8.000000e-12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>21102463</td>\n",
       "      <td>2010-11-21</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>10359299</td>\n",
       "      <td>ICAM3, TYK2, ICAM1</td>\n",
       "      <td>TYK2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000105397</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs12720356-G</td>\n",
       "      <td>rs12720356</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.084</td>\n",
       "      <td>1.000000e-12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>21102463</td>\n",
       "      <td>2010-11-21</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>19q13.33</td>\n",
       "      <td>48711017</td>\n",
       "      <td>FUT2, RASIP1</td>\n",
       "      <td>FUT2 - MAMSTR</td>\n",
       "      <td>ENSG00000176920</td>\n",
       "      <td>ENSG00000176909</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5066.0</td>\n",
       "      <td>1708.0</td>\n",
       "      <td>rs281379-A</td>\n",
       "      <td>rs281379</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.487</td>\n",
       "      <td>7.000000e-12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>21658281</td>\n",
       "      <td>2011-06-10</td>\n",
       "      <td>Sudden cardiac arrest</td>\n",
       "      <td>19q13.32</td>\n",
       "      <td>45804148</td>\n",
       "      <td>RSHL1</td>\n",
       "      <td>RSPH6A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000104941</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs8111071-?</td>\n",
       "      <td>rs8111071</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.09</td>\n",
       "      <td>4.000000e-06</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>19118814</td>\n",
       "      <td>2009-01-03</td>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>19q13.31</td>\n",
       "      <td>44110055</td>\n",
       "      <td>ZNF224</td>\n",
       "      <td>AC021092.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000186019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs2061333-?</td>\n",
       "      <td>rs2061333</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>2.000000e-06</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178774</th>\n",
       "      <td>30038396</td>\n",
       "      <td>2018-07-23</td>\n",
       "      <td>Highest math class taken (MTAG)</td>\n",
       "      <td>19q13.11</td>\n",
       "      <td>32493978</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>DPY19L3 - PDCD5</td>\n",
       "      <td>ENSG00000178904</td>\n",
       "      <td>ENSG00000105185</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8083.0</td>\n",
       "      <td>87212.0</td>\n",
       "      <td>rs7250367-A</td>\n",
       "      <td>rs7250367</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.8239</td>\n",
       "      <td>4.000000e-08</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178775</th>\n",
       "      <td>30038396</td>\n",
       "      <td>2018-07-23</td>\n",
       "      <td>Highest math class taken (MTAG)</td>\n",
       "      <td>19q13.12</td>\n",
       "      <td>35746344</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>PSENEN, AD000671.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000205155, ENSG00000188223</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs2293688-C</td>\n",
       "      <td>rs2293688</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.6589</td>\n",
       "      <td>5.000000e-10</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178776</th>\n",
       "      <td>30038396</td>\n",
       "      <td>2018-07-23</td>\n",
       "      <td>Highest math class taken (MTAG)</td>\n",
       "      <td>19q13.2</td>\n",
       "      <td>40589394</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>SHKBP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000160410</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs76000727-A</td>\n",
       "      <td>rs76000727</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.089</td>\n",
       "      <td>3.000000e-08</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178777</th>\n",
       "      <td>30038396</td>\n",
       "      <td>2018-07-23</td>\n",
       "      <td>Highest math class taken (MTAG)</td>\n",
       "      <td>19q13.42</td>\n",
       "      <td>53939797</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>CACNG7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000105605</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs307913-A</td>\n",
       "      <td>rs307913</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.8516</td>\n",
       "      <td>6.000000e-12</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178778</th>\n",
       "      <td>30038396</td>\n",
       "      <td>2018-07-23</td>\n",
       "      <td>Highest math class taken (MTAG)</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>58530356</td>\n",
       "      <td>Intergenic</td>\n",
       "      <td>ZBTB45</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000119574</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs7253433-A</td>\n",
       "      <td>rs7253433</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.3255</td>\n",
       "      <td>5.000000e-09</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6817 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        pumbmed_id  study_date                          disease    region  \\\n",
       "10        21102463  2010-11-21                  Crohn's disease   19p13.3   \n",
       "33        21102463  2010-11-21                  Crohn's disease   19p13.2   \n",
       "40        21102463  2010-11-21                  Crohn's disease  19q13.33   \n",
       "110       21658281  2011-06-10            Sudden cardiac arrest  19q13.32   \n",
       "118       19118814  2009-01-03              Alzheimer's disease  19q13.31   \n",
       "...            ...         ...                              ...       ...   \n",
       "178774    30038396  2018-07-23  Highest math class taken (MTAG)  19q13.11   \n",
       "178775    30038396  2018-07-23  Highest math class taken (MTAG)  19q13.12   \n",
       "178776    30038396  2018-07-23  Highest math class taken (MTAG)   19q13.2   \n",
       "178777    30038396  2018-07-23  Highest math class taken (MTAG)  19q13.42   \n",
       "178778    30038396  2018-07-23  Highest math class taken (MTAG)  19q13.43   \n",
       "\n",
       "         chr_pos      reported_genes        mapped_genes upstream_gene_id  \\\n",
       "10       1124836         GPX4, SBNO2        SBNO2, SBNO2              NaN   \n",
       "33      10359299  ICAM3, TYK2, ICAM1                TYK2              NaN   \n",
       "40      48711017        FUT2, RASIP1       FUT2 - MAMSTR  ENSG00000176920   \n",
       "110     45804148               RSHL1              RSPH6A              NaN   \n",
       "118     44110055              ZNF224          AC021092.1              NaN   \n",
       "...          ...                 ...                 ...              ...   \n",
       "178774  32493978          Intergenic     DPY19L3 - PDCD5  ENSG00000178904   \n",
       "178775  35746344          Intergenic  PSENEN, AD000671.1              NaN   \n",
       "178776  40589394          Intergenic              SHKBP1              NaN   \n",
       "178777  53939797          Intergenic              CACNG7              NaN   \n",
       "178778  58530356          Intergenic              ZBTB45              NaN   \n",
       "\n",
       "       downstream_gene_id                      snp_gene_ids  \\\n",
       "10                    NaN  ENSG00000064932, ENSG00000278788   \n",
       "33                    NaN                   ENSG00000105397   \n",
       "40        ENSG00000176909                               NaN   \n",
       "110                   NaN                   ENSG00000104941   \n",
       "118                   NaN                   ENSG00000186019   \n",
       "...                   ...                               ...   \n",
       "178774    ENSG00000105185                               NaN   \n",
       "178775                NaN  ENSG00000205155, ENSG00000188223   \n",
       "178776                NaN                   ENSG00000160410   \n",
       "178777                NaN                   ENSG00000105605   \n",
       "178778                NaN                   ENSG00000119574   \n",
       "\n",
       "        upstream_gene_distance  dowstream_gene_distance  \\\n",
       "10                         NaN                      NaN   \n",
       "33                         NaN                      NaN   \n",
       "40                      5066.0                   1708.0   \n",
       "110                        NaN                      NaN   \n",
       "118                        NaN                      NaN   \n",
       "...                        ...                      ...   \n",
       "178774                  8083.0                  87212.0   \n",
       "178775                     NaN                      NaN   \n",
       "178776                     NaN                      NaN   \n",
       "178777                     NaN                      NaN   \n",
       "178778                     NaN                      NaN   \n",
       "\n",
       "       strongest_snp_risk_allele        snps             context  intergenic  \\\n",
       "10                    rs740495-G    rs740495      intron_variant         0.0   \n",
       "33                  rs12720356-G  rs12720356    missense_variant         0.0   \n",
       "40                    rs281379-A    rs281379  intergenic_variant         1.0   \n",
       "110                  rs8111071-?   rs8111071      intron_variant         0.0   \n",
       "118                  rs2061333-?   rs2061333      intron_variant         0.0   \n",
       "...                          ...         ...                 ...         ...   \n",
       "178774               rs7250367-A   rs7250367  intergenic_variant         1.0   \n",
       "178775               rs2293688-C   rs2293688      intron_variant         0.0   \n",
       "178776              rs76000727-A  rs76000727      intron_variant         0.0   \n",
       "178777                rs307913-A    rs307913      intron_variant         0.0   \n",
       "178778               rs7253433-A   rs7253433      intron_variant         0.0   \n",
       "\n",
       "       risk_allele_frequency       p_value  multiple_snp_semicolon  \\\n",
       "10                     0.247  8.000000e-12                   False   \n",
       "33                     0.084  1.000000e-12                   False   \n",
       "40                     0.487  7.000000e-12                   False   \n",
       "110                     0.09  4.000000e-06                   False   \n",
       "118                       NR  2.000000e-06                   False   \n",
       "...                      ...           ...                     ...   \n",
       "178774                0.8239  4.000000e-08                   False   \n",
       "178775                0.6589  5.000000e-10                   False   \n",
       "178776                 0.089  3.000000e-08                   False   \n",
       "178777                0.8516  6.000000e-12                   False   \n",
       "178778                0.3255  5.000000e-09                   False   \n",
       "\n",
       "        multiple_snp_colon  \n",
       "10                   False  \n",
       "33                   False  \n",
       "40                   False  \n",
       "110                  False  \n",
       "118                  False  \n",
       "...                    ...  \n",
       "178774               False  \n",
       "178775               False  \n",
       "178776               False  \n",
       "178777               False  \n",
       "178778               False  \n",
       "\n",
       "[6817 rows x 20 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_catalog['multiple_snp_semicolon'] = gwas_catalog['snps'].str.contains(';')\n",
    "gwas_catalog = gwas_catalog[gwas_catalog['multiple_snp_semicolon'] == False]\n",
    "gwas_catalog['multiple_snp_colon'] = gwas_catalog['snps'].str.contains(',')\n",
    "gwas_catalog = gwas_catalog[gwas_catalog['multiple_snp_colon'] == False]\n",
    "gwas_catalog[gwas_catalog ['snps'].str.match('rs')]\n",
    "gwas_catalog = gwas_catalog[gwas_catalog.region.notnull()]\n",
    "gwas_catalog = gwas_catalog[gwas_catalog ['region'].str.match('19')]\n",
    "gwas_catalog['p_value'] = gwas_catalog['p_value'].astype(float)\n",
    "\n",
    "\n",
    "gwas_catalog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pumbmed_id</th>\n",
       "      <th>study_date</th>\n",
       "      <th>disease</th>\n",
       "      <th>region</th>\n",
       "      <th>chr_pos</th>\n",
       "      <th>reported_genes</th>\n",
       "      <th>mapped_genes</th>\n",
       "      <th>upstream_gene_id</th>\n",
       "      <th>downstream_gene_id</th>\n",
       "      <th>snp_gene_ids</th>\n",
       "      <th>...</th>\n",
       "      <th>dowstream_gene_distance</th>\n",
       "      <th>strongest_snp_risk_allele</th>\n",
       "      <th>snps</th>\n",
       "      <th>context</th>\n",
       "      <th>intergenic</th>\n",
       "      <th>risk_allele_frequency</th>\n",
       "      <th>p_value</th>\n",
       "      <th>multiple_snp_semicolon</th>\n",
       "      <th>multiple_snp_colon</th>\n",
       "      <th>snp_length_str</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64907</th>\n",
       "      <td>30698716</td>\n",
       "      <td>2019-01-29</td>\n",
       "      <td>HDL cholesterol levels in current drinkers</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>8364439</td>\n",
       "      <td>ANGPTL4</td>\n",
       "      <td>ANGPTL4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000167772</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs116843064-A</td>\n",
       "      <td>rs116843064</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>2.000000e-14</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91016</th>\n",
       "      <td>29777097</td>\n",
       "      <td>2018-05-18</td>\n",
       "      <td>Family history of Alzheimer's disease</td>\n",
       "      <td>19q13.32</td>\n",
       "      <td>44842530</td>\n",
       "      <td>NR</td>\n",
       "      <td>BCAM - NECTIN2</td>\n",
       "      <td>ENSG00000187244</td>\n",
       "      <td>ENSG00000130202</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3645.0</td>\n",
       "      <td>rs111371860-?</td>\n",
       "      <td>rs111371860</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>2.000000e-18</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125304</th>\n",
       "      <td>26053186</td>\n",
       "      <td>2015-06-08</td>\n",
       "      <td>3-hydroxypropylmercapturic acid levels in smokers</td>\n",
       "      <td>19q13.42</td>\n",
       "      <td>54650073</td>\n",
       "      <td>NR</td>\n",
       "      <td>LILRB4, LILRB4, LILRB4, LILRB4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000186818, ENSG00000276042, ENSG00000278...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs139806220-?</td>\n",
       "      <td>rs139806220</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>5.000000e-08</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90992</th>\n",
       "      <td>29875488</td>\n",
       "      <td>2018-06-06</td>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>7700507</td>\n",
       "      <td>NR</td>\n",
       "      <td>FCER2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000104921</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs113893384-C</td>\n",
       "      <td>rs113893384</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.706</td>\n",
       "      <td>5.000000e-19</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91009</th>\n",
       "      <td>29777097</td>\n",
       "      <td>2018-05-18</td>\n",
       "      <td>Family history of Alzheimer's disease</td>\n",
       "      <td>19q13.32</td>\n",
       "      <td>44723891</td>\n",
       "      <td>NR</td>\n",
       "      <td>AC243964.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000266903</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs150820726-?</td>\n",
       "      <td>rs150820726</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>3.000000e-08</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89032</th>\n",
       "      <td>29507422</td>\n",
       "      <td>2018-03-05</td>\n",
       "      <td>Total cholesterol levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>11116926</td>\n",
       "      <td>NR</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000130164</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs688-C</td>\n",
       "      <td>rs688</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.578</td>\n",
       "      <td>1.000000e-25</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158350</th>\n",
       "      <td>29507422</td>\n",
       "      <td>2018-03-05</td>\n",
       "      <td>Low density lipoprotein cholesterol levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>11116926</td>\n",
       "      <td>NR</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000130164</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs688-C</td>\n",
       "      <td>rs688</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>2.000000e-32</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158349</th>\n",
       "      <td>29507422</td>\n",
       "      <td>2018-03-05</td>\n",
       "      <td>Low density lipoprotein cholesterol levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>11116926</td>\n",
       "      <td>NR</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000130164</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs688-C</td>\n",
       "      <td>rs688</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.578</td>\n",
       "      <td>5.000000e-30</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89033</th>\n",
       "      <td>29507422</td>\n",
       "      <td>2018-03-05</td>\n",
       "      <td>Total cholesterol levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>11116926</td>\n",
       "      <td>NR</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000130164</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs688-C</td>\n",
       "      <td>rs688</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>4.000000e-28</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58731</th>\n",
       "      <td>31217584</td>\n",
       "      <td>2019-06-19</td>\n",
       "      <td>Low density lipoprotein cholesterol levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>11116926</td>\n",
       "      <td>NR</td>\n",
       "      <td>LDLR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000130164</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs688-?</td>\n",
       "      <td>rs688</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>1.000000e-08</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6817 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        pumbmed_id  study_date  \\\n",
       "64907     30698716  2019-01-29   \n",
       "91016     29777097  2018-05-18   \n",
       "125304    26053186  2015-06-08   \n",
       "90992     29875488  2018-06-06   \n",
       "91009     29777097  2018-05-18   \n",
       "...            ...         ...   \n",
       "89032     29507422  2018-03-05   \n",
       "158350    29507422  2018-03-05   \n",
       "158349    29507422  2018-03-05   \n",
       "89033     29507422  2018-03-05   \n",
       "58731     31217584  2019-06-19   \n",
       "\n",
       "                                                  disease    region   chr_pos  \\\n",
       "64907          HDL cholesterol levels in current drinkers   19p13.2   8364439   \n",
       "91016               Family history of Alzheimer's disease  19q13.32  44842530   \n",
       "125304  3-hydroxypropylmercapturic acid levels in smokers  19q13.42  54650073   \n",
       "90992                                Blood protein levels   19p13.2   7700507   \n",
       "91009               Family history of Alzheimer's disease  19q13.32  44723891   \n",
       "...                                                   ...       ...       ...   \n",
       "89032                            Total cholesterol levels   19p13.2  11116926   \n",
       "158350         Low density lipoprotein cholesterol levels   19p13.2  11116926   \n",
       "158349         Low density lipoprotein cholesterol levels   19p13.2  11116926   \n",
       "89033                            Total cholesterol levels   19p13.2  11116926   \n",
       "58731          Low density lipoprotein cholesterol levels   19p13.2  11116926   \n",
       "\n",
       "       reported_genes                    mapped_genes upstream_gene_id  \\\n",
       "64907         ANGPTL4                         ANGPTL4              NaN   \n",
       "91016              NR                  BCAM - NECTIN2  ENSG00000187244   \n",
       "125304             NR  LILRB4, LILRB4, LILRB4, LILRB4              NaN   \n",
       "90992              NR                           FCER2              NaN   \n",
       "91009              NR                      AC243964.3              NaN   \n",
       "...               ...                             ...              ...   \n",
       "89032              NR                            LDLR              NaN   \n",
       "158350             NR                            LDLR              NaN   \n",
       "158349             NR                            LDLR              NaN   \n",
       "89033              NR                            LDLR              NaN   \n",
       "58731              NR                            LDLR              NaN   \n",
       "\n",
       "       downstream_gene_id                                       snp_gene_ids  \\\n",
       "64907                 NaN                                    ENSG00000167772   \n",
       "91016     ENSG00000130202                                                NaN   \n",
       "125304                NaN  ENSG00000186818, ENSG00000276042, ENSG00000278...   \n",
       "90992                 NaN                                    ENSG00000104921   \n",
       "91009                 NaN                                    ENSG00000266903   \n",
       "...                   ...                                                ...   \n",
       "89032                 NaN                                    ENSG00000130164   \n",
       "158350                NaN                                    ENSG00000130164   \n",
       "158349                NaN                                    ENSG00000130164   \n",
       "89033                 NaN                                    ENSG00000130164   \n",
       "58731                 NaN                                    ENSG00000130164   \n",
       "\n",
       "        ...  dowstream_gene_distance  strongest_snp_risk_allele         snps  \\\n",
       "64907   ...                      NaN              rs116843064-A  rs116843064   \n",
       "91016   ...                   3645.0              rs111371860-?  rs111371860   \n",
       "125304  ...                      NaN              rs139806220-?  rs139806220   \n",
       "90992   ...                      NaN              rs113893384-C  rs113893384   \n",
       "91009   ...                      NaN              rs150820726-?  rs150820726   \n",
       "...     ...                      ...                        ...          ...   \n",
       "89032   ...                      NaN                    rs688-C        rs688   \n",
       "158350  ...                      NaN                    rs688-C        rs688   \n",
       "158349  ...                      NaN                    rs688-C        rs688   \n",
       "89033   ...                      NaN                    rs688-C        rs688   \n",
       "58731   ...                      NaN                    rs688-?        rs688   \n",
       "\n",
       "                   context intergenic  risk_allele_frequency       p_value  \\\n",
       "64907     missense_variant        0.0                     NR  2.000000e-14   \n",
       "91016   intergenic_variant        1.0                     NR  2.000000e-18   \n",
       "125304      intron_variant        0.0                     NR  5.000000e-08   \n",
       "90992       intron_variant        0.0                  0.706  5.000000e-19   \n",
       "91009       intron_variant        0.0                     NR  3.000000e-08   \n",
       "...                    ...        ...                    ...           ...   \n",
       "89032   synonymous_variant        0.0                  0.578  1.000000e-25   \n",
       "158350  synonymous_variant        0.0                     NR  2.000000e-32   \n",
       "158349  synonymous_variant        0.0                  0.578  5.000000e-30   \n",
       "89033   synonymous_variant        0.0                     NR  4.000000e-28   \n",
       "58731   synonymous_variant        0.0                     NR  1.000000e-08   \n",
       "\n",
       "        multiple_snp_semicolon  multiple_snp_colon  snp_length_str  \n",
       "64907                    False               False              11  \n",
       "91016                    False               False              11  \n",
       "125304                   False               False              11  \n",
       "90992                    False               False              11  \n",
       "91009                    False               False              11  \n",
       "...                        ...                 ...             ...  \n",
       "89032                    False               False               5  \n",
       "158350                   False               False               5  \n",
       "158349                   False               False               5  \n",
       "89033                    False               False               5  \n",
       "58731                    False               False               5  \n",
       "\n",
       "[6817 rows x 21 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_catalog['snp_length_str'] = gwas_catalog['snps'].str.len()\n",
    "gwas_catalog.sort_values('snp_length_str',ascending=False,inplace=True)\n",
    "gwas_catalog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "del gwas_catalog[\"multiple_snp_semicolon\"]\n",
    "del gwas_catalog[\"multiple_snp_colon\"]\n",
    "del gwas_catalog[\"snp_length_str\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Fundimos a informação do dataframe dos SNPs (gerada anteriormente), com a informação acerca dos SNPs derivada do dataset GWAS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene</th>\n",
       "      <th>functional_consequence</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>reference</th>\n",
       "      <th>alternative</th>\n",
       "      <th>easian_frequency</th>\n",
       "      <th>european_frequency</th>\n",
       "      <th>...</th>\n",
       "      <th>upstream_gene_id</th>\n",
       "      <th>downstream_gene_id</th>\n",
       "      <th>snp_gene_ids</th>\n",
       "      <th>upstream_gene_distance</th>\n",
       "      <th>dowstream_gene_distance</th>\n",
       "      <th>strongest_snp_risk_allele</th>\n",
       "      <th>context</th>\n",
       "      <th>intergenic</th>\n",
       "      <th>risk_allele_frequency</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs10001</td>\n",
       "      <td>STXBP2</td>\n",
       "      <td>coding_sequence_variant,genic_downstream_trans...</td>\n",
       "      <td>benign</td>\n",
       "      <td>19</td>\n",
       "      <td>7711221</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.7034</td>\n",
       "      <td>0.4036</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1000237</td>\n",
       "      <td>GATAD2A</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19518316</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.3185</td>\n",
       "      <td>0.3370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1000244</td>\n",
       "      <td>COPE</td>\n",
       "      <td>upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19030467</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1000245</td>\n",
       "      <td>COPE</td>\n",
       "      <td>intron_variant,upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19031087</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1000262</td>\n",
       "      <td>RPL18A</td>\n",
       "      <td>upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>17970627</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2470</td>\n",
       "      <td>0.2286</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index     gene                             functional_consequence  \\\n",
       "0    rs10001   STXBP2  coding_sequence_variant,genic_downstream_trans...   \n",
       "1  rs1000237  GATAD2A   intron_variant,genic_upstream_transcript_variant   \n",
       "2  rs1000244     COPE                        upstream_transcript_variant   \n",
       "3  rs1000245     COPE         intron_variant,upstream_transcript_variant   \n",
       "4  rs1000262   RPL18A                        upstream_transcript_variant   \n",
       "\n",
       "  clinical_significance  chromosome  snp_position reference alternative  \\\n",
       "0                benign          19       7711221         T           C   \n",
       "1                  None          19      19518316         T           A   \n",
       "2                  None          19      19030467         A           G   \n",
       "3                  None          19      19031087         C           G   \n",
       "4                  None          19      17970627         C           G   \n",
       "\n",
       "   easian_frequency  european_frequency  ...  upstream_gene_id  \\\n",
       "0            0.7034              0.4036  ...               NaN   \n",
       "1            0.3185              0.3370  ...               NaN   \n",
       "2            0.0000              0.0000  ...               NaN   \n",
       "3            0.0000              0.0000  ...               NaN   \n",
       "4            0.2470              0.2286  ...               NaN   \n",
       "\n",
       "   downstream_gene_id  snp_gene_ids  upstream_gene_distance  \\\n",
       "0                 NaN           NaN                     NaN   \n",
       "1                 NaN           NaN                     NaN   \n",
       "2                 NaN           NaN                     NaN   \n",
       "3                 NaN           NaN                     NaN   \n",
       "4                 NaN           NaN                     NaN   \n",
       "\n",
       "  dowstream_gene_distance strongest_snp_risk_allele context intergenic  \\\n",
       "0                     NaN                       NaN     NaN        NaN   \n",
       "1                     NaN                       NaN     NaN        NaN   \n",
       "2                     NaN                       NaN     NaN        NaN   \n",
       "3                     NaN                       NaN     NaN        NaN   \n",
       "4                     NaN                       NaN     NaN        NaN   \n",
       "\n",
       "  risk_allele_frequency p_value  \n",
       "0                   NaN     NaN  \n",
       "1                   NaN     NaN  \n",
       "2                   NaN     NaN  \n",
       "3                   NaN     NaN  \n",
       "4                   NaN     NaN  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = df1.set_index('snp_ref_code').join(gwas_catalog.set_index('snps'))\n",
    "\n",
    "df2 = df2.reset_index()\n",
    "df2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['index', 'gene', 'functional_consequence', 'clinical_significance',\n",
       "       'chromosome', 'snp_position', 'reference', 'alternative',\n",
       "       'easian_frequency', 'european_frequency', 'african_frequency',\n",
       "       'american_frequency', 'sasian_frequency', 'pumbmed_id', 'study_date',\n",
       "       'disease', 'region', 'chr_pos', 'reported_genes', 'mapped_genes',\n",
       "       'upstream_gene_id', 'downstream_gene_id', 'snp_gene_ids',\n",
       "       'upstream_gene_distance', 'dowstream_gene_distance',\n",
       "       'strongest_snp_risk_allele', 'context', 'intergenic',\n",
       "       'risk_allele_frequency', 'p_value'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "del df2[\"pumbmed_id\"]\n",
    "del df2[\"study_date\"]\n",
    "del df2[\"strongest_snp_risk_allele\"]\n",
    "del df2[\"risk_allele_frequency\"]\n",
    "del df2[\"p_value\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene</th>\n",
       "      <th>functional_consequence</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>reference</th>\n",
       "      <th>alternative</th>\n",
       "      <th>easian_frequency</th>\n",
       "      <th>european_frequency</th>\n",
       "      <th>...</th>\n",
       "      <th>chr_pos</th>\n",
       "      <th>reported_genes</th>\n",
       "      <th>mapped_genes</th>\n",
       "      <th>upstream_gene_id</th>\n",
       "      <th>downstream_gene_id</th>\n",
       "      <th>snp_gene_ids</th>\n",
       "      <th>upstream_gene_distance</th>\n",
       "      <th>dowstream_gene_distance</th>\n",
       "      <th>context</th>\n",
       "      <th>intergenic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs10001</td>\n",
       "      <td>STXBP2</td>\n",
       "      <td>coding_sequence_variant,genic_downstream_trans...</td>\n",
       "      <td>benign</td>\n",
       "      <td>19</td>\n",
       "      <td>7711221</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.7034</td>\n",
       "      <td>0.4036</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1000237</td>\n",
       "      <td>GATAD2A</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19518316</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.3185</td>\n",
       "      <td>0.3370</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1000244</td>\n",
       "      <td>COPE</td>\n",
       "      <td>upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19030467</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1000245</td>\n",
       "      <td>COPE</td>\n",
       "      <td>intron_variant,upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19031087</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1000262</td>\n",
       "      <td>RPL18A</td>\n",
       "      <td>upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>17970627</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2470</td>\n",
       "      <td>0.2286</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1710170</th>\n",
       "      <td>rs999814</td>\n",
       "      <td>IQCN</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>18368969</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.2371</td>\n",
       "      <td>0.2724</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1710171</th>\n",
       "      <td>rs999815</td>\n",
       "      <td>IQCN</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>18368988</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.2381</td>\n",
       "      <td>0.2704</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1710172</th>\n",
       "      <td>rs999850</td>\n",
       "      <td>DPY19L3</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>32901045</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.6399</td>\n",
       "      <td>0.6402</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1710173</th>\n",
       "      <td>rs999854</td>\n",
       "      <td>CACNA1A</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>13616206</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.2569</td>\n",
       "      <td>0.0109</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1710174</th>\n",
       "      <td>rs999967</td>\n",
       "      <td>ADGRE2</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>14863522</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.5169</td>\n",
       "      <td>0.2654</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1710175 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             index     gene  \\\n",
       "0          rs10001   STXBP2   \n",
       "1        rs1000237  GATAD2A   \n",
       "2        rs1000244     COPE   \n",
       "3        rs1000245     COPE   \n",
       "4        rs1000262   RPL18A   \n",
       "...            ...      ...   \n",
       "1710170   rs999814     IQCN   \n",
       "1710171   rs999815     IQCN   \n",
       "1710172   rs999850  DPY19L3   \n",
       "1710173   rs999854  CACNA1A   \n",
       "1710174   rs999967   ADGRE2   \n",
       "\n",
       "                                    functional_consequence  \\\n",
       "0        coding_sequence_variant,genic_downstream_trans...   \n",
       "1         intron_variant,genic_upstream_transcript_variant   \n",
       "2                              upstream_transcript_variant   \n",
       "3               intron_variant,upstream_transcript_variant   \n",
       "4                              upstream_transcript_variant   \n",
       "...                                                    ...   \n",
       "1710170                                     intron_variant   \n",
       "1710171                                     intron_variant   \n",
       "1710172   intron_variant,genic_upstream_transcript_variant   \n",
       "1710173                                     intron_variant   \n",
       "1710174                                     intron_variant   \n",
       "\n",
       "        clinical_significance  chromosome  snp_position reference alternative  \\\n",
       "0                      benign          19       7711221         T           C   \n",
       "1                        None          19      19518316         T           A   \n",
       "2                        None          19      19030467         A           G   \n",
       "3                        None          19      19031087         C           G   \n",
       "4                        None          19      17970627         C           G   \n",
       "...                       ...         ...           ...       ...         ...   \n",
       "1710170                  None          19      18368969         G           A   \n",
       "1710171                  None          19      18368988         G           A   \n",
       "1710172                  None          19      32901045         T           A   \n",
       "1710173                  None          19      13616206         G           A   \n",
       "1710174                  None          19      14863522         A           G   \n",
       "\n",
       "         easian_frequency  european_frequency  ...  chr_pos  reported_genes  \\\n",
       "0                  0.7034              0.4036  ...      NaN             NaN   \n",
       "1                  0.3185              0.3370  ...      NaN             NaN   \n",
       "2                  0.0000              0.0000  ...      NaN             NaN   \n",
       "3                  0.0000              0.0000  ...      NaN             NaN   \n",
       "4                  0.2470              0.2286  ...      NaN             NaN   \n",
       "...                   ...                 ...  ...      ...             ...   \n",
       "1710170            0.2371              0.2724  ...      NaN             NaN   \n",
       "1710171            0.2381              0.2704  ...      NaN             NaN   \n",
       "1710172            0.6399              0.6402  ...      NaN             NaN   \n",
       "1710173            0.2569              0.0109  ...      NaN             NaN   \n",
       "1710174            0.5169              0.2654  ...      NaN             NaN   \n",
       "\n",
       "         mapped_genes upstream_gene_id downstream_gene_id snp_gene_ids  \\\n",
       "0                 NaN              NaN                NaN          NaN   \n",
       "1                 NaN              NaN                NaN          NaN   \n",
       "2                 NaN              NaN                NaN          NaN   \n",
       "3                 NaN              NaN                NaN          NaN   \n",
       "4                 NaN              NaN                NaN          NaN   \n",
       "...               ...              ...                ...          ...   \n",
       "1710170           NaN              NaN                NaN          NaN   \n",
       "1710171           NaN              NaN                NaN          NaN   \n",
       "1710172           NaN              NaN                NaN          NaN   \n",
       "1710173           NaN              NaN                NaN          NaN   \n",
       "1710174           NaN              NaN                NaN          NaN   \n",
       "\n",
       "        upstream_gene_distance dowstream_gene_distance context intergenic  \n",
       "0                          NaN                     NaN     NaN        NaN  \n",
       "1                          NaN                     NaN     NaN        NaN  \n",
       "2                          NaN                     NaN     NaN        NaN  \n",
       "3                          NaN                     NaN     NaN        NaN  \n",
       "4                          NaN                     NaN     NaN        NaN  \n",
       "...                        ...                     ...     ...        ...  \n",
       "1710170                    NaN                     NaN     NaN        NaN  \n",
       "1710171                    NaN                     NaN     NaN        NaN  \n",
       "1710172                    NaN                     NaN     NaN        NaN  \n",
       "1710173                    NaN                     NaN     NaN        NaN  \n",
       "1710174                    NaN                     NaN     NaN        NaN  \n",
       "\n",
       "[1710175 rows x 25 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2['chromosome'] = df2['chromosome'].astype(int)\n",
    "\n",
    "df2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- O processo de introdução no SQL dos cerca de 1.7 milhões de SNPs é muito moroso. Por essa razão, optamos por remover todas as instâncias de SNP que sabíamos não estar associados a doenças (por informação derivada da Study_dimension).\n",
    "- Isto dá-nos 6557 SNPs no cromossoma 19, que têm uma associação genética identificada na Study Dimension.\n",
    "- Idealmente, todos os SNPs deveriam ter sido introduzidos na SNP Dimension."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df2[df2.disease.notnull()]\n",
    "del df2[\"disease\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene</th>\n",
       "      <th>functional_consequence</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>reference</th>\n",
       "      <th>alternative</th>\n",
       "      <th>easian_frequency</th>\n",
       "      <th>european_frequency</th>\n",
       "      <th>...</th>\n",
       "      <th>chr_pos</th>\n",
       "      <th>reported_genes</th>\n",
       "      <th>mapped_genes</th>\n",
       "      <th>upstream_gene_id</th>\n",
       "      <th>downstream_gene_id</th>\n",
       "      <th>snp_gene_ids</th>\n",
       "      <th>upstream_gene_distance</th>\n",
       "      <th>dowstream_gene_distance</th>\n",
       "      <th>context</th>\n",
       "      <th>intergenic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>DOT1L</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>2206575</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1687</td>\n",
       "      <td>0.3241</td>\n",
       "      <td>...</td>\n",
       "      <td>2206576</td>\n",
       "      <td>NR</td>\n",
       "      <td>DOT1L</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000104885</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>CRTC1</td>\n",
       "      <td>3_prime_UTR_variant,genic_downstream_transcrip...</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>18892728</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.1042</td>\n",
       "      <td>0.2823</td>\n",
       "      <td>...</td>\n",
       "      <td>18781918</td>\n",
       "      <td>COMP</td>\n",
       "      <td>CRTC1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000105662</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>16456260</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.6062</td>\n",
       "      <td>0.7296</td>\n",
       "      <td>...</td>\n",
       "      <td>1.63454e+07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KLF2 - AC020917.3</td>\n",
       "      <td>ENSG00000127528</td>\n",
       "      <td>ENSG00000277825</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17575.0</td>\n",
       "      <td>7013.0</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>rs10048489</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>40343129</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.1012</td>\n",
       "      <td>0.0825</td>\n",
       "      <td>...</td>\n",
       "      <td>39852489</td>\n",
       "      <td>FCGBP</td>\n",
       "      <td>FBL - FCGBP</td>\n",
       "      <td>ENSG00000280548</td>\n",
       "      <td>ENSG00000281123</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6110.0</td>\n",
       "      <td>10834.0</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>rs1005636</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>43806559</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9990</td>\n",
       "      <td>0.8201</td>\n",
       "      <td>...</td>\n",
       "      <td>43302407</td>\n",
       "      <td>NR</td>\n",
       "      <td>CEACAMP11 - CEACAMP4</td>\n",
       "      <td>ENSG00000236123</td>\n",
       "      <td>ENSG00000230681</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18089.0</td>\n",
       "      <td>764.0</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1709685</th>\n",
       "      <td>rs9797885</td>\n",
       "      <td>TMEM91</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>41873001</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.4464</td>\n",
       "      <td>0.6750</td>\n",
       "      <td>...</td>\n",
       "      <td>41367096</td>\n",
       "      <td>NR</td>\n",
       "      <td>AC011462.1, TMEM91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000255730, ENSG00000142046</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1709721</th>\n",
       "      <td>rs9807841</td>\n",
       "      <td>QTRT1</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>10817656</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.7242</td>\n",
       "      <td>0.6024</td>\n",
       "      <td>...</td>\n",
       "      <td>10706980</td>\n",
       "      <td>QTRT1</td>\n",
       "      <td>QTRT1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000213339</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1709995</th>\n",
       "      <td>rs995834</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>28866596</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.5109</td>\n",
       "      <td>0.7137</td>\n",
       "      <td>...</td>\n",
       "      <td>28375689</td>\n",
       "      <td>LOC100420587</td>\n",
       "      <td>AC005580.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000267320</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1710023</th>\n",
       "      <td>rs9967620</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>46415696</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.0873</td>\n",
       "      <td>0.1779</td>\n",
       "      <td>...</td>\n",
       "      <td>45912438</td>\n",
       "      <td>GIPR</td>\n",
       "      <td>MYPOP - NANOS2</td>\n",
       "      <td>ENSG00000176182</td>\n",
       "      <td>ENSG00000188425</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9825.0</td>\n",
       "      <td>776.0</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1710121</th>\n",
       "      <td>rs998732</td>\n",
       "      <td>TM6SF2</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19</td>\n",
       "      <td>19378671</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1670</td>\n",
       "      <td>...</td>\n",
       "      <td>19267862</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TM6SF2, AC138430.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000213996, ENSG00000267629</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6557 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              index    gene  \\\n",
       "53        rs1003531   DOT1L   \n",
       "57          rs10038   CRTC1   \n",
       "70        rs1004246    None   \n",
       "90       rs10048489    None   \n",
       "140       rs1005636    None   \n",
       "...             ...     ...   \n",
       "1709685   rs9797885  TMEM91   \n",
       "1709721   rs9807841   QTRT1   \n",
       "1709995    rs995834    None   \n",
       "1710023   rs9967620    None   \n",
       "1710121    rs998732  TM6SF2   \n",
       "\n",
       "                                    functional_consequence  \\\n",
       "53                                          intron_variant   \n",
       "57       3_prime_UTR_variant,genic_downstream_transcrip...   \n",
       "70                                                    None   \n",
       "90                                                    None   \n",
       "140                                                   None   \n",
       "...                                                    ...   \n",
       "1709685   intron_variant,genic_upstream_transcript_variant   \n",
       "1709721                                     intron_variant   \n",
       "1709995                                               None   \n",
       "1710023                                               None   \n",
       "1710121                                     intron_variant   \n",
       "\n",
       "        clinical_significance  chromosome  snp_position reference alternative  \\\n",
       "53                       None          19       2206575         G           T   \n",
       "57                       None          19      18892728         C           A   \n",
       "70                       None          19      16456260         C           T   \n",
       "90                       None          19      40343129         G           C   \n",
       "140                      None          19      43806559         T           C   \n",
       "...                       ...         ...           ...       ...         ...   \n",
       "1709685                  None          19      41873001         A           G   \n",
       "1709721                  None          19      10817656         T           C   \n",
       "1709995                  None          19      28866596         G           A   \n",
       "1710023                  None          19      46415696         C           T   \n",
       "1710121                  None          19      19378671         A           G   \n",
       "\n",
       "         easian_frequency  european_frequency  ...      chr_pos  \\\n",
       "53                 0.1687              0.3241  ...      2206576   \n",
       "57                 0.1042              0.2823  ...     18781918   \n",
       "70                 0.6062              0.7296  ...  1.63454e+07   \n",
       "90                 0.1012              0.0825  ...     39852489   \n",
       "140                0.9990              0.8201  ...     43302407   \n",
       "...                   ...                 ...  ...          ...   \n",
       "1709685            0.4464              0.6750  ...     41367096   \n",
       "1709721            0.7242              0.6024  ...     10706980   \n",
       "1709995            0.5109              0.7137  ...     28375689   \n",
       "1710023            0.0873              0.1779  ...     45912438   \n",
       "1710121            0.0000              0.1670  ...     19267862   \n",
       "\n",
       "         reported_genes          mapped_genes upstream_gene_id  \\\n",
       "53                   NR                 DOT1L              NaN   \n",
       "57                 COMP                 CRTC1              NaN   \n",
       "70                  NaN     KLF2 - AC020917.3  ENSG00000127528   \n",
       "90                FCGBP           FBL - FCGBP  ENSG00000280548   \n",
       "140                  NR  CEACAMP11 - CEACAMP4  ENSG00000236123   \n",
       "...                 ...                   ...              ...   \n",
       "1709685              NR    AC011462.1, TMEM91              NaN   \n",
       "1709721           QTRT1                 QTRT1              NaN   \n",
       "1709995    LOC100420587            AC005580.1              NaN   \n",
       "1710023            GIPR        MYPOP - NANOS2  ENSG00000176182   \n",
       "1710121             NaN    TM6SF2, AC138430.1              NaN   \n",
       "\n",
       "        downstream_gene_id                      snp_gene_ids  \\\n",
       "53                     NaN                   ENSG00000104885   \n",
       "57                     NaN                   ENSG00000105662   \n",
       "70         ENSG00000277825                               NaN   \n",
       "90         ENSG00000281123                               NaN   \n",
       "140        ENSG00000230681                               NaN   \n",
       "...                    ...                               ...   \n",
       "1709685                NaN  ENSG00000255730, ENSG00000142046   \n",
       "1709721                NaN                   ENSG00000213339   \n",
       "1709995                NaN                   ENSG00000267320   \n",
       "1710023    ENSG00000188425                               NaN   \n",
       "1710121                NaN  ENSG00000213996, ENSG00000267629   \n",
       "\n",
       "        upstream_gene_distance dowstream_gene_distance              context  \\\n",
       "53                         NaN                     NaN       intron_variant   \n",
       "57                         NaN                     NaN  3_prime_UTR_variant   \n",
       "70                     17575.0                  7013.0   intergenic_variant   \n",
       "90                      6110.0                 10834.0   intergenic_variant   \n",
       "140                    18089.0                   764.0   intergenic_variant   \n",
       "...                        ...                     ...                  ...   \n",
       "1709685                    NaN                     NaN       intron_variant   \n",
       "1709721                    NaN                     NaN       intron_variant   \n",
       "1709995                    NaN                     NaN       intron_variant   \n",
       "1710023                 9825.0                   776.0   intergenic_variant   \n",
       "1710121                    NaN                     NaN       intron_variant   \n",
       "\n",
       "        intergenic  \n",
       "53             0.0  \n",
       "57             0.0  \n",
       "70             1.0  \n",
       "90             1.0  \n",
       "140            1.0  \n",
       "...            ...  \n",
       "1709685        0.0  \n",
       "1709721        0.0  \n",
       "1709995        0.0  \n",
       "1710023        1.0  \n",
       "1710121        0.0  \n",
       "\n",
       "[6557 rows x 24 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_catalog['reported_genes'] = '{' + gwas_catalog['reported_genes'].astype(str) + '}'\n",
    "gwas_catalog['mapped_genes'] = '{' + gwas_catalog['mapped_genes'].astype(str) + '}'\n",
    "gwas_catalog['snp_gene_ids'] = '{' + gwas_catalog['snp_gene_ids'].astype(str) + '}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df2.rename(columns={\"index\": \"snp_ref_code\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "del df2['reported_genes']\n",
    "del df2['mapped_genes']\n",
    "del df2['snp_gene_ids']\n",
    "del df2['intergenic']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['snp_ref_code', 'gene', 'functional_consequence',\n",
      "       'clinical_significance', 'chromosome', 'snp_position', 'reference',\n",
      "       'alternative', 'easian_frequency', 'european_frequency',\n",
      "       'african_frequency', 'american_frequency', 'sasian_frequency', 'region',\n",
      "       'chr_pos', 'upstream_gene_id', 'downstream_gene_id',\n",
      "       'upstream_gene_distance', 'dowstream_gene_distance', 'context'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "53              intron_variant\n",
       "57         3_prime_UTR_variant\n",
       "70          intergenic_variant\n",
       "90          intergenic_variant\n",
       "140         intergenic_variant\n",
       "                  ...         \n",
       "1709685         intron_variant\n",
       "1709721         intron_variant\n",
       "1709995         intron_variant\n",
       "1710023     intergenic_variant\n",
       "1710121         intron_variant\n",
       "Name: context, Length: 6557, dtype: object"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df2.columns)\n",
    "df2['context']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- O código seguinte, insere os dados filtrados dos estudos GWAS, na Study Dimension."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "connectToServer()\n",
    "curs = conn.cursor()\n",
    "\n",
    "cols = \",\".join([str(i) for i in gwas_catalog.columns.tolist()])\n",
    "\n",
    "lines = 0\n",
    "for i,row in gwas_catalog.iterrows():\n",
    "    try:\n",
    "        sql = \"INSERT INTO study_dim (\" + cols + \") VALUES (\" + \"%s,\"*(len(row)-1) + \"%s)\"\n",
    "        #print(sql, tuple(row))\n",
    "        curs.execute(sql, tuple(row))\n",
    "        lines += 1\n",
    "        if lines % 101:\n",
    "            conn.commit()\n",
    "        #if lines > 38:\n",
    "        #    break\n",
    "    except Exception as e:\n",
    "        print(e,tuple(row))\n",
    "\n",
    "curs.close()\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>pumbmed_id</th>\n",
       "      <th>study_date</th>\n",
       "      <th>disease</th>\n",
       "      <th>region</th>\n",
       "      <th>chr_pos</th>\n",
       "      <th>reported_genes</th>\n",
       "      <th>mapped_genes</th>\n",
       "      <th>upstream_gene_id</th>\n",
       "      <th>downstream_gene_id</th>\n",
       "      <th>snp_gene_ids</th>\n",
       "      <th>upstream_gene_distance</th>\n",
       "      <th>dowstream_gene_distance</th>\n",
       "      <th>strongest_snp_risk_allele</th>\n",
       "      <th>snps</th>\n",
       "      <th>snp_id_current</th>\n",
       "      <th>context</th>\n",
       "      <th>intergenic</th>\n",
       "      <th>risk_allele_frequency</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>30698716</td>\n",
       "      <td>2019-01-29</td>\n",
       "      <td>HDL cholesterol levels in current drinkers</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>8364439</td>\n",
       "      <td>[ANGPTL4]</td>\n",
       "      <td>[ANGPTL4]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000167772]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs116843064-A</td>\n",
       "      <td>rs116843064</td>\n",
       "      <td>None</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>2.000000e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>29777097</td>\n",
       "      <td>2018-05-18</td>\n",
       "      <td>Family history of Alzheimer's disease</td>\n",
       "      <td>19q13.32</td>\n",
       "      <td>44842530</td>\n",
       "      <td>[NR]</td>\n",
       "      <td>[BCAM - NECTIN2]</td>\n",
       "      <td>ENSG00000187244</td>\n",
       "      <td>ENSG00000130202</td>\n",
       "      <td>[nan]</td>\n",
       "      <td>21109.0</td>\n",
       "      <td>3645.0</td>\n",
       "      <td>rs111371860-?</td>\n",
       "      <td>rs111371860</td>\n",
       "      <td>None</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>2.000000e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>26053186</td>\n",
       "      <td>2015-06-08</td>\n",
       "      <td>3-hydroxypropylmercapturic acid levels in smokers</td>\n",
       "      <td>19q13.42</td>\n",
       "      <td>54650073</td>\n",
       "      <td>[NR]</td>\n",
       "      <td>[LILRB4, LILRB4, LILRB4, LILRB4]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000186818, ENSG00000276042, ENSG0000027...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs139806220-?</td>\n",
       "      <td>rs139806220</td>\n",
       "      <td>None</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>5.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>29875488</td>\n",
       "      <td>2018-06-06</td>\n",
       "      <td>Blood protein levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>7700507</td>\n",
       "      <td>[NR]</td>\n",
       "      <td>[FCER2]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000104921]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs113893384-C</td>\n",
       "      <td>rs113893384</td>\n",
       "      <td>None</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.706</td>\n",
       "      <td>5.000000e-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>29777097</td>\n",
       "      <td>2018-05-18</td>\n",
       "      <td>Family history of Alzheimer's disease</td>\n",
       "      <td>19q13.32</td>\n",
       "      <td>44723891</td>\n",
       "      <td>[NR]</td>\n",
       "      <td>[AC243964.3]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000266903]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs150820726-?</td>\n",
       "      <td>rs150820726</td>\n",
       "      <td>None</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>3.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6812</th>\n",
       "      <td>6813</td>\n",
       "      <td>29507422</td>\n",
       "      <td>2018-03-05</td>\n",
       "      <td>Total cholesterol levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>11116926</td>\n",
       "      <td>[NR]</td>\n",
       "      <td>[LDLR]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000130164]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs688-C</td>\n",
       "      <td>rs688</td>\n",
       "      <td>None</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.578</td>\n",
       "      <td>1.000000e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6813</th>\n",
       "      <td>6814</td>\n",
       "      <td>29507422</td>\n",
       "      <td>2018-03-05</td>\n",
       "      <td>Low density lipoprotein cholesterol levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>11116926</td>\n",
       "      <td>[NR]</td>\n",
       "      <td>[LDLR]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000130164]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs688-C</td>\n",
       "      <td>rs688</td>\n",
       "      <td>None</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>2.000000e-32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6814</th>\n",
       "      <td>6815</td>\n",
       "      <td>29507422</td>\n",
       "      <td>2018-03-05</td>\n",
       "      <td>Low density lipoprotein cholesterol levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>11116926</td>\n",
       "      <td>[NR]</td>\n",
       "      <td>[LDLR]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000130164]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs688-C</td>\n",
       "      <td>rs688</td>\n",
       "      <td>None</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.578</td>\n",
       "      <td>5.000000e-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6815</th>\n",
       "      <td>6816</td>\n",
       "      <td>29507422</td>\n",
       "      <td>2018-03-05</td>\n",
       "      <td>Total cholesterol levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>11116926</td>\n",
       "      <td>[NR]</td>\n",
       "      <td>[LDLR]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000130164]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs688-C</td>\n",
       "      <td>rs688</td>\n",
       "      <td>None</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>4.000000e-28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6816</th>\n",
       "      <td>6817</td>\n",
       "      <td>31217584</td>\n",
       "      <td>2019-06-19</td>\n",
       "      <td>Low density lipoprotein cholesterol levels</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>11116926</td>\n",
       "      <td>[NR]</td>\n",
       "      <td>[LDLR]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[ENSG00000130164]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>rs688-?</td>\n",
       "      <td>rs688</td>\n",
       "      <td>None</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NR</td>\n",
       "      <td>1.000000e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6817 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      study_id  pumbmed_id  study_date  \\\n",
       "0            1    30698716  2019-01-29   \n",
       "1            2    29777097  2018-05-18   \n",
       "2            3    26053186  2015-06-08   \n",
       "3            4    29875488  2018-06-06   \n",
       "4            5    29777097  2018-05-18   \n",
       "...        ...         ...         ...   \n",
       "6812      6813    29507422  2018-03-05   \n",
       "6813      6814    29507422  2018-03-05   \n",
       "6814      6815    29507422  2018-03-05   \n",
       "6815      6816    29507422  2018-03-05   \n",
       "6816      6817    31217584  2019-06-19   \n",
       "\n",
       "                                                disease    region   chr_pos  \\\n",
       "0            HDL cholesterol levels in current drinkers   19p13.2   8364439   \n",
       "1                 Family history of Alzheimer's disease  19q13.32  44842530   \n",
       "2     3-hydroxypropylmercapturic acid levels in smokers  19q13.42  54650073   \n",
       "3                                  Blood protein levels   19p13.2   7700507   \n",
       "4                 Family history of Alzheimer's disease  19q13.32  44723891   \n",
       "...                                                 ...       ...       ...   \n",
       "6812                           Total cholesterol levels   19p13.2  11116926   \n",
       "6813         Low density lipoprotein cholesterol levels   19p13.2  11116926   \n",
       "6814         Low density lipoprotein cholesterol levels   19p13.2  11116926   \n",
       "6815                           Total cholesterol levels   19p13.2  11116926   \n",
       "6816         Low density lipoprotein cholesterol levels   19p13.2  11116926   \n",
       "\n",
       "     reported_genes                      mapped_genes upstream_gene_id  \\\n",
       "0         [ANGPTL4]                         [ANGPTL4]              NaN   \n",
       "1              [NR]                  [BCAM - NECTIN2]  ENSG00000187244   \n",
       "2              [NR]  [LILRB4, LILRB4, LILRB4, LILRB4]              NaN   \n",
       "3              [NR]                           [FCER2]              NaN   \n",
       "4              [NR]                      [AC243964.3]              NaN   \n",
       "...             ...                               ...              ...   \n",
       "6812           [NR]                            [LDLR]              NaN   \n",
       "6813           [NR]                            [LDLR]              NaN   \n",
       "6814           [NR]                            [LDLR]              NaN   \n",
       "6815           [NR]                            [LDLR]              NaN   \n",
       "6816           [NR]                            [LDLR]              NaN   \n",
       "\n",
       "     downstream_gene_id                                       snp_gene_ids  \\\n",
       "0                   NaN                                  [ENSG00000167772]   \n",
       "1       ENSG00000130202                                              [nan]   \n",
       "2                   NaN  [ENSG00000186818, ENSG00000276042, ENSG0000027...   \n",
       "3                   NaN                                  [ENSG00000104921]   \n",
       "4                   NaN                                  [ENSG00000266903]   \n",
       "...                 ...                                                ...   \n",
       "6812                NaN                                  [ENSG00000130164]   \n",
       "6813                NaN                                  [ENSG00000130164]   \n",
       "6814                NaN                                  [ENSG00000130164]   \n",
       "6815                NaN                                  [ENSG00000130164]   \n",
       "6816                NaN                                  [ENSG00000130164]   \n",
       "\n",
       "     upstream_gene_distance dowstream_gene_distance strongest_snp_risk_allele  \\\n",
       "0                       NaN                     NaN             rs116843064-A   \n",
       "1                   21109.0                  3645.0             rs111371860-?   \n",
       "2                       NaN                     NaN             rs139806220-?   \n",
       "3                       NaN                     NaN             rs113893384-C   \n",
       "4                       NaN                     NaN             rs150820726-?   \n",
       "...                     ...                     ...                       ...   \n",
       "6812                    NaN                     NaN                   rs688-C   \n",
       "6813                    NaN                     NaN                   rs688-C   \n",
       "6814                    NaN                     NaN                   rs688-C   \n",
       "6815                    NaN                     NaN                   rs688-C   \n",
       "6816                    NaN                     NaN                   rs688-?   \n",
       "\n",
       "             snps snp_id_current             context intergenic  \\\n",
       "0     rs116843064           None    missense_variant        0.0   \n",
       "1     rs111371860           None  intergenic_variant        1.0   \n",
       "2     rs139806220           None      intron_variant        0.0   \n",
       "3     rs113893384           None      intron_variant        0.0   \n",
       "4     rs150820726           None      intron_variant        0.0   \n",
       "...           ...            ...                 ...        ...   \n",
       "6812        rs688           None  synonymous_variant        0.0   \n",
       "6813        rs688           None  synonymous_variant        0.0   \n",
       "6814        rs688           None  synonymous_variant        0.0   \n",
       "6815        rs688           None  synonymous_variant        0.0   \n",
       "6816        rs688           None  synonymous_variant        0.0   \n",
       "\n",
       "     risk_allele_frequency       p_value  \n",
       "0                       NR  2.000000e-14  \n",
       "1                       NR  2.000000e-18  \n",
       "2                       NR  5.000000e-08  \n",
       "3                    0.706  5.000000e-19  \n",
       "4                       NR  3.000000e-08  \n",
       "...                    ...           ...  \n",
       "6812                 0.578  1.000000e-25  \n",
       "6813                    NR  2.000000e-32  \n",
       "6814                 0.578  5.000000e-30  \n",
       "6815                    NR  4.000000e-28  \n",
       "6816                    NR  1.000000e-08  \n",
       "\n",
       "[6817 rows x 20 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select * from study_dim\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- O código seguinte, insere os dados acerca dos SNPs (filtrados, como mencionado previamente), na SNP dimension."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "connectToServer()\n",
    "curs = conn.cursor()\n",
    "\n",
    "cols = \",\".join([str(i) for i in df2.columns.tolist()])\n",
    "\n",
    "lines = 0\n",
    "for i,row in df2.iterrows():\n",
    "    try:\n",
    "        sql = \"INSERT INTO snp_dim (\" + cols + \") VALUES (\" + \"%s,\"*(len(row)-1) + \"%s)\"\n",
    "        #print(sql, tuple(row))\n",
    "        curs.execute(sql, tuple(row))\n",
    "        lines += 1\n",
    "        if lines % 101:\n",
    "            conn.commit()\n",
    "        #if lines > 2:\n",
    "        #    break\n",
    "    except Exception as e:\n",
    "        print(e,tuple(row))\n",
    "\n",
    "curs.close()\n",
    "conn.commit()\n",
    "conn.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>gene</th>\n",
       "      <th>functional_consequence</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>snp_position</th>\n",
       "      <th>reference</th>\n",
       "      <th>alternative</th>\n",
       "      <th>info_string</th>\n",
       "      <th>...</th>\n",
       "      <th>sasian_frequency</th>\n",
       "      <th>variation_type</th>\n",
       "      <th>consequence</th>\n",
       "      <th>mapped_gene</th>\n",
       "      <th>upstream_gene_id</th>\n",
       "      <th>downstream_gene_id</th>\n",
       "      <th>upstream_gene_distance</th>\n",
       "      <th>dowstream_gene_distance</th>\n",
       "      <th>region</th>\n",
       "      <th>chr_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs1003531</td>\n",
       "      <td>DOT1L</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19.0</td>\n",
       "      <td>2206575</td>\n",
       "      <td>G                                             ...</td>\n",
       "      <td>T                                             ...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1626</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19p13.3</td>\n",
       "      <td>2206576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>rs10038</td>\n",
       "      <td>CRTC1</td>\n",
       "      <td>3_prime_UTR_variant,genic_downstream_transcrip...</td>\n",
       "      <td>None</td>\n",
       "      <td>19.0</td>\n",
       "      <td>18892728</td>\n",
       "      <td>C                                             ...</td>\n",
       "      <td>A                                             ...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1442</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19p13.11</td>\n",
       "      <td>18781918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>rs1004246</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19.0</td>\n",
       "      <td>16456260</td>\n",
       "      <td>C                                             ...</td>\n",
       "      <td>T                                             ...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0.8006</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000127528</td>\n",
       "      <td>ENSG00000277825</td>\n",
       "      <td>17575.0</td>\n",
       "      <td>7013.0</td>\n",
       "      <td>19p13.11</td>\n",
       "      <td>16345449.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>rs10048489</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19.0</td>\n",
       "      <td>40343129</td>\n",
       "      <td>G                                             ...</td>\n",
       "      <td>C                                             ...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0358</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000280548</td>\n",
       "      <td>ENSG00000281123</td>\n",
       "      <td>6110.0</td>\n",
       "      <td>10834.0</td>\n",
       "      <td>19q13.2</td>\n",
       "      <td>39852489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>rs1005636</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19.0</td>\n",
       "      <td>43806559</td>\n",
       "      <td>T                                             ...</td>\n",
       "      <td>C                                             ...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0.9110</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000236123</td>\n",
       "      <td>ENSG00000230681</td>\n",
       "      <td>18089.0</td>\n",
       "      <td>764.0</td>\n",
       "      <td>19q13.31</td>\n",
       "      <td>43302407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6552</th>\n",
       "      <td>6553</td>\n",
       "      <td>rs9797885</td>\n",
       "      <td>TMEM91</td>\n",
       "      <td>intron_variant,genic_upstream_transcript_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19.0</td>\n",
       "      <td>41873001</td>\n",
       "      <td>A                                             ...</td>\n",
       "      <td>G                                             ...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6207</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19q13.2</td>\n",
       "      <td>41367096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6553</th>\n",
       "      <td>6554</td>\n",
       "      <td>rs9807841</td>\n",
       "      <td>QTRT1</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19.0</td>\n",
       "      <td>10817656</td>\n",
       "      <td>T                                             ...</td>\n",
       "      <td>C                                             ...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0.7239</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19p13.2</td>\n",
       "      <td>10706980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6554</th>\n",
       "      <td>6555</td>\n",
       "      <td>rs995834</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19.0</td>\n",
       "      <td>28866596</td>\n",
       "      <td>G                                             ...</td>\n",
       "      <td>A                                             ...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0.6350</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19q12</td>\n",
       "      <td>28375689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6555</th>\n",
       "      <td>6556</td>\n",
       "      <td>rs9967620</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>19.0</td>\n",
       "      <td>46415696</td>\n",
       "      <td>C                                             ...</td>\n",
       "      <td>T                                             ...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1646</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>ENSG00000176182</td>\n",
       "      <td>ENSG00000188425</td>\n",
       "      <td>9825.0</td>\n",
       "      <td>776.0</td>\n",
       "      <td>19q13.32</td>\n",
       "      <td>45912438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6556</th>\n",
       "      <td>6557</td>\n",
       "      <td>rs998732</td>\n",
       "      <td>TM6SF2</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>None</td>\n",
       "      <td>19.0</td>\n",
       "      <td>19378671</td>\n",
       "      <td>A                                             ...</td>\n",
       "      <td>G                                             ...</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0.1033</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19p13.11</td>\n",
       "      <td>19267862</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6557 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      snp_id snp_ref_code    gene  \\\n",
       "0          1    rs1003531   DOT1L   \n",
       "1          2      rs10038   CRTC1   \n",
       "2          3    rs1004246    None   \n",
       "3          4   rs10048489    None   \n",
       "4          5    rs1005636    None   \n",
       "...      ...          ...     ...   \n",
       "6552    6553    rs9797885  TMEM91   \n",
       "6553    6554    rs9807841   QTRT1   \n",
       "6554    6555     rs995834    None   \n",
       "6555    6556    rs9967620    None   \n",
       "6556    6557     rs998732  TM6SF2   \n",
       "\n",
       "                                 functional_consequence clinical_significance  \\\n",
       "0                                        intron_variant                  None   \n",
       "1     3_prime_UTR_variant,genic_downstream_transcrip...                  None   \n",
       "2                                                  None                  None   \n",
       "3                                                  None                  None   \n",
       "4                                                  None                  None   \n",
       "...                                                 ...                   ...   \n",
       "6552   intron_variant,genic_upstream_transcript_variant                  None   \n",
       "6553                                     intron_variant                  None   \n",
       "6554                                               None                  None   \n",
       "6555                                               None                  None   \n",
       "6556                                     intron_variant                  None   \n",
       "\n",
       "      chromosome snp_position  \\\n",
       "0           19.0      2206575   \n",
       "1           19.0     18892728   \n",
       "2           19.0     16456260   \n",
       "3           19.0     40343129   \n",
       "4           19.0     43806559   \n",
       "...          ...          ...   \n",
       "6552        19.0     41873001   \n",
       "6553        19.0     10817656   \n",
       "6554        19.0     28866596   \n",
       "6555        19.0     46415696   \n",
       "6556        19.0     19378671   \n",
       "\n",
       "                                              reference  \\\n",
       "0     G                                             ...   \n",
       "1     C                                             ...   \n",
       "2     C                                             ...   \n",
       "3     G                                             ...   \n",
       "4     T                                             ...   \n",
       "...                                                 ...   \n",
       "6552  A                                             ...   \n",
       "6553  T                                             ...   \n",
       "6554  G                                             ...   \n",
       "6555  C                                             ...   \n",
       "6556  A                                             ...   \n",
       "\n",
       "                                            alternative info_string  ...  \\\n",
       "0     T                                             ...        None  ...   \n",
       "1     A                                             ...        None  ...   \n",
       "2     T                                             ...        None  ...   \n",
       "3     C                                             ...        None  ...   \n",
       "4     C                                             ...        None  ...   \n",
       "...                                                 ...         ...  ...   \n",
       "6552  G                                             ...        None  ...   \n",
       "6553  C                                             ...        None  ...   \n",
       "6554  A                                             ...        None  ...   \n",
       "6555  T                                             ...        None  ...   \n",
       "6556  G                                             ...        None  ...   \n",
       "\n",
       "     sasian_frequency  variation_type  consequence  mapped_gene  \\\n",
       "0              0.1626            None         None         None   \n",
       "1              0.1442            None         None         None   \n",
       "2              0.8006            None         None         None   \n",
       "3              0.0358            None         None         None   \n",
       "4              0.9110            None         None         None   \n",
       "...               ...             ...          ...          ...   \n",
       "6552           0.6207            None         None         None   \n",
       "6553           0.7239            None         None         None   \n",
       "6554           0.6350            None         None         None   \n",
       "6555           0.1646            None         None         None   \n",
       "6556           0.1033            None         None         None   \n",
       "\n",
       "      upstream_gene_id  downstream_gene_id upstream_gene_distance  \\\n",
       "0                  NaN                 NaN                    NaN   \n",
       "1                  NaN                 NaN                    NaN   \n",
       "2      ENSG00000127528     ENSG00000277825                17575.0   \n",
       "3      ENSG00000280548     ENSG00000281123                 6110.0   \n",
       "4      ENSG00000236123     ENSG00000230681                18089.0   \n",
       "...                ...                 ...                    ...   \n",
       "6552               NaN                 NaN                    NaN   \n",
       "6553               NaN                 NaN                    NaN   \n",
       "6554               NaN                 NaN                    NaN   \n",
       "6555   ENSG00000176182     ENSG00000188425                 9825.0   \n",
       "6556               NaN                 NaN                    NaN   \n",
       "\n",
       "     dowstream_gene_distance    region     chr_pos  \n",
       "0                        NaN   19p13.3     2206576  \n",
       "1                        NaN  19p13.11    18781918  \n",
       "2                     7013.0  19p13.11  16345449.0  \n",
       "3                    10834.0   19q13.2    39852489  \n",
       "4                      764.0  19q13.31    43302407  \n",
       "...                      ...       ...         ...  \n",
       "6552                     NaN   19q13.2    41367096  \n",
       "6553                     NaN   19p13.2    10706980  \n",
       "6554                     NaN     19q12    28375689  \n",
       "6555                   776.0  19q13.32    45912438  \n",
       "6556                     NaN  19p13.11    19267862  \n",
       "\n",
       "[6557 rows x 25 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select * from snp_dim\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### c. Describe the process for inserting facts data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O seguinte processo será usado para inserir dados nas tabelas de factos:\n",
    "\n",
    "#### **SNP_Person_FACT** \n",
    "- Carregaríamos todos os dataframes das dimensões associadas com a fact table (person_dimension, population_dimension, gender_dimension, chromosome_dimension, allele_dimension). \n",
    "- De seguida, usando o dataset de variação genética do 1000 genomes project, iríamos começar a ler a primeira linha, referente ao primeiro SNP no cromossoma 19. A referência do cromossoma (que neste exercício académico é sempre o cromossoma 19) e do SNP, seria usada para determinar as chromosome_key no chromosome_dimension e a SNP_key na SNP dimension.\n",
    "- De seguida, essa linha (que contém informação genética acerca de todos os indíviduos no estudo) seria analisada para identificar o primeiro indíviduo (person_code). Esse person code, seria usado para determinar o person_key da Person Dimension.\n",
    "- A informação na Person_Dimension, permitiria chegar à population_key e gender_key (na person_dimension, temos informação acerca da população, país e género).\n",
    "- Para finalizar, a informação genética do indíviduo, seria lida para determinar o genótipo de cada um dos dois alelos do indíviduo (a informação genética representa-se como 0|0, 0|1, 1|0 ou 1|1. 0 representa o alelo de referência, 1 o alelo alternativo).\n",
    "#### Este processo permitiria introduzir a primeira linha de factos. Isto seria feito iterativamente, para inserir todos os factos na tabela.\n",
    "\n",
    "#### **GWAS_Disease FACT** \n",
    "- Carregaríamos todos os dataframes das dimensões associadas com a facts table (snp_dimension, disease_dimension e study_dimension). \n",
    "- Usando o dataframe do study_dimension, definiriamos o study_key para introduzir na tabela de factos (refere um estudo realizado).\n",
    "- A informação acerca da doença (no mesmo dataframe), seria usada para encontrar a doença na disease_dimension e, consequentemente, o disease_key. \n",
    "- Ainda a partir da dataframe da study_dimension, iríamos identificar o SNP_Reference_code, que iríamos usar para identificar o snp_key, na snp_dimension.\n",
    "- Para finalizar, já com a informação atómica contida nas chaves study_key, disease_key, snp_key, iríamos adicionar o p-value presente na study_dimension, como uma medida não aditiva do facto.\n",
    "#### Este processo permitiria introduzir a primeira linha de factos. Isto seria feito iterativamente, para inserir todos os factos na tabela.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Código para inserção de factos na Disease_gwas_fact table."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Como as tabelas de SQL das dimensões são pequenas, vamos carregá-las com a informação essencial para identificar os factos.\n",
    "- De seguida, críamos queries sobre esses dataframes e produzimos uma função que produz uma nova dataframe que é usada para preencher a Disease GWAS fact table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL = \"select study_id,snps,disease,p_value from study_dim\"\n",
    "study_dim  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "SQL = \"select snp_id,snp_ref_code from snp_dim\"\n",
    "snp_dim  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "SQL = \"select disease_key,disease_name,alternative_names from disease_dimension\"\n",
    "diseases_dimension  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['study_id', 'snps', 'disease', 'p_value'], dtype='object')"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "study_dim.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    snp_id = snp_dim.loc[snp_dim['snp_ref_code']=='rs1003531']['snp_id'].values[0]\n",
    "except:\n",
    "    snp_id = None\n",
    "#df.loc[df['column_name'] == some_value]\n",
    "snp_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_query = 'byssinosis'\n",
    "\n",
    "disease_id =  diseases_dimension[diseases_dimension ['disease_name'].str.match(my_query,case=False)]['disease_key'].values[0]\n",
    "\n",
    "disease_id\n",
    "                      \n",
    "#disease_name = test3.loc[test3['disease_name']==my_query]\n",
    "#if disease_name.empty:\n",
    "#    disease_name =  test3[test3 ['alternative_names'].str.match('.*'+my_query+'.*',case=False)]\n",
    "#disease_name['disease_key'].values[0]\n",
    "                      \n",
    "#my_disease_key = disease_name['disease_key'].values[0]\n",
    "\n",
    "\n",
    "\n",
    "#print('###',my_disease_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_id</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>disease_id</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18</td>\n",
       "      <td>1066</td>\n",
       "      <td>302</td>\n",
       "      <td>5.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31</td>\n",
       "      <td>671</td>\n",
       "      <td>302</td>\n",
       "      <td>7.000000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>40</td>\n",
       "      <td>1095</td>\n",
       "      <td>918</td>\n",
       "      <td>7.000000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>102</td>\n",
       "      <td>1053</td>\n",
       "      <td>7014</td>\n",
       "      <td>3.000000e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>133</td>\n",
       "      <td>652</td>\n",
       "      <td>7160</td>\n",
       "      <td>2.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>6744</td>\n",
       "      <td>6400</td>\n",
       "      <td>302</td>\n",
       "      <td>9.000000e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>6781</td>\n",
       "      <td>5742</td>\n",
       "      <td>9273</td>\n",
       "      <td>3.000000e-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>6784</td>\n",
       "      <td>5742</td>\n",
       "      <td>9273</td>\n",
       "      <td>3.000000e-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>6785</td>\n",
       "      <td>5742</td>\n",
       "      <td>9273</td>\n",
       "      <td>2.000000e-35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>6788</td>\n",
       "      <td>5209</td>\n",
       "      <td>4388</td>\n",
       "      <td>8.000000e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>532 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     study_id  snp_id  disease_id       p_value\n",
       "0          18    1066         302  5.000000e-07\n",
       "1          31     671         302  7.000000e-09\n",
       "2          40    1095         918  7.000000e-09\n",
       "3         102    1053        7014  3.000000e-23\n",
       "4         133     652        7160  2.000000e-08\n",
       "..        ...     ...         ...           ...\n",
       "527      6744    6400         302  9.000000e-13\n",
       "528      6781    5742        9273  3.000000e-39\n",
       "529      6784    5742        9273  3.000000e-39\n",
       "530      6785    5742        9273  2.000000e-35\n",
       "531      6788    5209        4388  8.000000e-06\n",
       "\n",
       "[532 rows x 4 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_facts = []\n",
    "for i,row in study_dim.iterrows():\n",
    "    study_id,snps,disease,p_value = row[\"study_id\"],row[\"snps\"],row[\"disease\"],row[\"p_value\"]\n",
    "    #print(study_id,snps,disease,p_value)\n",
    "    try:\n",
    "        snp_id = snp_dim.loc[snp_dim['snp_ref_code']==snps]['snp_id'].values[0]\n",
    "    except:\n",
    "        snp_id = ''\n",
    "    try:\n",
    "        disease_id =  diseases_dimension[diseases_dimension ['disease_name'].str.match(disease,case=False)]['disease_key'].values[0]\n",
    "    except:\n",
    "        disease_id = ''\n",
    "    #if disease_name.empty:\n",
    "    #    disease_name =  test3[test3 ['alternative_names'].str.match('.*'+disease+'.*',case=False)]\n",
    "    if snp_id != '' and disease_id != '':\n",
    "        #print(study_id,snp_id,disease_id)\n",
    "        snp_facts.append([study_id,snp_id,disease_id,p_value])\n",
    "    \n",
    "snp_facts_df = pd.DataFrame(snp_facts,columns=['study_id','snp_id','disease_id','p_value'])\n",
    "\n",
    "snp_facts_df\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['study_id', 'snp_id', 'disease_id', 'p_value'], dtype='object')"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp_facts_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- A função seguinte, insere a dataframe na tabela de factos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "connectToServer()\n",
    "curs = conn.cursor()\n",
    "\n",
    "cols = \",\".join([str(i) for i in snp_facts_df.columns.tolist()])\n",
    "\n",
    "lines = 0\n",
    "for i,row in snp_facts_df.iterrows():\n",
    "    try:\n",
    "        sql = \"INSERT INTO disease_gwas_fact (\" + cols + \") VALUES (\" + \"%s,\"*(len(row)-1) + \"%s)\"\n",
    "        #print(sql, tuple(row))\n",
    "        curs.execute(sql, tuple(row))\n",
    "        lines += 1\n",
    "        if lines % 101:\n",
    "            conn.commit()\n",
    "        #if lines > 2:\n",
    "        #    break\n",
    "    except Exception as e:\n",
    "        print(e,tuple(row))\n",
    "\n",
    "curs.close()\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_id</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>study_id</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>302</td>\n",
       "      <td>1066</td>\n",
       "      <td>18</td>\n",
       "      <td>5.000000e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>302</td>\n",
       "      <td>671</td>\n",
       "      <td>31</td>\n",
       "      <td>7.000000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>918</td>\n",
       "      <td>1095</td>\n",
       "      <td>40</td>\n",
       "      <td>7.000000e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7014</td>\n",
       "      <td>1053</td>\n",
       "      <td>102</td>\n",
       "      <td>3.000000e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7160</td>\n",
       "      <td>652</td>\n",
       "      <td>133</td>\n",
       "      <td>2.000000e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>527</th>\n",
       "      <td>302</td>\n",
       "      <td>6400</td>\n",
       "      <td>6744</td>\n",
       "      <td>9.000000e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>528</th>\n",
       "      <td>9273</td>\n",
       "      <td>5742</td>\n",
       "      <td>6781</td>\n",
       "      <td>3.000000e-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>529</th>\n",
       "      <td>9273</td>\n",
       "      <td>5742</td>\n",
       "      <td>6784</td>\n",
       "      <td>3.000000e-39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>530</th>\n",
       "      <td>9273</td>\n",
       "      <td>5742</td>\n",
       "      <td>6785</td>\n",
       "      <td>2.000000e-35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>531</th>\n",
       "      <td>4388</td>\n",
       "      <td>5209</td>\n",
       "      <td>6788</td>\n",
       "      <td>8.000000e-06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>532 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     disease_id  snp_id  study_id       p_value\n",
       "0           302    1066        18  5.000000e-07\n",
       "1           302     671        31  7.000000e-09\n",
       "2           918    1095        40  7.000000e-09\n",
       "3          7014    1053       102  3.000000e-23\n",
       "4          7160     652       133  2.000000e-08\n",
       "..          ...     ...       ...           ...\n",
       "527         302    6400      6744  9.000000e-13\n",
       "528        9273    5742      6781  3.000000e-39\n",
       "529        9273    5742      6784  3.000000e-39\n",
       "530        9273    5742      6785  2.000000e-35\n",
       "531        4388    5209      6788  8.000000e-06\n",
       "\n",
       "[532 rows x 4 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select * from disease_gwas_fact\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Código para inserção de factos na snp_person_fact table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "conn = ''\n",
    "def connectToServer():\n",
    "    global conn\n",
    "    conn = pg.connect(host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Primeiro vamos por as tabelas de sql mais pequenas em dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>allele_key</th>\n",
       "      <th>allele_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Reference</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Alternative</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   allele_key  allele_type\n",
       "0           0    Reference\n",
       "1           1  Alternative"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select allele_key,allele_type from allele_dimension\"\n",
    "allele_dim = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "allele_dim"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Esta dimensão não tem chave substituta, vamos usar a allele_key directamente"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_id</th>\n",
       "      <th>snp_ref_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>rs1003531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>rs10038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>rs1004246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>rs10048489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>rs1005636</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   snp_id snp_ref_code\n",
       "0       1    rs1003531\n",
       "1       2      rs10038\n",
       "2       3    rs1004246\n",
       "3       4   rs10048489\n",
       "4       5    rs1005636"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select snp_id,snp_ref_code from snp_dim\"\n",
    "snp_dim = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "snp_dim.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    snp_id = snp_dim.loc[snp_dim['snp_ref_code']=='rs10048489']['snp_id'].values[0]\n",
    "except:\n",
    "    snp_id = None\n",
    "snp_id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Para os snps optámos por fazer uma chamada SQL em vez de converter em dataframe. Isto porque, quando completa, esta tabela teria ~2milhões de entradas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sql = \"select snp_id from snp_dim where snp_ref_code = %s\"\n",
    "snp_dim = getSQLfromQuery(sql, ('rs10048489',), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "try: \n",
    "    snp_id = snp_dim['snp_id'].values[0]\n",
    "except:\n",
    "    snp_id = None\n",
    "snp_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome_key</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome_key chromosome\n",
       "0               1          1\n",
       "1               2          2\n",
       "2               3          3\n",
       "3               4          4\n",
       "4               5          5"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select chromosome_key,chromosome from chromosome_dimension\"\n",
    "chromosome_dimension = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "chromosome_dimension.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "19"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    chromosome_id = chromosome_dimension.loc[chromosome_dimension['chromosome']=='19']['chromosome_key'].values[0]\n",
    "except:\n",
    "    chromosome_id = None\n",
    "chromosome_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gender_key</th>\n",
       "      <th>gender</th>\n",
       "      <th>country_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>male</td>\n",
       "      <td>BGD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>female</td>\n",
       "      <td>BGD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>male</td>\n",
       "      <td>COL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>female</td>\n",
       "      <td>COL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>male</td>\n",
       "      <td>ESP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gender_key  gender country_code\n",
       "0           1    male          BGD\n",
       "1           2  female          BGD\n",
       "2           7    male          COL\n",
       "3           8  female          COL\n",
       "4           9    male          ESP"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select gender_key,gender,country_code from gender_dimension\"\n",
    "gender_dim = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "gender_dim.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gender = 'female'\n",
    "country_code = 'COL'\n",
    "try:\n",
    "    gender_row = gender_dim.loc[(gender_dim['gender'] == gender) & (gender_dim['country_code'] == country_code)]\n",
    "    gender_id = gender_row['gender_key'].values[0]\n",
    "except:\n",
    "    gender_id = None\n",
    "    \n",
    "gender_id\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>person_key</th>\n",
       "      <th>country_code</th>\n",
       "      <th>person_code</th>\n",
       "      <th>population_code</th>\n",
       "      <th>country</th>\n",
       "      <th>gender</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GBR</td>\n",
       "      <td>HG00096</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>GBR</td>\n",
       "      <td>HG00097</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>GBR</td>\n",
       "      <td>HG00099</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>GBR</td>\n",
       "      <td>HG00100</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>GBR</td>\n",
       "      <td>HG00101</td>\n",
       "      <td>GBR</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>male</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   person_key country_code person_code population_code         country  gender\n",
       "0           1          GBR     HG00096             GBR  United Kingdom    male\n",
       "1           2          GBR     HG00097             GBR  United Kingdom  female\n",
       "2           3          GBR     HG00099             GBR  United Kingdom  female\n",
       "3           4          GBR     HG00100             GBR  United Kingdom  female\n",
       "4           5          GBR     HG00101             GBR  United Kingdom    male"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select person_key,country_code,person_code,population_code,country,gender from person_dimension\"\n",
    "person_dimension = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "person_dimension.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gender = 'female'\n",
    "country_code = 'COL'\n",
    "try:\n",
    "    gender_row = gender_dim.loc[(gender_dim['gender'] == gender) & (gender_dim['country_code'] == country_code)]\n",
    "    gender_id = gender_row['gender_key'].values[0]\n",
    "except:\n",
    "    gender_id = None\n",
    "    \n",
    "gender_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4, 'GBR', 'GBR', 'female')"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "person_code = 'HG00100'\n",
    "try:\n",
    "    person_row = person_dimension.loc[person_dimension['person_code']==person_code]\n",
    "    #print(person_row)\n",
    "    person_id = person_row['person_key'].values[0]\n",
    "    population_code = person_row['population_code'].values[0]\n",
    "    country_code = person_row['country_code'].values[0]\n",
    "    gender = person_row['gender'].values[0]\n",
    "except:\n",
    "    person_id,population_code,country_code,gender = None,None,None\n",
    "    \n",
    "person_id,population_code,country_code,gender"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population_key</th>\n",
       "      <th>population_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>CDX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>CHB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>JPT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>KHV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>CHS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   population_key population_code\n",
       "0               1             CDX\n",
       "1               2             CHB\n",
       "2               3             JPT\n",
       "3               4             KHV\n",
       "4               5             CHS"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select population_key,population_code from population_dimension\"\n",
    "population_dim = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "population_dim.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "population_code = 'JPT'\n",
    "try:\n",
    "    population_row = population_dim.loc[population_dim['population_code'] == population_code]\n",
    "    population_key = population_row['population_key'].values[0]\n",
    "except:\n",
    "    population_key = None\n",
    "    \n",
    "population_key"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Sabendo já como fazer queries sobre todos os dataframes vamos criar um programa que traduz as product_keys em ids das tabelas dimensão, para inserer na facts table.\n",
    "\n",
    "- Aqui limitámos o número de linhas que vamos processar a título de exemplo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "i = 0\n",
    "fact_snp = []\n",
    "with open(\"ALL.chr19.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf\") as f:\n",
    "    for l in f:\n",
    "        l = l.strip()\n",
    "        if not l.startswith(\"##\"):\n",
    "            if i == 0:\n",
    "                i += 1\n",
    "                continue\n",
    "            l = l.split(\"\\t\")\n",
    "            curr_snp = l[2]\n",
    "            curr_chrom = l[0]\n",
    "            l = l[9:]\n",
    "            n_entries = len(l)\n",
    "            #print(l)\n",
    "            if i == 252:\n",
    "                my_columns = l\n",
    "                #print(len(my_columns))\n",
    "                #print(my_columns)\n",
    "            if i > 252:\n",
    "                for f in range(n_entries):\n",
    "                    person_code = my_columns[f]\n",
    "                    allele1,allele2 = l[f].split('|')\n",
    "                    #print(curr_chrom,curr_snp,person_code,allele1,allele2)\n",
    "                    try:\n",
    "                        person_row = person_dimension.loc[person_dimension['person_code']==person_code]\n",
    "                        person_id = person_row['person_key'].values[0]\n",
    "                        population_code = person_row['population_code'].values[0]\n",
    "                        country_code = person_row['country_code'].values[0]\n",
    "                        gender = person_row['gender'].values[0]\n",
    "                    except:\n",
    "                        person_id,population_code,country_code,gender = None,None,None,None\n",
    "                    try:\n",
    "                        person_row = person_dimension.loc[person_dimension['person_code']==person_code]\n",
    "                        person_id = person_row['person_key'].values[0]\n",
    "                        population_code = person_row['population_code'].values[0]\n",
    "                        gender = person_row['gender'].values[0]\n",
    "                    except:\n",
    "                        person_id,population_code,gender = None,None,None\n",
    "                    try:\n",
    "                        chromosome_id = chromosome_dimension.loc[chromosome_dimension['chromosome']==curr_chrom]['chromosome_key'].values[0]\n",
    "                    except:\n",
    "                        chromosome_id = None\n",
    "                    try: \n",
    "                        sql = \"select snp_id from snp_dim where snp_ref_code = %s\"\n",
    "                        snp_dim = getSQLfromQuery(sql, (curr_snp,), host=\"localhost\",database=\"test1\", user=\"eli\", password=\"pwd\")\n",
    "                        snp_id = snp_dim['snp_id'].values[0]\n",
    "                    except:\n",
    "                        snp_id = None\n",
    "                    try:\n",
    "                        population_row = population_dim.loc[population_dim['population_code'] == population_code]\n",
    "                        population_key = population_row['population_key'].values[0]\n",
    "                    except:\n",
    "                        population_key = None\n",
    "                    try:\n",
    "                        gender_row = gender_dim.loc[(gender_dim['gender'] == gender) & (gender_dim['country_code'] == country_code)]\n",
    "                        gender_id = gender_row['gender_key'].values[0]\n",
    "                    except:\n",
    "                        gender_id = None\n",
    "                    #print(curr_chrom,curr_snp,person_code,population_code,gender,allele1,allele2)\n",
    "                    insert_me = [chromosome_id,snp_id,person_id,population_key,gender_id,allele1,allele2]\n",
    "                    #print(insert_me)\n",
    "                    fact_snp.append(insert_me)\n",
    "                #print(l)\n",
    "                #snps_alleles.append(l)\n",
    "        i += 1\n",
    "        if i > 255: #limitámos o número de linhas que vamos processar aqui\n",
    "            break\n",
    "\n",
    "snps_alleles_df = pd.DataFrame(fact_snp,columns=['chromossome_id','snp_id','person_id','population_id','gender_id','allele_1','allele_2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromossome_id</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>person_id</th>\n",
       "      <th>population_id</th>\n",
       "      <th>gender_id</th>\n",
       "      <th>allele_1</th>\n",
       "      <th>allele_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>18</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>18</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>18</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>5</td>\n",
       "      <td>18</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7507</th>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>2500</td>\n",
       "      <td>7</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7508</th>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>2501</td>\n",
       "      <td>7</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7509</th>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>2502</td>\n",
       "      <td>7</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7510</th>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>2503</td>\n",
       "      <td>7</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7511</th>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>2504</td>\n",
       "      <td>7</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7512 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromossome_id snp_id  person_id  population_id  gender_id allele_1  \\\n",
       "0                 19   None          1             18         13        0   \n",
       "1                 19   None          2             18         14        0   \n",
       "2                 19   None          3             18         14        0   \n",
       "3                 19   None          4             18         14        0   \n",
       "4                 19   None          5             18         13        0   \n",
       "...              ...    ...        ...            ...        ...      ...   \n",
       "7507              19   None       2500              7         34        0   \n",
       "7508              19   None       2501              7         34        0   \n",
       "7509              19   None       2502              7         34        0   \n",
       "7510              19   None       2503              7         34        0   \n",
       "7511              19   None       2504              7         34        0   \n",
       "\n",
       "     allele_2  \n",
       "0           0  \n",
       "1           0  \n",
       "2           0  \n",
       "3           0  \n",
       "4           0  \n",
       "...       ...  \n",
       "7507        0  \n",
       "7508        0  \n",
       "7509        0  \n",
       "7510        0  \n",
       "7511        0  \n",
       "\n",
       "[7512 rows x 7 columns]"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snps_alleles_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Finalmente inserimos o dataframe da tabela de factos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "connectToServer()\n",
    "curs = conn.cursor()\n",
    "\n",
    "cols = \",\".join([str(i) for i in snps_alleles_df.columns.tolist()])\n",
    "\n",
    "lines = 0\n",
    "for i,row in snps_alleles_df.iterrows():\n",
    "    try:\n",
    "        sql = \"INSERT INTO snp_person_fact (\" + cols + \") VALUES (\" + \"%s,\"*(len(row)-1) + \"%s)\"\n",
    "        #print(sql, tuple(row))\n",
    "        curs.execute(sql, tuple(row))\n",
    "        lines += 1\n",
    "        if lines % 101:\n",
    "            conn.commit()\n",
    "        #if lines > 2:\n",
    "        #    break\n",
    "    except Exception as e:\n",
    "        print(e,tuple(row))\n",
    "\n",
    "curs.close()\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gender_id</th>\n",
       "      <th>population_id</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>person_id</th>\n",
       "      <th>chromossome_id</th>\n",
       "      <th>allele_1</th>\n",
       "      <th>allele_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>None</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>14</td>\n",
       "      <td>18</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14</td>\n",
       "      <td>18</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>18</td>\n",
       "      <td>None</td>\n",
       "      <td>4</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>None</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7507</th>\n",
       "      <td>34</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "      <td>2500</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7508</th>\n",
       "      <td>34</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "      <td>2501</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7509</th>\n",
       "      <td>34</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "      <td>2502</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7510</th>\n",
       "      <td>34</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "      <td>2503</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7511</th>\n",
       "      <td>34</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "      <td>2504</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7512 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gender_id  population_id snp_id  person_id  chromossome_id  allele_1  \\\n",
       "0            13             18   None          1              19         0   \n",
       "1            14             18   None          2              19         0   \n",
       "2            14             18   None          3              19         0   \n",
       "3            14             18   None          4              19         0   \n",
       "4            13             18   None          5              19         0   \n",
       "...         ...            ...    ...        ...             ...       ...   \n",
       "7507         34              7   None       2500              19         0   \n",
       "7508         34              7   None       2501              19         0   \n",
       "7509         34              7   None       2502              19         0   \n",
       "7510         34              7   None       2503              19         0   \n",
       "7511         34              7   None       2504              19         0   \n",
       "\n",
       "      allele_2  \n",
       "0            0  \n",
       "1            0  \n",
       "2            0  \n",
       "3            0  \n",
       "4            0  \n",
       "...        ...  \n",
       "7507         0  \n",
       "7508         0  \n",
       "7509         0  \n",
       "7510         0  \n",
       "7511         0  \n",
       "\n",
       "[7512 rows x 7 columns]"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"select * from snp_person_fact\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Do a critical assessment of the work\n",
    "### a. Describe potential issues with the ETL procedure used\n",
    "\n",
    "O processo de ETL foi um desafio árduo, visto que não foi possível encontrar um único repositório com todos os datasets/dados necessários. Projectos genómicos desta dimensão e âmbito, necessitam de aglomerar dados de várias fontes diferentes. Descarregar, organizar, filtrar, compatibilizar estes dados é um desafio enorme. Neste sentido, recorremos a dados de fontes como, NCBI (informação acerca de SNPs e cromossomas), 1000 genomes project, The World Data Bank, GWAS Catalogue e Wikidata. Tendo em conta a dimensão dos dados a agregar e a necessidade de os normalizar e compatibilizar, é fácil entender a complexidade deste exercício.\n",
    "\n",
    "Os problemas que encontramos na obtenção de datasets adequados, podem ser ilustrados com o desafio em obter dados agregados acerca de doenças humanas. Na realidade, na nossa busca, o único repositório que encontramos que agrega informação acerca de doenças humanas é o site https://www.malacards.org/. Este site aglomera dados de 74 (!!) fontes distintas para obter informação rica acerca de doenças. Infelizmente, não há maneira de aceder aos datasets deste site. Chegamos inclusivamente a contactar o Help Desk, mais foi-nos dito que para obter os dados precisávamos de um acordo institucional. Esta pequena estória, exemplifica bem a importância de usar várias fontes para enriquecer os dados. Por isso mesmo, a maioria das nossas tabelas de dimensões resultam da fusão de dados de datasets diferentes (por exemplo, os dados para a tabela population Dimension, vieram de dados do dataset 1000 genomes e de datasets do World Data Bank). \n",
    "\n",
    "Voltando ao desafio com a dimensão Disesases (um dos grandes desafios, que nos levou, sem exagero, meses!), não se encontrou nenhum dataset que contivesse os dados necessários. Tentamos recorrer a ontologias como Doid e Human Diseases Ontologies, mas mesmo assim os dados eram \"pobre\". Após muita procura, acabamos por utilizar o Wikidata, uma solução \"knowledge base\", onde se encontraram todos os dados necessário para a dimensão. Para extrair e construir o dataset \"caseiro\" recorreu-se à ferramenta Wikidata Query Service na qual se utiliza a linguagem SPARQL para o desenvolvimento das queries, que nos permitiram obter os dados para carregar na nossa dimensão de doenças.\n",
    "\n",
    "**A complexidade e dificuldade que foi o processo de \"data acquisition\", transpirou para o processo de ETL. De facto, vários problemas emergiram durante este processo de \"escultura\" e compatibilização de dados de diferentes fontes.**\n",
    "Heis alguns dos problemas com os quais nos deparamos (alguns dos quais não foram completamente resolvidos):\n",
    "- Redundância entre os dados de diferentes fontes, que tentamos normalizar o melhor que conseguimos.\n",
    "- Problemas de ter campos em datasets com entradas múltiplas (a solução actual, foi transformar estes campos em arrays).\n",
    "- Problemas de modelação das Study and Disease Dimension, nomeadamente no que diz respeito aos nomes das doenças. Muitas doenças têm nomes alternativos, pelo que optamos por incluir na Disease Dimension, um campo com \"Alternative_Names\". Infelizmente, este campo não foi ainda modelado da melhor forma e é uma solução a explorar no futuro.\n",
    "- Problemas com indíces demográficos em falta para as dimensões Gender e Population. Preencher estes dados necessitaria de mais tempo e mais fontes de dados demográficos. De qualquer modo, este é um aspecto a considerar em versões futuras do DW, que serão guiadas pelos vários campos de \"negócio\", por forma a enriquecer de maneira substancial os dados demográficos.\n",
    "- Na modelação dos dados demográficos, tivemos que decidir como modelar estes dados no tempo. Dado a natureza do DW actual, decidimos extrair os dados demográficos relativos ao ano em que o 1000 genomes project foi publicado, 2015. No futuro, poderemos melhorar esta solução, incluindo mais anos, médias das últimas décadas, adicionar indíces dos últimos anos disponíveis, etc.). Estes dados demográficos, se crescerem significativamente, podem inclusive ser modelados de maneira diferente, em outtriggers, em dimensões slow changing (SCD), etc...\n",
    "\n",
    "**Dadas as dimensões como GWAS e SNP Dimension e a dificuldade em modular estas dimensões tomamos várias decisões estratégicas que se justificam pelo teor académico deste exercício:**\n",
    "- Focamos a nossa análise genética só no cromossoma 19 (e que já assim são cerca de 1.8 milhões de linhas de SNPs).\n",
    "- Existem vários tipos de variações genéticas (SNPs, CNVs, INDELs, etc.). Para simplificar a nossa análises usamos só dados de SNPs com um único alelo alternativo.\n",
    "- No study Dimension usamos apenas os estudos de associação que identificam um único SNP no cromossoma 19 associado a uma determinada doença.\n",
    "- Dada a dimensão da SNP dimension, optámos por modelar os nossos dados para inserir apenas a informação relativa a SNPs que foram reportados no GWAS catalogue, como tendo uma associação genética a uma doença (e que de facto constituem os dados que vão ser introduzidos na tabela de factos). \n",
    "- Na Study Dimension, estão incluídos estudos acerca de associação genética com doenças, mas também com outros traços e características humanas. Isto cria uma dificuldade acrescida, na selecção de dados de associação relacionados apenas com doenças.\n",
    "- Devido à complexidade na modelação destas duas dimensões, resta ainda muito trabalho para fazer na optimização dos dados respectivos.\n",
    "\n",
    "Devido a uma enorme esforço de coordenação, conseguiu-se solucionar a maioria dos problemas verificados durante o processo de ETL. Porém, a fiabilidade dos dados poderá não ser ainda a melhor (dados desatualizados e/ou com falhas). Próximas versões deste DW deverão concentrar-se em limpar e harmonizar os dados.\n",
    "\n",
    "De salientar que houve uma decisão de modelação de última hora: por forma a lidar com os campos Allele_1 e Allele_2 na SNP_Person_Fact, criou-se uma Allele_Dimension, cujo único propósito é identificar se o alelo é o a referência, ou o alternativo. Achamos que esta \"Role-Playing Dimension\" irá ajudar na inteligibilidade da leitura do esquema, pelos end-users.\n",
    "\n",
    "Por fim, acreditamos que os erros a evitar em modelação dimensional foram evitados.\n",
    "\n",
    "#### Nota: \n",
    "- Um erro de última hora foi detectado quando tentámos inserir os factos na Disease GWAS fact. Detectámos uma incompatibilidade na nomenclatura usada para descrever as doenças nos dados obtidos do GWAS dataset e da Wikidata. Para resolver este problema, fomos forçados a recorrer à comparação por expressões regulares. Isto não é ideal mas permite-nos, para já, evitar falsos positivos. Os dados serão, no futuro, tratados para resolver este tipo de problemas.\n",
    "- Nesta versão do Notebook, e após uma noite inteira a tentar resolver problemas de incompatibilidade com os dados, conseguimos criar e injectar a tabela de factos SNP Person Fact.\n",
    "- O código para criação da Study Dimension e SNP Dimension está ainda muito complexo e necessita ser simplificado e melhorado. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. Compare your schema to the one previously defined in phase I\n",
    "\n",
    "O formato essencial do nosso esquema de dados, não mudou dramaticamente em comparação com aquele proposto na fase 1 do projecto. As mudanças operadas, resultaram da consolidação dos dados, do refinamento da informação guardada em cada dimensão, da conformação de dimensões, e da tentativa de melhor o DW. \n",
    "\n",
    "Heis alguns dos exemplos de inovação operados nesta fase do projecto:\n",
    "- Vários dados demográficos foram adicionados às Gender e Population Dimensions. Este exercício académico, pretende ser apenas exemplificativo. Muitos mais dados e, uma muito maior riqueza do esquema, poderia ser facilmente atingida modelando uma maior quantidade de dados no nosso DW.\n",
    "- Em quase todas as populações, aumentamos o número de colunas, por forma a facilitar a leitura e inteligibilidade do dados. Preocupamo-nos também em adicionar atributos de uma forma conformada, por forma a permitir a ligação entre dimensões no futuro (por exemplo, em exercícios de \"drill-across\" através de queries em diferentes tabelas de factos).\n",
    "- Adicionamos vários parámetros métricos à Chromosome Dimension, e parámetros pré-calculados, como densidade de genes por megabase. \n",
    "- Adicionaram-se vários atributos à dimensão SNP, incluindo contexto genómico, consequências funcionais, genes associados.\n",
    "- A Disease Dimension foi completamente renovada, com novos atributos derivados dos dados que obtemos.\n",
    "- A Study Dimension agrega mais informação relativamente aos genes associados com os SNPs identificados no estudo e valores probabilisticos da associação.\n",
    "- A Disesase GWAS Fact passou a ter uma medida não aditiva, o p-value derivado do estudo de associação genética. Este valor permite medir, para o evento atómico, a probabilidade de um  SNP estar associado com uma doença.\n",
    "- A SNP Person Fact, passou a ser uma factless table. Esta decisão derivou da necessidade de modelar a informação genética relativamente aos alelos de cada indíviduo (para cada SNP). A informação que tínhamos planeado colocar em cada linha de factos era, para cada um dos dois alelos, o valor 0 representando um alelo de referência, ou o valor 1 representando um alelo alternativo. Como estas medidas seriam pouco informativas e difíceis de entender, críamos uma Allele_Dimension, onde a informação acerca do tipo de alelo (Reference ou Alternative) pode ser facilmente obtida."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### c. Discuss the issues for updating the data warehouse with novel data\n",
    "\n",
    "O nosso Data Warehouse é relativamente limitado em termos de updates no tempo. Além do mais, não existem hierarquias nas dimensões, pelo que a modelação no tempo acaba por ser facilitada. Em termos de crescimento, este acontecerá apenas quando novos estudos GWAS forem publicados, ou novos dados genéticos estiverem disponíveis. Neste sentido, e no ponto de vista do exercício académico, estas alterações seriam periódicas e consistiriam de adicionar novas linhas às tabelas de dimensões respectivas (com consequente aumento dos factos, para cada estudo e para cada indíviduos cujo genoma fosse sequenciado). Por isso, a grande preocupação será a de garantir que novos dados são adicionados na ordem correcta (e que dados antigos não são reintroduzidos).\n",
    "\n",
    "No que diz respeito a Slow Changing Dimensions (SCDs), as únicas SCD que podemos considerar são aquelas que agregam informação demográfica. Nos actuais esquemas de negócio, não existe interesse em guardar historial dos dados, pelo que estas dimensões comportar-se-iam com SCD do tipo 1, simplesmente substituíndo os indíces demográficos por valores actualizados. Se existisse a necessidade de se referir a dados demográficos históricos, poder-se-ia considerar o uso de SCDs de tipo IV, onde se usaria uma \"mini-dimension\" com o histórico dos dados (de 5 em 5 anos) e que não obrigaria a grandes alterações ao esquema actual."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
