{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HuGS-DW: Human GWAS-SNP Cross-Reference Data Warehouse\n",
    "# Tecnologia de Processamento de dados 2019-2020 - Phase 3\n",
    "\n",
    "#### Elias Barreira 40821, José Matos 49652, Miguel Casanova 24475, Telmo Silva 54013\n",
    "#### Horas de contribuição: 15h por elemento"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. OLAP queries\n",
    "### a. Check and revise the queries defined in Phase I. Ensure that the queries are meaningful for the processes at hand"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As queries que propusemos na fase 1 do projecto, debruçam-se sobre questões relacionadas com filogenias e variabilidade genética humana. Mais precisamente, este data warehouse vai permitir uma abordagem analítica acerca de como variantes genéticas estão distribuídas de forma distinta entre diferentes populações humanas, ou mesmo, prever a predisposição de certas populações para determinadas doenças.\n",
    "\n",
    "Este data warehouse pretende responder às necessidades de vários \"processos de negócio\", tais como:\n",
    "- **Laboratórios de investigação** - servindo como uma ferramenta para análise da Variabilidade Genética Populacional;\n",
    "- **Departamentos médicos de doenças genéticas** - providenciando uma ferramenta que permite estudar a associação entre doenças a variações genéticas;\n",
    "- **Seguradoras e outras instituições de análise de background e risco** - este data warehouse permitirá a previsão da prevalência de certas variantes genéticas de risco, em diferentes populações humanas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As queries propostas inicialmente foram:\n",
    "\n",
    "1. Que populações têm uma maior frequência de variantes genéticas em genes vs regiões intergénicas e/ou reguladoras?\n",
    "2.  Existe uma relação entre variabilidade genética e aspectos socioeconómicos e culturais das populações? (maior taxas de mutações, maior isolamento populacional?)\n",
    "3. Em que populações uma determinada doença complexa tem maior probabilidade de ocorrer?\n",
    "4. As variantes genéticas identificadas estão em que contexto genómico? Existe uma maior associação de doenças com genes ou com regiões não codificantes?\n",
    "5. Em populações com igual frequência de alelos de risco para uma determinada doença, qual o impacto que aspectos socioeconómicos têm sobre a incidência da doença?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Na sua maioria, as queries formuladas inicialmente são muito relevantes e adequadas aos vários processos de negócio. Requerem porém alguma revisão, que se prende com os seguintes factos:\n",
    "\n",
    "**- Como temos muitos dados, optámos por compromissos:**\n",
    "- Inserimos apenas estudos em que se refere uma associação entre uma doença e um único SNP.\n",
    "- Inserimos apenas os SNPs que aparecem associados aos estudos (e consequentemente, a uma doença).\n",
    "- Reduzimos assim de 2 milhões de SNPs para alguns milhares.\n",
    "- Para cada pessoa inserimos a informação sobre os SNPs referidos acima.\n",
    "- Ao inserirmos alguns factos na SNP_Person_Fact, o snp_id virá necessariamente vazio (não foi criado o SNP).\n",
    "- Inserimos dummy data quando necessário para poder executar as queries obtendo resultados.\n",
    "\n",
    "**- A mudança da estrutura do DW, levou-nos a adaptar as nossas queries.**\n",
    "\n",
    "**- Dificuldade na obtenção de informação acerca de doenças, nomeadamente na sua incidência em diferentes populações, mortalidade, etc. Isto levou-nos a deixar algumas queries de lado, tal como a query 5.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mesmo tendo em conta estes precalços, as queries mantiveram-se mais ou menos as mesmas, sem grande necessidade de alterações. Salienta-se apenas as seguintes mudanças:\n",
    "1. A query 1 foi ligeiramente adaptada, para se focar no cálculdo das frequências de variantes genéticas em diferentes Populações, Países, géneros, etc. Esta mudança deveu-se essencialmente ao facto de termos julgado que a query original se sobrepunha à query 4.\n",
    "2. A query 5 tornou-se inviável, essencialmente porque não obtivemos dados relativamente à incidência e mortalidade das doenças, por país e/ou população.\n",
    "\n",
    "As restantes queries, por manterem a sua viabilidade e interesse para os processos de negócio definidos, mantiveram-se na mesma."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. For each query type identify dimensions, attributes and measures for each one "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 1 - Quais as frequências de variantes genéticas em diferentes populações, Países, géneros?\n",
    "Este tipo de queries vai orbitar à volta da tabela de factos snp_person_fact. Vai usar essencialmente dados de:\n",
    "- **Dimensões:** SNP_dimension, Population_Dimension, Person_Dimension, Gender_Dimension\n",
    "- **Atributos:** Usaremos vários atributos, nomeadamente, snp_key, snp_ref_code, context, gender_key, gender, population_key, population_description, super_population, country\n",
    "- **Medidas:** A snp_person_fact table é uma factless table. Porém, usamos a informação das colunas allele_1 e allele_2, da snp_person_fact, para conseguir calcular a frequência de alelos de referência e alternativos."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 2 - Existe uma relação entre variabilidade genética e aspectos socioeconómicos e culturais das populações? \n",
    "\n",
    "Estas queries vão centrar-se em torno da mesma estrutura de dados das queries anteriores, porém focando-se em aspectos demográficos vários. Assim, vai usar essencialmente:\n",
    "- **Dimensões:** SNP_dimension, Population_Dimension, Person_Dimension, Gender_Dimension\n",
    "- **Atributos:** Usaremos vários atributos, nomeadamente, snp_key, snp_ref_code, context, gender_key, gender, population_key, population_description, super_population, country e, principlamente, atributos demográficos como mortality_per_1000_habitants, gross_national_income e gross_national_income_group\n",
    "- **Medidas:** A snp_person_fact table é uma factless table. Porém, usamos a informação das colunas allele_1 e allele_2, da snp_person_fact, para conseguir calcular a frequência de alelos de referência e alternativos."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 3 - Em que populações uma determinada doença complexa tem maior probabilidade de ocorrer?\n",
    "\n",
    "Este conjunto de queries marca os primeiros exercícios de drill-across diferentes tabelas de factos. Vai usar informação acerca da associação genética de certas variantes de risco com determinadas doenças (extraída da disease_gwas_fact table), para procurar informação acerca da variabilidade dessas variantes em diferentes populações (a partir da snp_person_fact table).\n",
    "\n",
    "- **Dimensões:** Person_Dimension, Disease_Dimension, Population_Dimension\n",
    "- **Atributos:** Os atributos usados serão, super_population, population_description, person_key, snp_id, disease_id, disease_name\n",
    "- **Medidas:** Como nas queries anteriores, não temos medidadas aditivas associadas aos factos e que sejam relevantes para esta query. Porém, mais uma vez, a informação nas colunas allele_1 e allele_2, da snp_person_fact, vão permitir calcular a frequência de alelos de referência e alternativos em determinadas populações."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 4 - As variantes genéticas identificadas estão em que contexto genómico? Existe uma maior associação de doenças com genes ou com regiões não codificantes?\n",
    "\n",
    "Estas queries centram-se em torno das doenças e de SNPs de risco identificados em vários estudos (os factos da disease_gwas_fact table). Assim, usa-se essencialmente a informação contida em:\n",
    "- **Dimensões:** Disease_Dimension, SNP_Dimension\n",
    "- **Atributos:** Os atributos usados nestas queries são snp_ref_code, context e disease_name\n",
    "- **Medidas:** Não se usam as medidas da tabela disease_gwas_fact. Usa-se o context (da SNP_Dimension), para agregar dados por tipo de localização genómica onde se encontrar variantes genéticas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 5 - Em populações com igual frequência de alelos de risco para uma determinada doença, qual o impacto que aspectos socioeconómicos têm sobre a incidência da doença?\n",
    "\n",
    "Estas queries fariam o drill-across entre as tabelas de factos Disease_GWAS_fact e SNP_Person_fact, para explorar como os aspectos ambientais estão relacionados com a gravidade da doença em certas populações. Por não termos informações acerca da incidência das doenças em diferentes Países e/ou populações, estas queries não são possíveis. Se fossem possíveis, estas queries iriam necessitar dos seguintes dados:\n",
    "- **Dimensões:** Disease_Dimension, Population_Dimension\n",
    "- **Atributos:** Os atributos que seriam usados seriam population_key, snp_key, disease_key, population_description, life_expectancy_birth, gross_national_income. Para além destes, necessitariamos de dados acerca da incidência das doenças em grupos populacionais.\n",
    "- **Medidas:** Estes queries não usariam medidas das tabelas de factos. Usariam a informação acerca da população, para calcular a frequência que as variantes genéticas de risco têm nestas populações. Poderíamos considerar usar a medida semi-aditiva da tabela Disease_GWAS_fact, para filtrar as associações genéticas identificadas entre SNPs e doenças, com maior probabilidade."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### c. Write the SQL queries and execute them. Make sure the results are understandable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Comecemos por carregar as funcões que nos vão ajudar com os comandos SQL."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import psycopg2 as pg\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def pandafy(rows, colnames):\n",
    "    N=len(colnames)\n",
    "    D={cn: [] for cn in colnames}\n",
    "    for row in rows:\n",
    "        for i in range(N): D[colnames[i]].append(row[i])\n",
    "    pdfy=pd.DataFrame(D)\n",
    "    return pdfy\n",
    "\n",
    "def excuteSingleSQLstatement(sql, host, database, user, password):\n",
    "    conn = pg.connect(host=host,database=database, user=user, password=password)\n",
    "    cur = conn.cursor()\n",
    "    cur.execute(sql)\n",
    "    cur.close()\n",
    "    conn.commit()\n",
    "    conn.close()    \n",
    "\n",
    "def getSQLfromQuery(sql, params, host, database, user, password):\n",
    "    conn = pg.connect(host=host, database=database, user=user, password=password)\n",
    "    cur  = conn.cursor()    \n",
    "    if len(params)==0:\n",
    "        cur.execute(sql)\n",
    "    else: \n",
    "        cur.execute(sql, params)\n",
    "    data = cur.fetchall()\n",
    "    colnames = [desc[0] for desc in cur.description]\n",
    "    cur.close()\n",
    "    conn.close()\n",
    "    df=pandafy(data, colnames)\n",
    "    return df\n",
    "\n",
    "def excuteManySQLstatements(sql_list, host, database, user, password):\n",
    "    conn = pg.connect(host=host,database=database, user=user, password=password)\n",
    "    cur = conn.cursor()\n",
    "    for sql in sql_list: cur.execute(sql)\n",
    "    cur.close()\n",
    "    conn.commit()\n",
    "    conn.close()   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 1 - Quais as frequências de variantes genéticas em diferentes populações, Países, géneros?\n",
    "Elaboramos aqui um conjunto de queries que visam fazer a análise da frequência alélica dos vários SNPs, em determinados contextos (populacionais, países, géneros)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **A query seguinte, vai calcular a percentagem de alelos de referência e alternativos nas várias sub-populações.**\n",
    "- Esta query permite-nos ter uma noção da variabilidade genética encontrada em várias populações."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population_description</th>\n",
       "      <th>reference_frequency</th>\n",
       "      <th>alternative_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Luhya in Webuye, Kenya</td>\n",
       "      <td>0.986532</td>\n",
       "      <td>0.013468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>African Caribbean in Barbados</td>\n",
       "      <td>0.989583</td>\n",
       "      <td>0.010417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mende in Sierra Leone</td>\n",
       "      <td>0.990196</td>\n",
       "      <td>0.009804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Yoruba in Ibadan, Nigeria</td>\n",
       "      <td>0.990741</td>\n",
       "      <td>0.009259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>0.990975</td>\n",
       "      <td>0.009025</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Esan in Nigeria</td>\n",
       "      <td>0.991582</td>\n",
       "      <td>0.008418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>African Ancestry in Southwest US</td>\n",
       "      <td>0.994536</td>\n",
       "      <td>0.005464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Gambian in Western Division, The Gambia</td>\n",
       "      <td>0.995575</td>\n",
       "      <td>0.004425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Toscani in Italy</td>\n",
       "      <td>0.996885</td>\n",
       "      <td>0.003115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Colombian in Medellin, Colombia</td>\n",
       "      <td>0.998227</td>\n",
       "      <td>0.001773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Utah residents with Northern and Western Europ...</td>\n",
       "      <td>0.998316</td>\n",
       "      <td>0.001684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Indian Telugu in the UK</td>\n",
       "      <td>0.998366</td>\n",
       "      <td>0.001634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Puerto Rican in Puerto Rico</td>\n",
       "      <td>0.998397</td>\n",
       "      <td>0.001603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Bengali in Bangladesh</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Japanese in Tokyo, Japan</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Punjabi in Lahore,Pakistan</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Southern Han Chinese, China</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Han Chinese in Beijing, China</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Iberian populations in Spain</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Gujarati Indian in Houston,TX</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Finnish in Finland</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Mexican Ancestry in Los Angeles, California</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Peruvian in Lima, Peru</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Sri Lankan Tamil in the UK</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Chinese Dai in Xishuangbanna, China</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Kinh in Ho Chi Minh City, Vietnam</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               population_description  reference_frequency  \\\n",
       "0                              Luhya in Webuye, Kenya             0.986532   \n",
       "1                       African Caribbean in Barbados             0.989583   \n",
       "2                               Mende in Sierra Leone             0.990196   \n",
       "3                           Yoruba in Ibadan, Nigeria             0.990741   \n",
       "4                     British in England and Scotland             0.990975   \n",
       "5                                     Esan in Nigeria             0.991582   \n",
       "6                    African Ancestry in Southwest US             0.994536   \n",
       "7             Gambian in Western Division, The Gambia             0.995575   \n",
       "8                                    Toscani in Italy             0.996885   \n",
       "9                     Colombian in Medellin, Colombia             0.998227   \n",
       "10  Utah residents with Northern and Western Europ...             0.998316   \n",
       "11                            Indian Telugu in the UK             0.998366   \n",
       "12                        Puerto Rican in Puerto Rico             0.998397   \n",
       "13                              Bengali in Bangladesh             1.000000   \n",
       "14                           Japanese in Tokyo, Japan             1.000000   \n",
       "15                         Punjabi in Lahore,Pakistan             1.000000   \n",
       "16                        Southern Han Chinese, China             1.000000   \n",
       "17                      Han Chinese in Beijing, China             1.000000   \n",
       "18                       Iberian populations in Spain             1.000000   \n",
       "19                      Gujarati Indian in Houston,TX             1.000000   \n",
       "20                                 Finnish in Finland             1.000000   \n",
       "21        Mexican Ancestry in Los Angeles, California             1.000000   \n",
       "22                             Peruvian in Lima, Peru             1.000000   \n",
       "23                         Sri Lankan Tamil in the UK             1.000000   \n",
       "24                Chinese Dai in Xishuangbanna, China             1.000000   \n",
       "25                  Kinh in Ho Chi Minh City, Vietnam             1.000000   \n",
       "\n",
       "    alternative_frequency  \n",
       "0                0.013468  \n",
       "1                0.010417  \n",
       "2                0.009804  \n",
       "3                0.009259  \n",
       "4                0.009025  \n",
       "5                0.008418  \n",
       "6                0.005464  \n",
       "7                0.004425  \n",
       "8                0.003115  \n",
       "9                0.001773  \n",
       "10               0.001684  \n",
       "11               0.001634  \n",
       "12               0.001603  \n",
       "13               0.000000  \n",
       "14               0.000000  \n",
       "15               0.000000  \n",
       "16               0.000000  \n",
       "17               0.000000  \n",
       "18               0.000000  \n",
       "19               0.000000  \n",
       "20               0.000000  \n",
       "21               0.000000  \n",
       "22               0.000000  \n",
       "23               0.000000  \n",
       "24               0.000000  \n",
       "25               0.000000  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "SELECT population_description, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "        (cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, population_dimension P\n",
    "WHERE P.population_key = SP.population_id\n",
    "GROUP BY population_description\n",
    "ORDER BY alternative_frequency DESC\n",
    "\"\"\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **A query seguinte, vai calcular, por SNP e para cada Super-população e sub-população, as frequências do alelo de referência e do alelo alternativo.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# As we mentioned above, we had to introduce some changes in the snp_person_fact table.\n",
    "# As the snp_id was missing (for the reasons already explained above), we manually added the snp_id.\n",
    "SQL = \"\"\"\n",
    "    alter table snp_person_fact add id serial;\n",
    "    update snp_person_fact\n",
    "    set snp_id = 1\n",
    "    where id <= 2504; \n",
    "    update snp_person_fact\n",
    "    set snp_id = 2\n",
    "    where id > 2504 and id <= 5008;\n",
    "    update snp_person_fact\n",
    "    set snp_id = 3\n",
    "    where id > 5008 and id <= 7512;\n",
    "    ALTER TABLE snp_person_fact \n",
    "    DROP COLUMN id;\n",
    "    \"\"\"\n",
    "#excuteSingleSQLstatement(SQL,host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gender_id</th>\n",
       "      <th>population_id</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>person_id</th>\n",
       "      <th>chromossome_id</th>\n",
       "      <th>allele_1</th>\n",
       "      <th>allele_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>1053</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>1053</td>\n",
       "      <td>2502</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>1053</td>\n",
       "      <td>2503</td>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>1053</td>\n",
       "      <td>2504</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>4179</td>\n",
       "      <td>56</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7515</th>\n",
       "      <td>34</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>2500</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7516</th>\n",
       "      <td>34</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>2501</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7517</th>\n",
       "      <td>34</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>2502</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7518</th>\n",
       "      <td>34</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>2503</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7519</th>\n",
       "      <td>34</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>2504</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7520 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      gender_id  population_id  snp_id  person_id  chromossome_id  allele_1  \\\n",
       "0            13             18    1053          1              19         1   \n",
       "1            13             18    1053       2502              19         1   \n",
       "2            13             18    1053       2503              19         1   \n",
       "3            13             18    1053       2504              19         0   \n",
       "4            13             19    4179         56              19         0   \n",
       "...         ...            ...     ...        ...             ...       ...   \n",
       "7515         34              7       3       2500              19         0   \n",
       "7516         34              7       3       2501              19         0   \n",
       "7517         34              7       3       2502              19         0   \n",
       "7518         34              7       3       2503              19         0   \n",
       "7519         34              7       3       2504              19         0   \n",
       "\n",
       "      allele_2  \n",
       "0            0  \n",
       "1            1  \n",
       "2            0  \n",
       "3            1  \n",
       "4            0  \n",
       "...        ...  \n",
       "7515         0  \n",
       "7516         0  \n",
       "7517         0  \n",
       "7518         0  \n",
       "7519         0  \n",
       "\n",
       "[7520 rows x 7 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "SELECT *\n",
    "FROM snp_person_fact\n",
    "\"\"\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>super_population</th>\n",
       "      <th>reference_frequency</th>\n",
       "      <th>alternative_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>African</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>American</td>\n",
       "      <td>0.998559</td>\n",
       "      <td>0.001441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>European</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>0.998978</td>\n",
       "      <td>0.001022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>African</td>\n",
       "      <td>0.982602</td>\n",
       "      <td>0.017398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>American</td>\n",
       "      <td>0.998559</td>\n",
       "      <td>0.001441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>European</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>African</td>\n",
       "      <td>0.990923</td>\n",
       "      <td>0.009077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>American</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>East Asian</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>European</td>\n",
       "      <td>0.997018</td>\n",
       "      <td>0.002982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rs117710327</td>\n",
       "      <td>European</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rs4808075</td>\n",
       "      <td>European</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>rs4808075</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   snp_ref_code super_population  reference_frequency  alternative_frequency\n",
       "0     rs1003531          African             1.000000               0.000000\n",
       "1     rs1003531         American             0.998559               0.001441\n",
       "2     rs1003531       East Asian             1.000000               0.000000\n",
       "3     rs1003531         European             1.000000               0.000000\n",
       "4     rs1003531      South Asian             0.998978               0.001022\n",
       "5       rs10038          African             0.982602               0.017398\n",
       "6       rs10038         American             0.998559               0.001441\n",
       "7       rs10038       East Asian             1.000000               0.000000\n",
       "8       rs10038         European             1.000000               0.000000\n",
       "9       rs10038      South Asian             1.000000               0.000000\n",
       "10    rs1004246          African             0.990923               0.009077\n",
       "11    rs1004246         American             1.000000               0.000000\n",
       "12    rs1004246       East Asian             1.000000               0.000000\n",
       "13    rs1004246         European             0.997018               0.002982\n",
       "14    rs1004246      South Asian             1.000000               0.000000\n",
       "15  rs117710327         European             0.375000               0.625000\n",
       "16    rs4808075         European             1.000000               0.000000\n",
       "17    rs4808075      South Asian             1.000000               0.000000"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "SELECT snp_ref_code, super_population, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, population_dimension P, snp_dim S\n",
    "WHERE P.population_key = SP.population_id AND S.snp_id=SP.snp_id\n",
    "GROUP BY super_population, snp_ref_code\n",
    "ORDER BY snp_ref_code\n",
    "\"\"\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Vamos organizar um pouco a maneira como visualizamos o resultado da query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>reference_frequency</th>\n",
       "      <th>alternative_frequency</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>super_population</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">rs1003531</th>\n",
       "      <th>African</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>American</th>\n",
       "      <td>0.998559</td>\n",
       "      <td>0.001441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>East Asian</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>European</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>South Asian</th>\n",
       "      <td>0.998978</td>\n",
       "      <td>0.001022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">rs10038</th>\n",
       "      <th>African</th>\n",
       "      <td>0.982602</td>\n",
       "      <td>0.017398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>American</th>\n",
       "      <td>0.998559</td>\n",
       "      <td>0.001441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>East Asian</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>European</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>South Asian</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">rs1004246</th>\n",
       "      <th>African</th>\n",
       "      <td>0.990923</td>\n",
       "      <td>0.009077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>American</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>East Asian</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>European</th>\n",
       "      <td>0.997018</td>\n",
       "      <td>0.002982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>South Asian</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs117710327</th>\n",
       "      <th>European</th>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">rs4808075</th>\n",
       "      <th>European</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>South Asian</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               reference_frequency  alternative_frequency\n",
       "snp_ref_code super_population                                            \n",
       "rs1003531    African                      1.000000               0.000000\n",
       "             American                     0.998559               0.001441\n",
       "             East Asian                   1.000000               0.000000\n",
       "             European                     1.000000               0.000000\n",
       "             South Asian                  0.998978               0.001022\n",
       "rs10038      African                      0.982602               0.017398\n",
       "             American                     0.998559               0.001441\n",
       "             East Asian                   1.000000               0.000000\n",
       "             European                     1.000000               0.000000\n",
       "             South Asian                  1.000000               0.000000\n",
       "rs1004246    African                      0.990923               0.009077\n",
       "             American                     1.000000               0.000000\n",
       "             East Asian                   1.000000               0.000000\n",
       "             European                     0.997018               0.002982\n",
       "             South Asian                  1.000000               0.000000\n",
       "rs117710327  European                     0.375000               0.625000\n",
       "rs4808075    European                     1.000000               0.000000\n",
       "             South Asian                  1.000000               0.000000"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pivDf = df.pivot(index='snp_ref_code', columns='super_population', values=['reference_frequency', 'alternative_frequency'])\n",
    "pivDf = pivDf.stack()\n",
    "pivDf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **A query seguinte, vai calcular, por SNP e para cada Super-população e sub-população, as frequências do alelo de referência e do alelo alternativo.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>super_population</th>\n",
       "      <th>population_description</th>\n",
       "      <th>reference_frequency</th>\n",
       "      <th>alternative_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>African</td>\n",
       "      <td>African Ancestry in Southwest US</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>African</td>\n",
       "      <td>African Caribbean in Barbados</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>African</td>\n",
       "      <td>Esan in Nigeria</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>African</td>\n",
       "      <td>Gambian in Western Division, The Gambia</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>African</td>\n",
       "      <td>Luhya in Webuye, Kenya</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>Punjabi in Lahore,Pakistan</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>Sri Lankan Tamil in the UK</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>rs117710327</td>\n",
       "      <td>European</td>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>0.375</td>\n",
       "      <td>0.625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>rs4808075</td>\n",
       "      <td>European</td>\n",
       "      <td>Finnish in Finland</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>rs4808075</td>\n",
       "      <td>South Asian</td>\n",
       "      <td>Bengali in Bangladesh</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>81 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   snp_ref_code super_population                   population_description  \\\n",
       "0     rs1003531          African         African Ancestry in Southwest US   \n",
       "1     rs1003531          African            African Caribbean in Barbados   \n",
       "2     rs1003531          African                          Esan in Nigeria   \n",
       "3     rs1003531          African  Gambian in Western Division, The Gambia   \n",
       "4     rs1003531          African                   Luhya in Webuye, Kenya   \n",
       "..          ...              ...                                      ...   \n",
       "76    rs1004246      South Asian               Punjabi in Lahore,Pakistan   \n",
       "77    rs1004246      South Asian               Sri Lankan Tamil in the UK   \n",
       "78  rs117710327         European          British in England and Scotland   \n",
       "79    rs4808075         European                       Finnish in Finland   \n",
       "80    rs4808075      South Asian                    Bengali in Bangladesh   \n",
       "\n",
       "    reference_frequency  alternative_frequency  \n",
       "0                 1.000                  0.000  \n",
       "1                 1.000                  0.000  \n",
       "2                 1.000                  0.000  \n",
       "3                 1.000                  0.000  \n",
       "4                 1.000                  0.000  \n",
       "..                  ...                    ...  \n",
       "76                1.000                  0.000  \n",
       "77                1.000                  0.000  \n",
       "78                0.375                  0.625  \n",
       "79                1.000                  0.000  \n",
       "80                1.000                  0.000  \n",
       "\n",
       "[81 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "SELECT snp_ref_code, super_population, population_description, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, population_dimension P, snp_dim S\n",
    "WHERE P.population_key = SP.population_id AND S.snp_id=SP.snp_id\n",
    "GROUP BY super_population, population_description, snp_ref_code\n",
    "ORDER BY snp_ref_code\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df\n",
    "#with pd.option_context('display.max_rows', None, 'display.max_columns', None):  \n",
    "#    display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **A query seguinte, vai calcular, por SNP, País e sub-população, as frequências do alelo de referência e do alelo alternativo.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>country</th>\n",
       "      <th>population_description</th>\n",
       "      <th>reference_frequency</th>\n",
       "      <th>alternative_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>Bengali in Bangladesh</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>African Caribbean in Barbados</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>China</td>\n",
       "      <td>Chinese Dai in Xishuangbanna, China</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>China</td>\n",
       "      <td>Han Chinese in Beijing, China</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>China</td>\n",
       "      <td>Southern Han Chinese, China</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>Kinh in Ho Chi Minh City, Vietnam</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>rs117710327</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>rs117710327</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>Gujarati Indian in Houston,TX</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>rs4808075</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>Bengali in Bangladesh</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>rs4808075</td>\n",
       "      <td>Finland</td>\n",
       "      <td>Finnish in Finland</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   snp_ref_code                   country  \\\n",
       "0     rs1003531                Bangladesh   \n",
       "1     rs1003531                  Barbados   \n",
       "2     rs1003531                     China   \n",
       "3     rs1003531                     China   \n",
       "4     rs1003531                     China   \n",
       "..          ...                       ...   \n",
       "77    rs1004246                   Vietnam   \n",
       "78  rs117710327            United Kingdom   \n",
       "79  rs117710327  United States of America   \n",
       "80    rs4808075                Bangladesh   \n",
       "81    rs4808075                   Finland   \n",
       "\n",
       "                 population_description  reference_frequency  \\\n",
       "0                 Bengali in Bangladesh             1.000000   \n",
       "1         African Caribbean in Barbados             1.000000   \n",
       "2   Chinese Dai in Xishuangbanna, China             1.000000   \n",
       "3         Han Chinese in Beijing, China             1.000000   \n",
       "4           Southern Han Chinese, China             1.000000   \n",
       "..                                  ...                  ...   \n",
       "77    Kinh in Ho Chi Minh City, Vietnam             1.000000   \n",
       "78      British in England and Scotland             0.500000   \n",
       "79        Gujarati Indian in Houston,TX             0.333333   \n",
       "80                Bengali in Bangladesh             1.000000   \n",
       "81                   Finnish in Finland             1.000000   \n",
       "\n",
       "    alternative_frequency  \n",
       "0                0.000000  \n",
       "1                0.000000  \n",
       "2                0.000000  \n",
       "3                0.000000  \n",
       "4                0.000000  \n",
       "..                    ...  \n",
       "77               0.000000  \n",
       "78               0.500000  \n",
       "79               0.666667  \n",
       "80               0.000000  \n",
       "81               0.000000  \n",
       "\n",
       "[82 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "SELECT snp_ref_code, country, population_description, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, person_dimension P, snp_dim S\n",
    "WHERE P.person_key = SP.person_id AND S.snp_id=SP.snp_id\n",
    "GROUP BY country, population_description, snp_ref_code\n",
    "ORDER BY snp_ref_code\n",
    "\"\"\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df\n",
    "\n",
    "#with pd.option_context('display.max_rows', None, 'display.max_columns', None):  \n",
    "#    display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **A query seguinte vai calcular as frequências alélicas de um SNP, por género**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>gender</th>\n",
       "      <th>reference_frequency</th>\n",
       "      <th>alternative_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>female</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>male</td>\n",
       "      <td>0.999189</td>\n",
       "      <td>0.000811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>female</td>\n",
       "      <td>0.994493</td>\n",
       "      <td>0.005507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>male</td>\n",
       "      <td>0.995945</td>\n",
       "      <td>0.004055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>female</td>\n",
       "      <td>0.996853</td>\n",
       "      <td>0.003147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>male</td>\n",
       "      <td>0.997161</td>\n",
       "      <td>0.002839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs117710327</td>\n",
       "      <td>male</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs4808075</td>\n",
       "      <td>male</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  snp_ref_code  gender  reference_frequency  alternative_frequency\n",
       "0    rs1003531  female             1.000000               0.000000\n",
       "1    rs1003531    male             0.999189               0.000811\n",
       "2      rs10038  female             0.994493               0.005507\n",
       "3      rs10038    male             0.995945               0.004055\n",
       "4    rs1004246  female             0.996853               0.003147\n",
       "5    rs1004246    male             0.997161               0.002839\n",
       "6  rs117710327    male             0.375000               0.625000\n",
       "7    rs4808075    male             1.000000               0.000000"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "SELECT snp_ref_code, gender, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, gender_dimension G, snp_dim S\n",
    "WHERE G.gender_key = SP.gender_id AND S.snp_id=SP.snp_id\n",
    "GROUP BY snp_ref_code, gender\n",
    "\"\"\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 2 - Existe uma relação entre variabilidade genética e aspectos socioeconómicos e culturais das populações? (maior taxas de mutações, maior isolamento populacional?)\n",
    "Com este tipo de queries, pretendemos determinar se aspectos socioeconómicos e culturais das populações, podem traduzir-se numa maior ou menos variabilidade genética."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **A query seguinte, vai calcular para cada SNP e para cada país, as frequências alélicas encontradas. Associa esta informação ao produto interno bruto de cada País.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>country</th>\n",
       "      <th>gross_national_income</th>\n",
       "      <th>gross_national_income_group</th>\n",
       "      <th>reference_frequency</th>\n",
       "      <th>alternative_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>7330</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>0.994681</td>\n",
       "      <td>0.005319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>44280</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>0.998305</td>\n",
       "      <td>0.001695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>1220</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>1290</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>1360</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>1970</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>2880</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Peru</td>\n",
       "      <td>6340</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>China</td>\n",
       "      <td>7910</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>15270</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>19420</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Spain</td>\n",
       "      <td>28430</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Italy</td>\n",
       "      <td>32980</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Japan</td>\n",
       "      <td>38840</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Finland</td>\n",
       "      <td>47160</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>56720</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>550</td>\n",
       "      <td>Below 1k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>620</td>\n",
       "      <td>Below 1k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>15270</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>0.973958</td>\n",
       "      <td>0.026042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>1290</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>0.974747</td>\n",
       "      <td>0.025253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>2880</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>0.978261</td>\n",
       "      <td>0.021739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>550</td>\n",
       "      <td>Below 1k</td>\n",
       "      <td>0.994118</td>\n",
       "      <td>0.005882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>19420</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>0.995192</td>\n",
       "      <td>0.004808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>620</td>\n",
       "      <td>Below 1k</td>\n",
       "      <td>0.995575</td>\n",
       "      <td>0.004425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>56720</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>0.996942</td>\n",
       "      <td>0.003058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>1970</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Peru</td>\n",
       "      <td>6340</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>7330</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>China</td>\n",
       "      <td>7910</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Spain</td>\n",
       "      <td>28430</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Italy</td>\n",
       "      <td>32980</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Japan</td>\n",
       "      <td>38840</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>44280</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Finland</td>\n",
       "      <td>47160</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>1220</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>1360</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>550</td>\n",
       "      <td>Below 1k</td>\n",
       "      <td>0.976471</td>\n",
       "      <td>0.023529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>1290</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>0.984848</td>\n",
       "      <td>0.015152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Italy</td>\n",
       "      <td>32980</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>0.990654</td>\n",
       "      <td>0.009346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>620</td>\n",
       "      <td>Below 1k</td>\n",
       "      <td>0.991150</td>\n",
       "      <td>0.008850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>15270</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>0.994792</td>\n",
       "      <td>0.005208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>2880</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>0.995169</td>\n",
       "      <td>0.004831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>56720</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>0.998471</td>\n",
       "      <td>0.001529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>1220</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>1360</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>1970</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Peru</td>\n",
       "      <td>6340</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>7330</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>China</td>\n",
       "      <td>7910</td>\n",
       "      <td>4k - 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>19420</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Spain</td>\n",
       "      <td>28430</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Japan</td>\n",
       "      <td>38840</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>44280</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Finland</td>\n",
       "      <td>47160</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>rs117710327</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>44280</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>rs4808075</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>1220</td>\n",
       "      <td>1k - 4k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>rs4808075</td>\n",
       "      <td>Finland</td>\n",
       "      <td>47160</td>\n",
       "      <td>Above 14k</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   snp_ref_code                   country  gross_national_income  \\\n",
       "0     rs1003531                  Colombia                   7330   \n",
       "1     rs1003531            United Kingdom                  44280   \n",
       "2     rs1003531                Bangladesh                   1220   \n",
       "3     rs1003531                     Kenya                   1290   \n",
       "4     rs1003531                  Pakistan                   1360   \n",
       "5     rs1003531                   Vietnam                   1970   \n",
       "6     rs1003531                   Nigeria                   2880   \n",
       "7     rs1003531                      Peru                   6340   \n",
       "8     rs1003531                     China                   7910   \n",
       "9     rs1003531                  Barbados                  15270   \n",
       "10    rs1003531               Puerto Rico                  19420   \n",
       "11    rs1003531                     Spain                  28430   \n",
       "12    rs1003531                     Italy                  32980   \n",
       "13    rs1003531                     Japan                  38840   \n",
       "14    rs1003531                   Finland                  47160   \n",
       "15    rs1003531  United States of America                  56720   \n",
       "16    rs1003531              Sierra Leone                    550   \n",
       "17    rs1003531                The Gambia                    620   \n",
       "18      rs10038                  Barbados                  15270   \n",
       "19      rs10038                     Kenya                   1290   \n",
       "20      rs10038                   Nigeria                   2880   \n",
       "21      rs10038              Sierra Leone                    550   \n",
       "22      rs10038               Puerto Rico                  19420   \n",
       "23      rs10038                The Gambia                    620   \n",
       "24      rs10038  United States of America                  56720   \n",
       "25      rs10038                   Vietnam                   1970   \n",
       "26      rs10038                      Peru                   6340   \n",
       "27      rs10038                  Colombia                   7330   \n",
       "28      rs10038                     China                   7910   \n",
       "29      rs10038                     Spain                  28430   \n",
       "30      rs10038                     Italy                  32980   \n",
       "31      rs10038                     Japan                  38840   \n",
       "32      rs10038            United Kingdom                  44280   \n",
       "33      rs10038                   Finland                  47160   \n",
       "34      rs10038                Bangladesh                   1220   \n",
       "35      rs10038                  Pakistan                   1360   \n",
       "36    rs1004246              Sierra Leone                    550   \n",
       "37    rs1004246                     Kenya                   1290   \n",
       "38    rs1004246                     Italy                  32980   \n",
       "39    rs1004246                The Gambia                    620   \n",
       "40    rs1004246                  Barbados                  15270   \n",
       "41    rs1004246                   Nigeria                   2880   \n",
       "42    rs1004246  United States of America                  56720   \n",
       "43    rs1004246                Bangladesh                   1220   \n",
       "44    rs1004246                  Pakistan                   1360   \n",
       "45    rs1004246                   Vietnam                   1970   \n",
       "46    rs1004246                      Peru                   6340   \n",
       "47    rs1004246                  Colombia                   7330   \n",
       "48    rs1004246                     China                   7910   \n",
       "49    rs1004246               Puerto Rico                  19420   \n",
       "50    rs1004246                     Spain                  28430   \n",
       "51    rs1004246                     Japan                  38840   \n",
       "52    rs1004246            United Kingdom                  44280   \n",
       "53    rs1004246                   Finland                  47160   \n",
       "54  rs117710327            United Kingdom                  44280   \n",
       "55    rs4808075                Bangladesh                   1220   \n",
       "56    rs4808075                   Finland                  47160   \n",
       "\n",
       "   gross_national_income_group  reference_frequency  alternative_frequency  \n",
       "0                     4k - 14k             0.994681               0.005319  \n",
       "1                    Above 14k             0.998305               0.001695  \n",
       "2                      1k - 4k             1.000000               0.000000  \n",
       "3                      1k - 4k             1.000000               0.000000  \n",
       "4                      1k - 4k             1.000000               0.000000  \n",
       "5                      1k - 4k             1.000000               0.000000  \n",
       "6                      1k - 4k             1.000000               0.000000  \n",
       "7                     4k - 14k             1.000000               0.000000  \n",
       "8                     4k - 14k             1.000000               0.000000  \n",
       "9                    Above 14k             1.000000               0.000000  \n",
       "10                   Above 14k             1.000000               0.000000  \n",
       "11                   Above 14k             1.000000               0.000000  \n",
       "12                   Above 14k             1.000000               0.000000  \n",
       "13                   Above 14k             1.000000               0.000000  \n",
       "14                   Above 14k             1.000000               0.000000  \n",
       "15                   Above 14k             1.000000               0.000000  \n",
       "16                    Below 1k             1.000000               0.000000  \n",
       "17                    Below 1k             1.000000               0.000000  \n",
       "18                   Above 14k             0.973958               0.026042  \n",
       "19                     1k - 4k             0.974747               0.025253  \n",
       "20                     1k - 4k             0.978261               0.021739  \n",
       "21                    Below 1k             0.994118               0.005882  \n",
       "22                   Above 14k             0.995192               0.004808  \n",
       "23                    Below 1k             0.995575               0.004425  \n",
       "24                   Above 14k             0.996942               0.003058  \n",
       "25                     1k - 4k             1.000000               0.000000  \n",
       "26                    4k - 14k             1.000000               0.000000  \n",
       "27                    4k - 14k             1.000000               0.000000  \n",
       "28                    4k - 14k             1.000000               0.000000  \n",
       "29                   Above 14k             1.000000               0.000000  \n",
       "30                   Above 14k             1.000000               0.000000  \n",
       "31                   Above 14k             1.000000               0.000000  \n",
       "32                   Above 14k             1.000000               0.000000  \n",
       "33                   Above 14k             1.000000               0.000000  \n",
       "34                     1k - 4k             1.000000               0.000000  \n",
       "35                     1k - 4k             1.000000               0.000000  \n",
       "36                    Below 1k             0.976471               0.023529  \n",
       "37                     1k - 4k             0.984848               0.015152  \n",
       "38                   Above 14k             0.990654               0.009346  \n",
       "39                    Below 1k             0.991150               0.008850  \n",
       "40                   Above 14k             0.994792               0.005208  \n",
       "41                     1k - 4k             0.995169               0.004831  \n",
       "42                   Above 14k             0.998471               0.001529  \n",
       "43                     1k - 4k             1.000000               0.000000  \n",
       "44                     1k - 4k             1.000000               0.000000  \n",
       "45                     1k - 4k             1.000000               0.000000  \n",
       "46                    4k - 14k             1.000000               0.000000  \n",
       "47                    4k - 14k             1.000000               0.000000  \n",
       "48                    4k - 14k             1.000000               0.000000  \n",
       "49                   Above 14k             1.000000               0.000000  \n",
       "50                   Above 14k             1.000000               0.000000  \n",
       "51                   Above 14k             1.000000               0.000000  \n",
       "52                   Above 14k             1.000000               0.000000  \n",
       "53                   Above 14k             1.000000               0.000000  \n",
       "54                   Above 14k             0.375000               0.625000  \n",
       "55                     1k - 4k             1.000000               0.000000  \n",
       "56                   Above 14k             1.000000               0.000000  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "SELECT snp_ref_code, PD.country, gross_national_income, gross_national_income_group,  (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, snp_dim S, population_dimension PD\n",
    "WHERE  S.snp_id=SP.snp_id AND PD.population_key = SP.population_id\n",
    "GROUP BY snp_ref_code, gross_national_income, gross_national_income_group, PD.country\n",
    "ORDER BY snp_ref_code, Alternative_frequency DESC\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **A query seguinte, vai calcular para cada SNP e para cada país, as frequências alélicas encontradas para cada género. Associa esta informação à taxa de mortalidade no País (dividida por género).**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>country</th>\n",
       "      <th>gender</th>\n",
       "      <th>mortality_per_1000_habitants</th>\n",
       "      <th>reference_frequency</th>\n",
       "      <th>alternative_frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>female</td>\n",
       "      <td>109.47</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>male</td>\n",
       "      <td>150.04</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>female</td>\n",
       "      <td>74.17</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>male</td>\n",
       "      <td>124.02</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>China</td>\n",
       "      <td>female</td>\n",
       "      <td>64.48</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>China</td>\n",
       "      <td>male</td>\n",
       "      <td>100.07</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>female</td>\n",
       "      <td>89.30</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>male</td>\n",
       "      <td>190.84</td>\n",
       "      <td>0.988372</td>\n",
       "      <td>0.011628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Finland</td>\n",
       "      <td>female</td>\n",
       "      <td>43.46</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Finland</td>\n",
       "      <td>male</td>\n",
       "      <td>94.25</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Italy</td>\n",
       "      <td>female</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Italy</td>\n",
       "      <td>male</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Japan</td>\n",
       "      <td>female</td>\n",
       "      <td>37.10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Japan</td>\n",
       "      <td>male</td>\n",
       "      <td>69.98</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>female</td>\n",
       "      <td>183.12</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>male</td>\n",
       "      <td>264.09</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>female</td>\n",
       "      <td>337.66</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>male</td>\n",
       "      <td>374.77</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>female</td>\n",
       "      <td>141.74</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>male</td>\n",
       "      <td>178.54</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Peru</td>\n",
       "      <td>female</td>\n",
       "      <td>94.74</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Peru</td>\n",
       "      <td>male</td>\n",
       "      <td>153.40</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>female</td>\n",
       "      <td>65.04</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>male</td>\n",
       "      <td>151.23</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>female</td>\n",
       "      <td>392.95</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>male</td>\n",
       "      <td>402.80</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Spain</td>\n",
       "      <td>female</td>\n",
       "      <td>38.81</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Spain</td>\n",
       "      <td>male</td>\n",
       "      <td>77.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>female</td>\n",
       "      <td>234.24</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>male</td>\n",
       "      <td>289.18</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>female</td>\n",
       "      <td>54.15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>male</td>\n",
       "      <td>84.74</td>\n",
       "      <td>0.996875</td>\n",
       "      <td>0.003125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>female</td>\n",
       "      <td>80.39</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>male</td>\n",
       "      <td>134.26</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>female</td>\n",
       "      <td>66.44</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>rs1003531</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>male</td>\n",
       "      <td>182.66</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>female</td>\n",
       "      <td>109.47</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>male</td>\n",
       "      <td>150.04</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>female</td>\n",
       "      <td>74.17</td>\n",
       "      <td>0.979592</td>\n",
       "      <td>0.020408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>male</td>\n",
       "      <td>124.02</td>\n",
       "      <td>0.968085</td>\n",
       "      <td>0.031915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>China</td>\n",
       "      <td>female</td>\n",
       "      <td>64.48</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>China</td>\n",
       "      <td>male</td>\n",
       "      <td>100.07</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>female</td>\n",
       "      <td>89.30</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>male</td>\n",
       "      <td>190.84</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Finland</td>\n",
       "      <td>female</td>\n",
       "      <td>43.46</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Finland</td>\n",
       "      <td>male</td>\n",
       "      <td>94.25</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Italy</td>\n",
       "      <td>female</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Italy</td>\n",
       "      <td>male</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Japan</td>\n",
       "      <td>female</td>\n",
       "      <td>37.10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Japan</td>\n",
       "      <td>male</td>\n",
       "      <td>69.98</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>female</td>\n",
       "      <td>183.12</td>\n",
       "      <td>0.963636</td>\n",
       "      <td>0.036364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>male</td>\n",
       "      <td>264.09</td>\n",
       "      <td>0.988636</td>\n",
       "      <td>0.011364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>female</td>\n",
       "      <td>337.66</td>\n",
       "      <td>0.980392</td>\n",
       "      <td>0.019608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>male</td>\n",
       "      <td>374.77</td>\n",
       "      <td>0.976190</td>\n",
       "      <td>0.023810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>female</td>\n",
       "      <td>141.74</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>male</td>\n",
       "      <td>178.54</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Peru</td>\n",
       "      <td>female</td>\n",
       "      <td>94.74</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Peru</td>\n",
       "      <td>male</td>\n",
       "      <td>153.40</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>female</td>\n",
       "      <td>65.04</td>\n",
       "      <td>0.990000</td>\n",
       "      <td>0.010000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>male</td>\n",
       "      <td>151.23</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>female</td>\n",
       "      <td>392.95</td>\n",
       "      <td>0.988372</td>\n",
       "      <td>0.011628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>male</td>\n",
       "      <td>402.80</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Spain</td>\n",
       "      <td>female</td>\n",
       "      <td>38.81</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Spain</td>\n",
       "      <td>male</td>\n",
       "      <td>77.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>female</td>\n",
       "      <td>234.24</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>male</td>\n",
       "      <td>289.18</td>\n",
       "      <td>0.990909</td>\n",
       "      <td>0.009091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>female</td>\n",
       "      <td>54.15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>male</td>\n",
       "      <td>84.74</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>female</td>\n",
       "      <td>80.39</td>\n",
       "      <td>0.993902</td>\n",
       "      <td>0.006098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>male</td>\n",
       "      <td>134.26</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>female</td>\n",
       "      <td>66.44</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>rs10038</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>male</td>\n",
       "      <td>182.66</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>female</td>\n",
       "      <td>109.47</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>male</td>\n",
       "      <td>150.04</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>female</td>\n",
       "      <td>74.17</td>\n",
       "      <td>0.989796</td>\n",
       "      <td>0.010204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Barbados</td>\n",
       "      <td>male</td>\n",
       "      <td>124.02</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>China</td>\n",
       "      <td>female</td>\n",
       "      <td>64.48</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>China</td>\n",
       "      <td>male</td>\n",
       "      <td>100.07</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>female</td>\n",
       "      <td>89.30</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Colombia</td>\n",
       "      <td>male</td>\n",
       "      <td>190.84</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Finland</td>\n",
       "      <td>female</td>\n",
       "      <td>43.46</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Finland</td>\n",
       "      <td>male</td>\n",
       "      <td>94.25</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Italy</td>\n",
       "      <td>female</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.990741</td>\n",
       "      <td>0.009259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Italy</td>\n",
       "      <td>male</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.990566</td>\n",
       "      <td>0.009434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Japan</td>\n",
       "      <td>female</td>\n",
       "      <td>37.10</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Japan</td>\n",
       "      <td>male</td>\n",
       "      <td>69.98</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>female</td>\n",
       "      <td>183.12</td>\n",
       "      <td>0.990909</td>\n",
       "      <td>0.009091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Kenya</td>\n",
       "      <td>male</td>\n",
       "      <td>264.09</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.022727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>female</td>\n",
       "      <td>337.66</td>\n",
       "      <td>0.990196</td>\n",
       "      <td>0.009804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Nigeria</td>\n",
       "      <td>male</td>\n",
       "      <td>374.77</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>female</td>\n",
       "      <td>141.74</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>male</td>\n",
       "      <td>178.54</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Peru</td>\n",
       "      <td>female</td>\n",
       "      <td>94.74</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Peru</td>\n",
       "      <td>male</td>\n",
       "      <td>153.40</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>female</td>\n",
       "      <td>65.04</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Puerto Rico</td>\n",
       "      <td>male</td>\n",
       "      <td>151.23</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>female</td>\n",
       "      <td>392.95</td>\n",
       "      <td>0.988372</td>\n",
       "      <td>0.011628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Sierra Leone</td>\n",
       "      <td>male</td>\n",
       "      <td>402.80</td>\n",
       "      <td>0.964286</td>\n",
       "      <td>0.035714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Spain</td>\n",
       "      <td>female</td>\n",
       "      <td>38.81</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Spain</td>\n",
       "      <td>male</td>\n",
       "      <td>77.00</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>female</td>\n",
       "      <td>234.24</td>\n",
       "      <td>0.982759</td>\n",
       "      <td>0.017241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>The Gambia</td>\n",
       "      <td>male</td>\n",
       "      <td>289.18</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>female</td>\n",
       "      <td>54.15</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>male</td>\n",
       "      <td>84.74</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>female</td>\n",
       "      <td>80.39</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>United States of America</td>\n",
       "      <td>male</td>\n",
       "      <td>134.26</td>\n",
       "      <td>0.996933</td>\n",
       "      <td>0.003067</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>female</td>\n",
       "      <td>66.44</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>rs1004246</td>\n",
       "      <td>Vietnam</td>\n",
       "      <td>male</td>\n",
       "      <td>182.66</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>rs117710327</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>male</td>\n",
       "      <td>84.74</td>\n",
       "      <td>0.375000</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>rs4808075</td>\n",
       "      <td>Bangladesh</td>\n",
       "      <td>male</td>\n",
       "      <td>84.74</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>rs4808075</td>\n",
       "      <td>Finland</td>\n",
       "      <td>male</td>\n",
       "      <td>84.74</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    snp_ref_code                   country  gender  \\\n",
       "0      rs1003531                Bangladesh  female   \n",
       "1      rs1003531                Bangladesh    male   \n",
       "2      rs1003531                  Barbados  female   \n",
       "3      rs1003531                  Barbados    male   \n",
       "4      rs1003531                     China  female   \n",
       "5      rs1003531                     China    male   \n",
       "6      rs1003531                  Colombia  female   \n",
       "7      rs1003531                  Colombia    male   \n",
       "8      rs1003531                   Finland  female   \n",
       "9      rs1003531                   Finland    male   \n",
       "10     rs1003531                     Italy  female   \n",
       "11     rs1003531                     Italy    male   \n",
       "12     rs1003531                     Japan  female   \n",
       "13     rs1003531                     Japan    male   \n",
       "14     rs1003531                     Kenya  female   \n",
       "15     rs1003531                     Kenya    male   \n",
       "16     rs1003531                   Nigeria  female   \n",
       "17     rs1003531                   Nigeria    male   \n",
       "18     rs1003531                  Pakistan  female   \n",
       "19     rs1003531                  Pakistan    male   \n",
       "20     rs1003531                      Peru  female   \n",
       "21     rs1003531                      Peru    male   \n",
       "22     rs1003531               Puerto Rico  female   \n",
       "23     rs1003531               Puerto Rico    male   \n",
       "24     rs1003531              Sierra Leone  female   \n",
       "25     rs1003531              Sierra Leone    male   \n",
       "26     rs1003531                     Spain  female   \n",
       "27     rs1003531                     Spain    male   \n",
       "28     rs1003531                The Gambia  female   \n",
       "29     rs1003531                The Gambia    male   \n",
       "30     rs1003531            United Kingdom  female   \n",
       "31     rs1003531            United Kingdom    male   \n",
       "32     rs1003531  United States of America  female   \n",
       "33     rs1003531  United States of America    male   \n",
       "34     rs1003531                   Vietnam  female   \n",
       "35     rs1003531                   Vietnam    male   \n",
       "36       rs10038                Bangladesh  female   \n",
       "37       rs10038                Bangladesh    male   \n",
       "38       rs10038                  Barbados  female   \n",
       "39       rs10038                  Barbados    male   \n",
       "40       rs10038                     China  female   \n",
       "41       rs10038                     China    male   \n",
       "42       rs10038                  Colombia  female   \n",
       "43       rs10038                  Colombia    male   \n",
       "44       rs10038                   Finland  female   \n",
       "45       rs10038                   Finland    male   \n",
       "46       rs10038                     Italy  female   \n",
       "47       rs10038                     Italy    male   \n",
       "48       rs10038                     Japan  female   \n",
       "49       rs10038                     Japan    male   \n",
       "50       rs10038                     Kenya  female   \n",
       "51       rs10038                     Kenya    male   \n",
       "52       rs10038                   Nigeria  female   \n",
       "53       rs10038                   Nigeria    male   \n",
       "54       rs10038                  Pakistan  female   \n",
       "55       rs10038                  Pakistan    male   \n",
       "56       rs10038                      Peru  female   \n",
       "57       rs10038                      Peru    male   \n",
       "58       rs10038               Puerto Rico  female   \n",
       "59       rs10038               Puerto Rico    male   \n",
       "60       rs10038              Sierra Leone  female   \n",
       "61       rs10038              Sierra Leone    male   \n",
       "62       rs10038                     Spain  female   \n",
       "63       rs10038                     Spain    male   \n",
       "64       rs10038                The Gambia  female   \n",
       "65       rs10038                The Gambia    male   \n",
       "66       rs10038            United Kingdom  female   \n",
       "67       rs10038            United Kingdom    male   \n",
       "68       rs10038  United States of America  female   \n",
       "69       rs10038  United States of America    male   \n",
       "70       rs10038                   Vietnam  female   \n",
       "71       rs10038                   Vietnam    male   \n",
       "72     rs1004246                Bangladesh  female   \n",
       "73     rs1004246                Bangladesh    male   \n",
       "74     rs1004246                  Barbados  female   \n",
       "75     rs1004246                  Barbados    male   \n",
       "76     rs1004246                     China  female   \n",
       "77     rs1004246                     China    male   \n",
       "78     rs1004246                  Colombia  female   \n",
       "79     rs1004246                  Colombia    male   \n",
       "80     rs1004246                   Finland  female   \n",
       "81     rs1004246                   Finland    male   \n",
       "82     rs1004246                     Italy  female   \n",
       "83     rs1004246                     Italy    male   \n",
       "84     rs1004246                     Japan  female   \n",
       "85     rs1004246                     Japan    male   \n",
       "86     rs1004246                     Kenya  female   \n",
       "87     rs1004246                     Kenya    male   \n",
       "88     rs1004246                   Nigeria  female   \n",
       "89     rs1004246                   Nigeria    male   \n",
       "90     rs1004246                  Pakistan  female   \n",
       "91     rs1004246                  Pakistan    male   \n",
       "92     rs1004246                      Peru  female   \n",
       "93     rs1004246                      Peru    male   \n",
       "94     rs1004246               Puerto Rico  female   \n",
       "95     rs1004246               Puerto Rico    male   \n",
       "96     rs1004246              Sierra Leone  female   \n",
       "97     rs1004246              Sierra Leone    male   \n",
       "98     rs1004246                     Spain  female   \n",
       "99     rs1004246                     Spain    male   \n",
       "100    rs1004246                The Gambia  female   \n",
       "101    rs1004246                The Gambia    male   \n",
       "102    rs1004246            United Kingdom  female   \n",
       "103    rs1004246            United Kingdom    male   \n",
       "104    rs1004246  United States of America  female   \n",
       "105    rs1004246  United States of America    male   \n",
       "106    rs1004246                   Vietnam  female   \n",
       "107    rs1004246                   Vietnam    male   \n",
       "108  rs117710327            United Kingdom    male   \n",
       "109    rs4808075                Bangladesh    male   \n",
       "110    rs4808075                   Finland    male   \n",
       "\n",
       "     mortality_per_1000_habitants  reference_frequency  alternative_frequency  \n",
       "0                          109.47             1.000000               0.000000  \n",
       "1                          150.04             1.000000               0.000000  \n",
       "2                           74.17             1.000000               0.000000  \n",
       "3                          124.02             1.000000               0.000000  \n",
       "4                           64.48             1.000000               0.000000  \n",
       "5                          100.07             1.000000               0.000000  \n",
       "6                           89.30             1.000000               0.000000  \n",
       "7                          190.84             0.988372               0.011628  \n",
       "8                           43.46             1.000000               0.000000  \n",
       "9                           94.25             1.000000               0.000000  \n",
       "10                           0.00             1.000000               0.000000  \n",
       "11                           0.00             1.000000               0.000000  \n",
       "12                          37.10             1.000000               0.000000  \n",
       "13                          69.98             1.000000               0.000000  \n",
       "14                         183.12             1.000000               0.000000  \n",
       "15                         264.09             1.000000               0.000000  \n",
       "16                         337.66             1.000000               0.000000  \n",
       "17                         374.77             1.000000               0.000000  \n",
       "18                         141.74             1.000000               0.000000  \n",
       "19                         178.54             1.000000               0.000000  \n",
       "20                          94.74             1.000000               0.000000  \n",
       "21                         153.40             1.000000               0.000000  \n",
       "22                          65.04             1.000000               0.000000  \n",
       "23                         151.23             1.000000               0.000000  \n",
       "24                         392.95             1.000000               0.000000  \n",
       "25                         402.80             1.000000               0.000000  \n",
       "26                          38.81             1.000000               0.000000  \n",
       "27                          77.00             1.000000               0.000000  \n",
       "28                         234.24             1.000000               0.000000  \n",
       "29                         289.18             1.000000               0.000000  \n",
       "30                          54.15             1.000000               0.000000  \n",
       "31                          84.74             0.996875               0.003125  \n",
       "32                          80.39             1.000000               0.000000  \n",
       "33                         134.26             1.000000               0.000000  \n",
       "34                          66.44             1.000000               0.000000  \n",
       "35                         182.66             1.000000               0.000000  \n",
       "36                         109.47             1.000000               0.000000  \n",
       "37                         150.04             1.000000               0.000000  \n",
       "38                          74.17             0.979592               0.020408  \n",
       "39                         124.02             0.968085               0.031915  \n",
       "40                          64.48             1.000000               0.000000  \n",
       "41                         100.07             1.000000               0.000000  \n",
       "42                          89.30             1.000000               0.000000  \n",
       "43                         190.84             1.000000               0.000000  \n",
       "44                          43.46             1.000000               0.000000  \n",
       "45                          94.25             1.000000               0.000000  \n",
       "46                           0.00             1.000000               0.000000  \n",
       "47                           0.00             1.000000               0.000000  \n",
       "48                          37.10             1.000000               0.000000  \n",
       "49                          69.98             1.000000               0.000000  \n",
       "50                         183.12             0.963636               0.036364  \n",
       "51                         264.09             0.988636               0.011364  \n",
       "52                         337.66             0.980392               0.019608  \n",
       "53                         374.77             0.976190               0.023810  \n",
       "54                         141.74             1.000000               0.000000  \n",
       "55                         178.54             1.000000               0.000000  \n",
       "56                          94.74             1.000000               0.000000  \n",
       "57                         153.40             1.000000               0.000000  \n",
       "58                          65.04             0.990000               0.010000  \n",
       "59                         151.23             1.000000               0.000000  \n",
       "60                         392.95             0.988372               0.011628  \n",
       "61                         402.80             1.000000               0.000000  \n",
       "62                          38.81             1.000000               0.000000  \n",
       "63                          77.00             1.000000               0.000000  \n",
       "64                         234.24             1.000000               0.000000  \n",
       "65                         289.18             0.990909               0.009091  \n",
       "66                          54.15             1.000000               0.000000  \n",
       "67                          84.74             1.000000               0.000000  \n",
       "68                          80.39             0.993902               0.006098  \n",
       "69                         134.26             1.000000               0.000000  \n",
       "70                          66.44             1.000000               0.000000  \n",
       "71                         182.66             1.000000               0.000000  \n",
       "72                         109.47             1.000000               0.000000  \n",
       "73                         150.04             1.000000               0.000000  \n",
       "74                          74.17             0.989796               0.010204  \n",
       "75                         124.02             1.000000               0.000000  \n",
       "76                          64.48             1.000000               0.000000  \n",
       "77                         100.07             1.000000               0.000000  \n",
       "78                          89.30             1.000000               0.000000  \n",
       "79                         190.84             1.000000               0.000000  \n",
       "80                          43.46             1.000000               0.000000  \n",
       "81                          94.25             1.000000               0.000000  \n",
       "82                           0.00             0.990741               0.009259  \n",
       "83                           0.00             0.990566               0.009434  \n",
       "84                          37.10             1.000000               0.000000  \n",
       "85                          69.98             1.000000               0.000000  \n",
       "86                         183.12             0.990909               0.009091  \n",
       "87                         264.09             0.977273               0.022727  \n",
       "88                         337.66             0.990196               0.009804  \n",
       "89                         374.77             1.000000               0.000000  \n",
       "90                         141.74             1.000000               0.000000  \n",
       "91                         178.54             1.000000               0.000000  \n",
       "92                          94.74             1.000000               0.000000  \n",
       "93                         153.40             1.000000               0.000000  \n",
       "94                          65.04             1.000000               0.000000  \n",
       "95                         151.23             1.000000               0.000000  \n",
       "96                         392.95             0.988372               0.011628  \n",
       "97                         402.80             0.964286               0.035714  \n",
       "98                          38.81             1.000000               0.000000  \n",
       "99                          77.00             1.000000               0.000000  \n",
       "100                        234.24             0.982759               0.017241  \n",
       "101                        289.18             1.000000               0.000000  \n",
       "102                         54.15             1.000000               0.000000  \n",
       "103                         84.74             1.000000               0.000000  \n",
       "104                         80.39             1.000000               0.000000  \n",
       "105                        134.26             0.996933               0.003067  \n",
       "106                         66.44             1.000000               0.000000  \n",
       "107                        182.66             1.000000               0.000000  \n",
       "108                         84.74             0.375000               0.625000  \n",
       "109                         84.74             1.000000               0.000000  \n",
       "110                         84.74             1.000000               0.000000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "SELECT  snp_ref_code, PD.country, gender,  mortality_per_1000_habitants, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, gender_dimension G, snp_dim S, population_dimension PD\n",
    "WHERE G.gender_key = SP.gender_id AND S.snp_id=SP.snp_id AND PD.population_key = SP.population_id\n",
    "GROUP BY snp_ref_code, gender, gross_national_income, PD.country, mortality_per_1000_habitants\n",
    "ORDER BY snp_ref_code, country\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df\n",
    "with pd.option_context('display.max_rows', None, 'display.max_columns', None): \n",
    "    display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 3: Em que populações uma determinada doença complexa tem maior probabilidade de ocorrer?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Com estas queries, pretendemos determinar quais populações vão ter uma maior frequência de variantes genéticas, associadas a determinadas doenças.\n",
    "\n",
    "- Neste caso, vamos usar asma para testar a nossa query. É uma doença interessante para testar, pois está associada a mais do que um SNP de risco. \n",
    "- Para testar a doença, inserimos \"dummy data\" na snp_person_fact, para o SNP 1053 e três indívíduos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comando SQL para inserir dados fictícios.\n",
    "SQL = \"\"\"\n",
    "    insert into snp_person_fact values (13,18,1053,1,19,1,0);\n",
    "    insert into snp_person_fact values (13,18,1053,2502,19,1,1);\n",
    "    insert into snp_person_fact values (13,18,1053,2503,19,1,0); \n",
    "    insert into snp_person_fact values (13,18,1053,2504,19,0,1);\n",
    "    \"\"\"\n",
    "#!!!!!!!! comentar esta proxima linha para evitar dados repetidos em nova execucao\n",
    "#excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **A query seguinte vai calcular a frequência absoluta de um SNP de risco associado a uma determinada doença, em indíviduos de determinadas sub-populações**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>super_population</th>\n",
       "      <th>population_description</th>\n",
       "      <th>abs_freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>South Asian</td>\n",
       "      <td>Gujarati Indian in Houston,TX</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>European</td>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  super_population           population_description  abs_freq\n",
       "0      South Asian    Gujarati Indian in Houston,TX         3\n",
       "1         European  British in England and Scotland         1"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complex_disease = 'asthma'\n",
    "\n",
    "SQL = \"\"\"\n",
    "SELECT super_population, population_description, count(population_description) AS abs_freq \n",
    "FROM person_dimension PD \n",
    "WHERE person_key in (\n",
    "  SELECT person_id \n",
    "  FROM snp_person_fact SP\n",
    "  WHERE snp_id in (\n",
    "    SELECT snp_id \n",
    "    FROM disease_gwas_fact DG\n",
    "    WHERE disease_id in (\n",
    "      SELECT disease_key \n",
    "      FROM disease_dimension D\n",
    "      WHERE disease_name like %s\n",
    "\t  )\n",
    "    )\n",
    "    AND (SP.allele_1 = 1 or SP.allele_2 = 1)\n",
    ")\n",
    "GROUP BY super_population, population_description \n",
    "ORDER BY abs_freq desc\n",
    "\"\"\"\n",
    "\n",
    "df1  = getSQLfromQuery(SQL, (complex_disease,), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **A query seguinte vai calcular a frequência absoluta e relativas de um SNP de risco associado a uma determinada doença, em todos os indíviduos de uma determinada sub-população**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL = \"\"\"\n",
    "SELECT count(*) \n",
    "FROM person_dimension \n",
    "WHERE population_description = %s\n",
    "\"\"\"\n",
    "\n",
    "result_array = []\n",
    "for p in df1['population_description']:\n",
    "    df2 = getSQLfromQuery(SQL, (p,), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "    result_array.append(df2['count'][0])\n",
    "    \n",
    "#for pop in df['population_description']:\n",
    "#    print(pop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>super_population</th>\n",
       "      <th>population_description</th>\n",
       "      <th>abs_freq</th>\n",
       "      <th>count</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>South Asian</td>\n",
       "      <td>Gujarati Indian in Houston,TX</td>\n",
       "      <td>3</td>\n",
       "      <td>103</td>\n",
       "      <td>0.029126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>European</td>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>1</td>\n",
       "      <td>91</td>\n",
       "      <td>0.010989</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  super_population           population_description  abs_freq  count      freq\n",
       "0      South Asian    Gujarati Indian in Houston,TX         3    103  0.029126\n",
       "1         European  British in England and Scotland         1     91  0.010989"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1['count'] = result_array\n",
    "df1['freq'] = df1['abs_freq'] / df1['count']\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 4 - As variantes genéticas identificadas estão em que contexto genómico? Existe uma maior associação de doenças com genes ou com regiões não codificantes?\n",
    "Nestas queries, procuramos entender o contexto genómico em que variantes genéticas associadas com doenças, se encontram."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comecemos por analisar em que contexto genómico se encontram os vários SNPs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>intergenic_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>regulatory_region_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>splice_donor_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>splice_region_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>stop_gained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>synonymous_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TF_binding_site_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               context\n",
       "0                  3_prime_UTR_variant\n",
       "1                  5_prime_UTR_variant\n",
       "2                   intergenic_variant\n",
       "3                       intron_variant\n",
       "4                     missense_variant\n",
       "5   non_coding_transcript_exon_variant\n",
       "6            regulatory_region_variant\n",
       "7                 splice_donor_variant\n",
       "8                splice_region_variant\n",
       "9                          stop_gained\n",
       "10                  synonymous_variant\n",
       "11             TF_binding_site_variant"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL =\"\"\"\n",
    "SELECT context \n",
    "FROM snp_dim \n",
    "GROUP BY context \n",
    "ORDER BY context\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Comecemos por fazer uma query para saber, por doença, qual o SNP de risco que lhe está associado e em que contexto genómico se encontra**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_name</th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>abdominal aortic aneurysm</td>\n",
       "      <td>rs6511720</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aggressive periodontitis</td>\n",
       "      <td>rs4284742</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>allergic rhinitis</td>\n",
       "      <td>rs11671925</td>\n",
       "      <td>TF_binding_site_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>allergic rhinitis</td>\n",
       "      <td>rs12973620</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>alopecia areata</td>\n",
       "      <td>rs17834679</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>urolithiasis</td>\n",
       "      <td>rs74956940</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>varicose veins</td>\n",
       "      <td>rs584768</td>\n",
       "      <td>intergenic_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>vitiligo</td>\n",
       "      <td>rs2304206</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>vitiligo</td>\n",
       "      <td>rs4807000</td>\n",
       "      <td>regulatory_region_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>vitiligo</td>\n",
       "      <td>rs6510827</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>384 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  disease_name snp_ref_code                    context\n",
       "0    abdominal aortic aneurysm    rs6511720             intron_variant\n",
       "1     Aggressive periodontitis    rs4284742             intron_variant\n",
       "2            allergic rhinitis   rs11671925    TF_binding_site_variant\n",
       "3            allergic rhinitis   rs12973620             intron_variant\n",
       "4              alopecia areata   rs17834679             intron_variant\n",
       "..                         ...          ...                        ...\n",
       "379               urolithiasis   rs74956940             intron_variant\n",
       "380             varicose veins     rs584768         intergenic_variant\n",
       "381                   vitiligo    rs2304206        5_prime_UTR_variant\n",
       "382                   vitiligo    rs4807000  regulatory_region_variant\n",
       "383                   vitiligo    rs6510827             intron_variant\n",
       "\n",
       "[384 rows x 3 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL =\"\"\"\n",
    "SELECT disease_name, snp_ref_code, context \n",
    "FROM disease_dimension D , snp_dim S, disease_gwas_fact DF\n",
    "WHERE DF.disease_id = D.disease_key  and S.snp_id = DF.snp_id\n",
    "GROUP BY disease_name, snp_ref_code, context \n",
    "ORDER BY disease_name\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Vamos agrupar as nossas variantes genéticas desta forma:\n",
    "#### non_coding:\n",
    "- 3_prime_UTR_variant\n",
    "- 5_prime_UTR_variant\n",
    "- intergenic_variant\n",
    "- intron_variant\n",
    "- non_coding_transcript_exon_variant\n",
    "- regulatory_region_variant\n",
    "- TF_binding_site_variant\n",
    "\n",
    "#### coding:\n",
    "- missense_variant\n",
    "- splice_donor_variant\n",
    "- splice_region_variant\n",
    "- stop_gained\n",
    "- synonymous_variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Existe uma maior associação de doenças com genes ou com regiões não codificantes?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **As próximas queries dão-nos uma lista de SNPs associadas às doenças no nosso DW, e a localização genómica desses SNPs**\n",
    "- Lista de doenças e SNPs de risco, em contextos \"non-coding\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_name</th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>non_coding_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>abdominal aortic aneurysm</td>\n",
       "      <td>rs6511720</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aggressive periodontitis</td>\n",
       "      <td>rs4284742</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>allergic rhinitis</td>\n",
       "      <td>rs11671925</td>\n",
       "      <td>TF_binding_site_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>allergic rhinitis</td>\n",
       "      <td>rs12973620</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>alopecia areata</td>\n",
       "      <td>rs17834679</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>urolithiasis</td>\n",
       "      <td>rs74956940</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>varicose veins</td>\n",
       "      <td>rs584768</td>\n",
       "      <td>intergenic_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>vitiligo</td>\n",
       "      <td>rs2304206</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>vitiligo</td>\n",
       "      <td>rs4807000</td>\n",
       "      <td>regulatory_region_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>vitiligo</td>\n",
       "      <td>rs6510827</td>\n",
       "      <td>intron_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>322 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  disease_name snp_ref_code         non_coding_context\n",
       "0    abdominal aortic aneurysm    rs6511720             intron_variant\n",
       "1     Aggressive periodontitis    rs4284742             intron_variant\n",
       "2            allergic rhinitis   rs11671925    TF_binding_site_variant\n",
       "3            allergic rhinitis   rs12973620             intron_variant\n",
       "4              alopecia areata   rs17834679             intron_variant\n",
       "..                         ...          ...                        ...\n",
       "317               urolithiasis   rs74956940             intron_variant\n",
       "318             varicose veins     rs584768         intergenic_variant\n",
       "319                   vitiligo    rs2304206        5_prime_UTR_variant\n",
       "320                   vitiligo    rs4807000  regulatory_region_variant\n",
       "321                   vitiligo    rs6510827             intron_variant\n",
       "\n",
       "[322 rows x 3 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL =\"\"\"\n",
    "SELECT disease_name, snp_ref_code, context AS non_coding_context\n",
    "FROM disease_dimension D , snp_dim S, disease_gwas_fact DF\n",
    "WHERE DF.disease_id = D.disease_key  and S.snp_id = DF.snp_id AND \n",
    "(context like '3_prime_UTR_variant' or context like '5_prime_UTR_variant' or context like 'intergenic_variant' or context like 'intron_variant' or context like 'non_coding_transcript_exon_variant' or context like 'regulatory_region_variant' or context like 'TF_binding_site_variant') \n",
    "GROUP BY disease_name, snp_ref_code, context \n",
    "ORDER BY disease_name\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Lista de doenças e SNPs de risco, em contextos \"coding\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>disease_name</th>\n",
       "      <th>snp_ref_code</th>\n",
       "      <th>coding_context</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>rs283814</td>\n",
       "      <td>synonymous_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>rs429358</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Alzheimer's disease</td>\n",
       "      <td>rs7412</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bipolar disorder</td>\n",
       "      <td>rs7250872</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>breast cancer</td>\n",
       "      <td>rs2363956</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>tuberculosis</td>\n",
       "      <td>rs1434579</td>\n",
       "      <td>stop_gained</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>type 1 diabetes mellitus 7</td>\n",
       "      <td>rs12720356</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>type 1 diabetes mellitus 7</td>\n",
       "      <td>rs2304256</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>type 1 diabetes mellitus 7</td>\n",
       "      <td>rs34536443</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>ulcerative colitis</td>\n",
       "      <td>rs12720356</td>\n",
       "      <td>missense_variant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  disease_name snp_ref_code      coding_context\n",
       "0          Alzheimer's disease     rs283814  synonymous_variant\n",
       "1          Alzheimer's disease     rs429358    missense_variant\n",
       "2          Alzheimer's disease       rs7412    missense_variant\n",
       "3             bipolar disorder    rs7250872    missense_variant\n",
       "4                breast cancer    rs2363956    missense_variant\n",
       "..                         ...          ...                 ...\n",
       "57                tuberculosis    rs1434579         stop_gained\n",
       "58  type 1 diabetes mellitus 7   rs12720356    missense_variant\n",
       "59  type 1 diabetes mellitus 7    rs2304256    missense_variant\n",
       "60  type 1 diabetes mellitus 7   rs34536443    missense_variant\n",
       "61          ulcerative colitis   rs12720356    missense_variant\n",
       "\n",
       "[62 rows x 3 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL =\"\"\"\n",
    "SELECT disease_name, snp_ref_code, context AS coding_context\n",
    "FROM disease_dimension D , snp_dim S, disease_gwas_fact DF\n",
    "WHERE DF.disease_id = D.disease_key  and S.snp_id = DF.snp_id AND \n",
    "(context like 'missense_variant' or context like 'splice_donor_variant' or context like 'splice_region_variant' or context like 'stop_gained' or context like 'synonymous_variant') \n",
    "GROUP BY disease_name, snp_ref_code, context \n",
    "ORDER BY disease_name\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **A próxima query dá-nos a frequência com que variações genéticas de risco, associadas a doenças, ocorrem em contexto \"non-coding\" e \"coding\"**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>non_coding</th>\n",
       "      <th>coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.838346</td>\n",
       "      <td>0.161654</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   non_coding    coding\n",
       "0    0.838346  0.161654"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL =\"\"\"\n",
    "SELECT cast(count(case when context like '3_prime_UTR_variant' or context like '5_prime_UTR_variant' or context like 'intergenic_variant' or context like 'intron_variant' or context like 'non_coding_transcript_exon_variant' or context like 'regulatory_region_variant' or context like 'TF_binding_site_variant' then 1 else null end)AS float)/count(context) AS non_coding, \n",
    "       cast(count(case when context like 'missense_variant' or context like 'splice_donor_variant' or context like 'splice_region_variant' or context like 'stop_gained' or context like 'synonymous_variant' then 1 else null end)AS float)/count(context) AS coding\n",
    "FROM disease_dimension D , snp_dim S, disease_gwas_fact DF\n",
    "WHERE DF.disease_id = D.disease_key  and S.snp_id = DF.snp_id\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Finalmente, vamos calcular o número de vezes e a frequência, que uma determinada variante genética de risco, se encontra num context genómico particular**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>context</th>\n",
       "      <th>count</th>\n",
       "      <th>frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>intron_variant</td>\n",
       "      <td>207</td>\n",
       "      <td>0.539062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>2</td>\n",
       "      <td>0.005208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>splice_region_variant</td>\n",
       "      <td>5</td>\n",
       "      <td>0.013021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>16</td>\n",
       "      <td>0.041667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>regulatory_region_variant</td>\n",
       "      <td>29</td>\n",
       "      <td>0.075521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TF_binding_site_variant</td>\n",
       "      <td>5</td>\n",
       "      <td>0.013021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>9</td>\n",
       "      <td>0.023438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>49</td>\n",
       "      <td>0.127604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>14</td>\n",
       "      <td>0.036458</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>missense_variant</td>\n",
       "      <td>46</td>\n",
       "      <td>0.119792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>stop_gained</td>\n",
       "      <td>2</td>\n",
       "      <td>0.005208</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               context  count  frequency\n",
       "0                       intron_variant    207   0.539062\n",
       "1                  5_prime_UTR_variant      2   0.005208\n",
       "2                splice_region_variant      5   0.013021\n",
       "3   non_coding_transcript_exon_variant     16   0.041667\n",
       "4            regulatory_region_variant     29   0.075521\n",
       "5              TF_binding_site_variant      5   0.013021\n",
       "6                   synonymous_variant      9   0.023438\n",
       "7                   intergenic_variant     49   0.127604\n",
       "8                  3_prime_UTR_variant     14   0.036458\n",
       "9                     missense_variant     46   0.119792\n",
       "10                         stop_gained      2   0.005208"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL =\"\"\"\n",
    "WITH myquery AS (\n",
    "SELECT disease_name, snp_ref_code, context \n",
    "FROM disease_dimension D , snp_dim S, disease_gwas_fact DF\n",
    "WHERE DF.disease_id = D.disease_key  and S.snp_id = DF.snp_id\n",
    "GROUP BY disease_name, snp_ref_code, context \n",
    "ORDER BY disease_name\n",
    ")\n",
    "\n",
    "SELECT context, count(context), count(context)/(select count(*) from myquery)::float AS frequency \n",
    "FROM myquery \n",
    "GROUP BY context\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query 5 - Em populações com igual frequência de alelos de risco para uma determinada doença, qual o impacto que aspectos socioeconómicos têm sobre a incidência da doença?\n",
    "Com estas queries, pretendiamos analisar como o contexto demográfico influencia a taxa de incidência de uma determinada doença. Por outras palavras, saber qual o peso do \"ambiente\" na manifestação de doenças.\n",
    "Infelizmente, não conseguimos obter os dados para a prevalência/incidência das doenças em diferentes populações. Por essa razão, esta query tornou-se inviável.\n",
    "\n",
    "Vamos de qualquer modo discutir, brevemente, o procedimento que teríamos seguido para elaborar esta query."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Comecemos por inserir factos ficticios para garantirmos que a query poderia funcionar. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL = \"\"\"\n",
    "insert into snp_person_fact values (13,19,4179,56,19,0,0);\n",
    "insert into snp_person_fact values (13,19,4179,57,19,0,0);\n",
    "insert into snp_person_fact values (13,6,4179,1347,19,0,0);\n",
    "insert into snp_person_fact values (13,6,4179,1348,19,0,0);\n",
    "\"\"\"\n",
    "#!!!!!!!! comentar esta proxima linha para evitar dados repetidos em nova execucao\n",
    "# excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- A query seguinte dá-nos a frequência absoluta por população.\n",
    "- De seguida, escolheríamos entre os resultados, quais as populações com a mesma frequência absoluta dos alelos de risco."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population_id</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   population_id  count\n",
       "0              6      2\n",
       "1             19      2"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_disease = 'cancer'\n",
    "\n",
    "SQL = \"\"\"\n",
    "SELECT population_id, count(population_id) \n",
    "FROM snp_person_fact \n",
    "WHERE snp_id IN  (\n",
    "  SELECT snp_id \n",
    "  FROM disease_gwas_fact \n",
    "  WHERE disease_id IN (\n",
    "    SELECT disease_key \n",
    "    FROM disease_dimension \n",
    "    WHERE disease_name = %s\n",
    "  )\n",
    ")\n",
    "GROUP BY population_id\n",
    "\"\"\"\n",
    "\n",
    "df1  = getSQLfromQuery(SQL, (test_disease,), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Descobertas as populações com as mesmas frequências aproximadas da doença, agora teríamos de comparar os factores económicos com a taxa de prevalência/incidência da doença (dados que não conseguimos obter).\n",
    "- Na tabela produzida por esta query, seria necessário ter também a prevalência da doença em ambas as populacoes para compararmos com factores sócio económicos."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>population_key</th>\n",
       "      <th>population_description</th>\n",
       "      <th>abs_freq</th>\n",
       "      <th>life_expectancy_birth</th>\n",
       "      <th>gross_national_income</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>6</td>\n",
       "      <td>Bengali in Bangladesh</td>\n",
       "      <td>2</td>\n",
       "      <td>71.51</td>\n",
       "      <td>1220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>Finnish in Finland</td>\n",
       "      <td>2</td>\n",
       "      <td>81.48</td>\n",
       "      <td>47160</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   population_key population_description  abs_freq  life_expectancy_birth  \\\n",
       "0               6  Bengali in Bangladesh         2                  71.51   \n",
       "1              19     Finnish in Finland         2                  81.48   \n",
       "\n",
       "   gross_national_income  \n",
       "0                   1220  \n",
       "1                  47160  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_disease = 'cancer'\n",
    "\n",
    "SQL = \"\"\"\n",
    "SELECT PD.population_key, PD.population_description, s1.abs_freq, PD.life_expectancy_birth, PD.gross_national_income \n",
    "FROM population_dimension PD\n",
    "INNER JOIN(\n",
    "  SELECT population_id, count(population_id) AS abs_freq \n",
    "  FROM snp_person_fact \n",
    "  WHERE snp_id in  (\n",
    "    SELECT snp_id \n",
    "    FROM disease_gwas_fact \n",
    "    WHERE disease_id in (\n",
    "      SELECT disease_key \n",
    "      FROM disease_dimension \n",
    "      WHERE disease_name = %s\n",
    "    )\n",
    "  )\n",
    "  GROUP BY population_id\n",
    ") AS s1 ON s1.population_id = PD.population_key\n",
    "WHERE PD.population_key IN (6,19)\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "df1  = getSQLfromQuery(SQL, (test_disease,), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Bottlenecks\n",
    "### a. Identify the most important bottlenecks and execute performance analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Os bottlenecks mais importantes no HuGS-DW, vão estar relacionados com queries que tenham quer percorrer as tabelas de factos e algumas das dimensões, como a SNP_dimension. A dimensão destas tabelas, tornarão queries que tenham que atravessar as tabelas e agregar dados dentro dela, em operações que podem ser muito lentas. \n",
    "\n",
    "Vamos por isso, a título de exemplo, executar alguns `EXPLAIN ANALYZE` a algumas das nossas queries."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Grupo de queries 1\n",
    "- **Query que calcula a percentagem de alelos de referência e alternativos nas várias sub-populações.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sort  (cost=547.56..547.66 rows=40 width=282) ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sort Key: ((((count(CASE WHEN (sp.allele_1 =...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Method: quicksort  Memory: 27kB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-&gt;  HashAggregate  (cost=544.90..546.50 rows...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Hash Join  (cost=10.90..300.50 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hash Cond: (sp.population_id = p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Seq Scan on snp_person_fact ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>-&gt;  Hash  (cost=10.40..10.40 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Buckets: 1024  Batches: 1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-&gt;  Seq Scan on population...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Total runtime: 14.745 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   Sort  (cost=547.56..547.66 rows=40 width=282) ...\n",
       "1     Sort Key: ((((count(CASE WHEN (sp.allele_1 =...\n",
       "2                Sort Method: quicksort  Memory: 27kB\n",
       "3     ->  HashAggregate  (cost=544.90..546.50 rows...\n",
       "4           ->  Hash Join  (cost=10.90..300.50 row...\n",
       "5                 Hash Cond: (sp.population_id = p...\n",
       "6                 ->  Seq Scan on snp_person_fact ...\n",
       "7                 ->  Hash  (cost=10.40..10.40 row...\n",
       "8                       Buckets: 1024  Batches: 1 ...\n",
       "9                       ->  Seq Scan on population...\n",
       "10                           Total runtime: 14.745 ms"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT population_description, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "        (cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, population_dimension P\n",
    "WHERE P.population_key = SP.population_id\n",
    "GROUP BY population_description\n",
    "ORDER BY alternative_frequency DESC\n",
    "\"\"\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "73 ms ± 2.95 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Query que calcula, por SNP e para cada Super-população e sub-população, as frequências do alelo de referência e do alelo alternativo.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GroupAggregate  (cost=2410.80..2993.60 rows=75...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-&gt;  Sort  (cost=2410.80..2429.60 rows=7520 w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Key: s.snp_ref_code, p.super_popu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sort Method: quicksort  Memory: 780kB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Hash Join  (cost=681.26..897.64 ro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hash Cond: (sp.population_id = p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Merge Join  (cost=670.36..78...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Merge Cond: (s.snp_id = sp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-&gt;  Index Scan using snp_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-&gt;  Sort  (cost=670.36..68...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Sort Key: sp.snp_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Sort Method: quickso...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>-&gt;  Seq Scan on snp_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>-&gt;  Hash  (cost=10.40..10.40 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Buckets: 1024  Batches: 1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>-&gt;  Seq Scan on population...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Total runtime: 37.053 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   GroupAggregate  (cost=2410.80..2993.60 rows=75...\n",
       "1     ->  Sort  (cost=2410.80..2429.60 rows=7520 w...\n",
       "2           Sort Key: s.snp_ref_code, p.super_popu...\n",
       "3               Sort Method: quicksort  Memory: 780kB\n",
       "4           ->  Hash Join  (cost=681.26..897.64 ro...\n",
       "5                 Hash Cond: (sp.population_id = p...\n",
       "6                 ->  Merge Join  (cost=670.36..78...\n",
       "7                       Merge Cond: (s.snp_id = sp...\n",
       "8                       ->  Index Scan using snp_d...\n",
       "9                       ->  Sort  (cost=670.36..68...\n",
       "10                                Sort Key: sp.snp_id\n",
       "11                            Sort Method: quickso...\n",
       "12                            ->  Seq Scan on snp_...\n",
       "13                ->  Hash  (cost=10.40..10.40 row...\n",
       "14                      Buckets: 1024  Batches: 1 ...\n",
       "15                      ->  Seq Scan on population...\n",
       "16                           Total runtime: 37.053 ms"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT snp_ref_code, super_population, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, population_dimension P, snp_dim S\n",
    "WHERE P.population_key = SP.population_id AND S.snp_id=SP.snp_id\n",
    "GROUP BY super_population, snp_ref_code\n",
    "ORDER BY snp_ref_code\n",
    "\"\"\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "118 ms ± 8.28 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Query que calcula, por SNP, País e sub-população, as frequências do alelo de referência e do alelo alternativo.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GroupAggregate  (cost=1470.90..2071.78 rows=75...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-&gt;  Sort  (cost=1470.90..1489.68 rows=7511 w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Key: s.snp_ref_code, p.country, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sort Method: external sort  Disk: 512kB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Hash Join  (cost=761.70..987.39 ro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hash Cond: (sp.person_id = p.per...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Merge Join  (cost=670.36..78...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Merge Cond: (s.snp_id = sp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-&gt;  Index Scan using snp_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-&gt;  Sort  (cost=670.36..68...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Sort Key: sp.snp_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Sort Method: quickso...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>-&gt;  Seq Scan on snp_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>-&gt;  Hash  (cost=60.04..60.04 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Buckets: 1024  Batches: 1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>-&gt;  Seq Scan on person_dim...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Total runtime: 302.062 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   GroupAggregate  (cost=1470.90..2071.78 rows=75...\n",
       "1     ->  Sort  (cost=1470.90..1489.68 rows=7511 w...\n",
       "2           Sort Key: s.snp_ref_code, p.country, p...\n",
       "3             Sort Method: external sort  Disk: 512kB\n",
       "4           ->  Hash Join  (cost=761.70..987.39 ro...\n",
       "5                 Hash Cond: (sp.person_id = p.per...\n",
       "6                 ->  Merge Join  (cost=670.36..78...\n",
       "7                       Merge Cond: (s.snp_id = sp...\n",
       "8                       ->  Index Scan using snp_d...\n",
       "9                       ->  Sort  (cost=670.36..68...\n",
       "10                                Sort Key: sp.snp_id\n",
       "11                            Sort Method: quickso...\n",
       "12                            ->  Seq Scan on snp_...\n",
       "13                ->  Hash  (cost=60.04..60.04 row...\n",
       "14                      Buckets: 1024  Batches: 1 ...\n",
       "15                      ->  Seq Scan on person_dim...\n",
       "16                          Total runtime: 302.062 ms"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT snp_ref_code, country, population_description, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, person_dimension P, snp_dim S\n",
    "WHERE P.person_key = SP.person_id AND S.snp_id=SP.snp_id\n",
    "GROUP BY country, population_description, snp_ref_code\n",
    "ORDER BY snp_ref_code\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "335 ms ± 32.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Grupo de queries 2\n",
    "- **Query que calcula para cada SNP e para cada país, as frequências alélicas encontradas para cada género. Associa esta informação à taxa de mortalidade no País (dividida por género).**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GroupAggregate  (cost=2568.51..3207.71 rows=75...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-&gt;  Sort  (cost=2568.51..2587.31 rows=7520 w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Key: s.snp_ref_code, pd.country, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sort Method: quicksort  Memory: 1009kB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Hash Join  (cost=683.07..1002.85 r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hash Cond: (sp.population_id = p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Hash Join  (cost=672.17..888...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Hash Cond: (sp.gender_id =...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-&gt;  Merge Join  (cost=670....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Merge Cond: (s.snp_i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-&gt;  Index Scan using...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>-&gt;  Sort  (cost=670....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Sort Key: sp.s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Sort Method: q...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>-&gt;  Seq Scan o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>-&gt;  Hash  (cost=1.36..1.36...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Buckets: 1024  Batch...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>-&gt;  Seq Scan on gend...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>-&gt;  Hash  (cost=10.40..10.40 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Buckets: 1024  Batches: 1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>-&gt;  Seq Scan on population...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Total runtime: 106.766 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   GroupAggregate  (cost=2568.51..3207.71 rows=75...\n",
       "1     ->  Sort  (cost=2568.51..2587.31 rows=7520 w...\n",
       "2           Sort Key: s.snp_ref_code, pd.country, ...\n",
       "3              Sort Method: quicksort  Memory: 1009kB\n",
       "4           ->  Hash Join  (cost=683.07..1002.85 r...\n",
       "5                 Hash Cond: (sp.population_id = p...\n",
       "6                 ->  Hash Join  (cost=672.17..888...\n",
       "7                       Hash Cond: (sp.gender_id =...\n",
       "8                       ->  Merge Join  (cost=670....\n",
       "9                             Merge Cond: (s.snp_i...\n",
       "10                            ->  Index Scan using...\n",
       "11                            ->  Sort  (cost=670....\n",
       "12                                  Sort Key: sp.s...\n",
       "13                                  Sort Method: q...\n",
       "14                                  ->  Seq Scan o...\n",
       "15                      ->  Hash  (cost=1.36..1.36...\n",
       "16                            Buckets: 1024  Batch...\n",
       "17                            ->  Seq Scan on gend...\n",
       "18                ->  Hash  (cost=10.40..10.40 row...\n",
       "19                      Buckets: 1024  Batches: 1 ...\n",
       "20                      ->  Seq Scan on population...\n",
       "21                          Total runtime: 106.766 ms"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "EXPLAIN ANALYZE SELECT  snp_ref_code, PD.country, gender,  mortality_per_1000_habitants, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, gender_dimension G, snp_dim S, population_dimension PD\n",
    "WHERE G.gender_key = SP.gender_id AND S.snp_id=SP.snp_id AND PD.population_key = SP.population_id\n",
    "GROUP BY snp_ref_code, gender, gross_national_income, PD.country, mortality_per_1000_habitants\n",
    "ORDER BY snp_ref_code, country\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "157 ms ± 8.02 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Grupo de queries 3\n",
    "- **Query que calcula a frequência absoluta de um SNP de risco associado a uma determinada doença, em indíviduos de determinadas sub-populações.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sort  (cost=787.50..787.60 rows=41 width=37) (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sort Key: (count(pd.population_description))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Method: quicksort  Memory: 25kB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-&gt;  HashAggregate  (cost=785.99..786.40 rows...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Nested Loop  (cost=733.82..785.69 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-&gt;  HashAggregate  (cost=733.82....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Nested Loop Semi Join ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Join Filter: (sp.snp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Rows Removed by Join...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-&gt;  Seq Scan on snp_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Filter: ((alle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Rows Removed b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>-&gt;  Materialize  (co...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>-&gt;  Hash Semi ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Hash Con...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>-&gt;  Seq ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>-&gt;  Hash...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Bu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>-&gt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>-&gt;  Index Scan using person_dime...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Index Cond: (person_key = ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Total runtime: 13.596 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   Sort  (cost=787.50..787.60 rows=41 width=37) (...\n",
       "1        Sort Key: (count(pd.population_description))\n",
       "2                Sort Method: quicksort  Memory: 25kB\n",
       "3     ->  HashAggregate  (cost=785.99..786.40 rows...\n",
       "4           ->  Nested Loop  (cost=733.82..785.69 ...\n",
       "5                 ->  HashAggregate  (cost=733.82....\n",
       "6                       ->  Nested Loop Semi Join ...\n",
       "7                             Join Filter: (sp.snp...\n",
       "8                             Rows Removed by Join...\n",
       "9                             ->  Seq Scan on snp_...\n",
       "10                                  Filter: ((alle...\n",
       "11                                  Rows Removed b...\n",
       "12                            ->  Materialize  (co...\n",
       "13                                  ->  Hash Semi ...\n",
       "14                                        Hash Con...\n",
       "15                                        ->  Seq ...\n",
       "16                                        ->  Hash...\n",
       "17                                              Bu...\n",
       "18                                              ->...\n",
       "19                                                ...\n",
       "20                                                ...\n",
       "21                ->  Index Scan using person_dime...\n",
       "22                      Index Cond: (person_key = ...\n",
       "23                           Total runtime: 13.596 ms"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complex_disease = 'asthma'\n",
    "\n",
    "SQL = \"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT super_population, population_description, count(population_description) AS abs_freq \n",
    "FROM person_dimension PD \n",
    "WHERE person_key in (\n",
    "  SELECT person_id \n",
    "  FROM snp_person_fact SP\n",
    "  WHERE snp_id in (\n",
    "    SELECT snp_id \n",
    "    FROM disease_gwas_fact DG\n",
    "    WHERE disease_id in (\n",
    "      SELECT disease_key \n",
    "      FROM disease_dimension D\n",
    "      WHERE disease_name like %s\n",
    "\t  )\n",
    "    )\n",
    "    AND (SP.allele_1 = 1 or SP.allele_2 = 1)\n",
    ")\n",
    "GROUP BY super_population, population_description \n",
    "ORDER BY abs_freq desc\n",
    "\"\"\"\n",
    "\n",
    "df1  = getSQLfromQuery(SQL, (complex_disease,), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "69.5 ms ± 1.91 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (complex_disease,), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Grupo de queries 4\n",
    "- **Query que calcula a frequência com que variações genéticas de risco, associadas a doenças, ocorrem em contexto \"non-coding\" e \"coding\".**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aggregate  (cost=1185.55..1185.58 rows=1 width...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-&gt;  Hash Join  (cost=1131.68..1164.27 rows=5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hash Cond: (df.snp_id = s.snp_id)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-&gt;  Hash Join  (cost=590.14..612.10 ro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Hash Cond: (df.disease_id = d.di...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-&gt;  Seq Scan on disease_gwas_fac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Hash  (cost=471.73..471.73 r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Buckets: 1024  Batches: 1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-&gt;  Seq Scan on disease_di...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-&gt;  Hash  (cost=459.57..459.57 rows=65...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Buckets: 1024  Batches: 1  Memor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>-&gt;  Seq Scan on snp_dim s  (cost...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Total runtime: 21.223 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   Aggregate  (cost=1185.55..1185.58 rows=1 width...\n",
       "1     ->  Hash Join  (cost=1131.68..1164.27 rows=5...\n",
       "2                   Hash Cond: (df.snp_id = s.snp_id)\n",
       "3           ->  Hash Join  (cost=590.14..612.10 ro...\n",
       "4                 Hash Cond: (df.disease_id = d.di...\n",
       "5                 ->  Seq Scan on disease_gwas_fac...\n",
       "6                 ->  Hash  (cost=471.73..471.73 r...\n",
       "7                       Buckets: 1024  Batches: 1 ...\n",
       "8                       ->  Seq Scan on disease_di...\n",
       "9           ->  Hash  (cost=459.57..459.57 rows=65...\n",
       "10                Buckets: 1024  Batches: 1  Memor...\n",
       "11                ->  Seq Scan on snp_dim s  (cost...\n",
       "12                           Total runtime: 21.223 ms"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL =\"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT cast(count(case when context like '3_prime_UTR_variant' or context like '5_prime_UTR_variant' or context like 'intergenic_variant' or context like 'intron_variant' or context like 'non_coding_transcript_exon_variant' or context like 'regulatory_region_variant' or context like 'TF_binding_site_variant' then 1 else null end)AS float)/count(context) AS non_coding, \n",
    "       cast(count(case when context like 'missense_variant' or context like 'splice_donor_variant' or context like 'splice_region_variant' or context like 'stop_gained' or context like 'synonymous_variant' then 1 else null end)AS float)/count(context) AS coding\n",
    "FROM disease_dimension D , snp_dim S, disease_gwas_fact DF\n",
    "WHERE DF.disease_id = D.disease_key  and S.snp_id = DF.snp_id\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77.3 ms ± 1.08 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sem grandes surpresas, as queries mais longas são as que utilizam a tabela snp_person_fact. Se considerarmos que esta tabela tem apenas uma fracção dos dados originais, facilmente conseguimos perceber como poderia dificultar o uso do DW para várias queries analíticas. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. Propose a plan for improving performance and implement it NOTE, if for improving performance, the existing database schema is to be changed, just describe the proposed changes – do NOT implement it!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Para aumentarmos a performance do nosso datawarehouse, poderíamos introduzir várias estratégias, como um planeamento eficaz de um desenho físico do sistema, o uso de índices, ou vistas materializadas."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### b.1 - Implementação de índices\n",
    "- O uso de índices permite um acesso eficaz a dados importantes. Os índices, não são mais que apontadares que o database management system pode usar para aumentar a velocidade de acesso aos dados.\n",
    "- No postgresql, os índices utilizados por defeito são B+ Tree indexes.\n",
    "- Vamos então tentar implementar estes indíces a alguns dos atributos das tabelas de factos e das dimensões de maior tamanho (SNP_Dimension e Disease_Dimension)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL = '''\n",
    "DROP INDEX IF EXISTS idx_SPF_person;\n",
    "DROP INDEX IF EXISTS idx_P_person;\n",
    "DROP INDEX IF EXISTS idx_SPF_snp;\n",
    "DROP INDEX IF EXISTS idx_S_snp;\n",
    "DROP INDEX IF EXISTS idx_DGF_disease;\n",
    "DROP INDEX IF EXISTS idx_D_disease;\n",
    "DROP INDEX IF EXISTS idx_DGF_snp;\n",
    "DROP INDEX IF EXISTS idx_DGF_study;\n",
    "DROP INDEX IF EXISTS idx_St_study;\n",
    "DROP INDEX IF EXISTS idx_snp_snp;\n",
    "DROP INDEX IF EXISTS idx_disease_name;\n",
    "\n",
    "'''\n",
    "#!!!!!!!! comentar esta proxima linha para evitar dados repetidos em nova execucao\n",
    "#excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL = \"\"\"\n",
    "CREATE INDEX idx_SPF_person ON snp_person_fact(person_id);\n",
    "CREATE INDEX idx_P_person ON person_dimension(person_key);\n",
    "CREATE INDEX idx_SPF_snp ON snp_person_fact(snp_id);\n",
    "CREATE INDEX idx_S_snp ON snp_dim(snp_id);\n",
    "CREATE INDEX idx_DGF_disease ON disease_gwas_fact(disease_id);\n",
    "CREATE INDEX idx_D_disease ON disease_dimension(disease_key);\n",
    "CREATE INDEX idx_DGF_snp ON disease_gwas_fact(snp_id);\n",
    "CREATE INDEX idx_DGF_study ON disease_gwas_fact(study_id);\n",
    "CREATE INDEX idx_St_study ON study_dim(study_id);\n",
    "CREATE INDEX idx_snp_snp ON snp_dim(snp_ref_code);\n",
    "CREATE INDEX idx_disease_name ON disease_dimension(disease_name);\n",
    "\"\"\"\n",
    "#!!!!!!!! comentar esta proxima linha para evitar dados repetidos em nova execucao\n",
    "#excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### b.2 - Uso de vistas materializadas\n",
    "- Em data warehouses, vistas materializadas podem ser usadas para pre-calcular e armazenar dados agregados. Normalmente, armazenam dados sumarizados. Podem também ser usados para pre-calcular joins de tabelas, com ou sem agregações. \n",
    "- Em suma, as vistas materializadas são usadas para aumentar a eficiência de joins ou agregações complicadas, para queries grandes ou importantes.\n",
    "- Infelizmente, a versão dos postgreSQL a correr nos servidores da FCUL, não permite a criação de vistas materializadas. De qualquer modo, a título académico, colocamos em baixo o código para a criação de vistas que consideramos importantes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Vista com dados acerca de variabilidade genética das diferentes populações, pré-calculada.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL  = \"\"\"\n",
    "CREATE /*MATERIALIZED*/ VIEW MV_freq_allele_population AS\n",
    "SELECT P.population_description, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "        (cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, population_dimension P\n",
    "WHERE P.population_key = SP.population_id\n",
    "GROUP BY population_description\n",
    "ORDER BY alternative_frequency DESC\n",
    "\"\"\"\n",
    "\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Vista com frequências alélicas de cada SNP, em diferentes super-populações.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL  = \"\"\"\n",
    "CREATE /*MATERIALIZED*/ VIEW MV_freq_snp_population as\n",
    "SELECT snp_ref_code, super_population, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "        (cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, population_dimension P, snp_dim S\n",
    "WHERE P.population_key = SP.population_id AND S.snp_id=SP.snp_id\n",
    "GROUP BY super_population, snp_ref_code\n",
    "ORDER BY snp_ref_code\n",
    "\"\"\"\n",
    "\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Vista que pré-calcula a frequência que variantes genéticas associadas a doenças, se encontram num contexto genómico particular.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL  = \"\"\"\n",
    "CREATE /*MATERIALIZED*/ VIEW MV_freq_allele_disease as\n",
    "with myquery as (\n",
    "select disease_name, snp_ref_code, context \n",
    "    from disease_dimension D , snp_dim S, disease_gwas_fact DF\n",
    "    where DF.disease_id = D.disease_key  and S.snp_id = DF.snp_id\n",
    "    group by disease_name, snp_ref_code, context \n",
    "   order by disease_name\n",
    ")\n",
    "\n",
    "select context,count(context),count(context)/(select count(*) from myquery)::float as frequency from myquery \n",
    "group by context\n",
    "\"\"\"\n",
    "\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Vista que agrega informação acerca de doenças, SNP de risco associado e o p_value para a sua associação com a doença.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "SQL =\"\"\"\n",
    "CREATE /*MATERIALIZED*/ VIEW MV_disease_pvalue AS\n",
    "SELECT DISTINCT D.disease_name, S.snp_ref_code, DF.p_value\n",
    "FROM disease_dimension D, disease_gwas_fact DF, snp_dim S\n",
    "WHERE DF.disease_id = D.disease_key\n",
    "ORDER BY p_value\n",
    "\"\"\"\n",
    "\n",
    "excuteSingleSQLstatement(SQL, host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### c. Verify that the proposed modifications have better performance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Para verificar se as modificações aumentaram a performance do HuGS-DW, vamos correr os `EXPLAIN ANALYZE` novamente."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### C.1 - Aumento de performance através do uso de índices\n",
    "#### Grupo de queries 1\n",
    "- **Query que calcula a percentagem de alelos de referência e alternativos nas várias sub-populações.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sort  (cost=547.56..547.66 rows=40 width=282) ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sort Key: ((((count(CASE WHEN (sp.allele_1 =...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Method: quicksort  Memory: 27kB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-&gt;  HashAggregate  (cost=544.90..546.50 rows...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Hash Join  (cost=10.90..300.50 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hash Cond: (sp.population_id = p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Seq Scan on snp_person_fact ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>-&gt;  Hash  (cost=10.40..10.40 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Buckets: 1024  Batches: 1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-&gt;  Seq Scan on population...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Total runtime: 7.220 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   Sort  (cost=547.56..547.66 rows=40 width=282) ...\n",
       "1     Sort Key: ((((count(CASE WHEN (sp.allele_1 =...\n",
       "2                Sort Method: quicksort  Memory: 27kB\n",
       "3     ->  HashAggregate  (cost=544.90..546.50 rows...\n",
       "4           ->  Hash Join  (cost=10.90..300.50 row...\n",
       "5                 Hash Cond: (sp.population_id = p...\n",
       "6                 ->  Seq Scan on snp_person_fact ...\n",
       "7                 ->  Hash  (cost=10.40..10.40 row...\n",
       "8                       Buckets: 1024  Batches: 1 ...\n",
       "9                       ->  Seq Scan on population...\n",
       "10                            Total runtime: 7.220 ms"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT population_description, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "        (cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, population_dimension P\n",
    "WHERE P.population_key = SP.population_id\n",
    "GROUP BY population_description\n",
    "ORDER BY alternative_frequency DESC\n",
    "\"\"\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "72.7 ms ± 2.96 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Query que calcula, por SNP e para cada Super-população e sub-população, as frequências do alelo de referência e do alelo alternativo.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GroupAggregate  (cost=2040.69..2623.49 rows=75...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-&gt;  Sort  (cost=2040.69..2059.49 rows=7520 w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Key: s.snp_ref_code, p.super_popu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sort Method: quicksort  Memory: 780kB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Hash Join  (cost=10.90..527.53 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hash Cond: (sp.population_id = p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Merge Join  (cost=0.00..413....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Merge Cond: (s.snp_id = sp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-&gt;  Index Scan using idx_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-&gt;  Index Scan using idx_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-&gt;  Hash  (cost=10.40..10.40 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Buckets: 1024  Batches: 1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>-&gt;  Seq Scan on population...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Total runtime: 62.611 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   GroupAggregate  (cost=2040.69..2623.49 rows=75...\n",
       "1     ->  Sort  (cost=2040.69..2059.49 rows=7520 w...\n",
       "2           Sort Key: s.snp_ref_code, p.super_popu...\n",
       "3               Sort Method: quicksort  Memory: 780kB\n",
       "4           ->  Hash Join  (cost=10.90..527.53 row...\n",
       "5                 Hash Cond: (sp.population_id = p...\n",
       "6                 ->  Merge Join  (cost=0.00..413....\n",
       "7                       Merge Cond: (s.snp_id = sp...\n",
       "8                       ->  Index Scan using idx_s...\n",
       "9                       ->  Index Scan using idx_s...\n",
       "10                ->  Hash  (cost=10.40..10.40 row...\n",
       "11                      Buckets: 1024  Batches: 1 ...\n",
       "12                      ->  Seq Scan on population...\n",
       "13                           Total runtime: 62.611 ms"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT snp_ref_code, super_population, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, population_dimension P, snp_dim S\n",
    "WHERE P.population_key = SP.population_id AND S.snp_id=SP.snp_id\n",
    "GROUP BY super_population, snp_ref_code\n",
    "ORDER BY snp_ref_code\n",
    "\"\"\"\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "119 ms ± 5.77 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Query que calcula, por SNP, País e sub-população, as frequências do alelo de referência e do alelo alternativo.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GroupAggregate  (cost=1100.80..1701.68 rows=75...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-&gt;  Sort  (cost=1100.80..1119.57 rows=7511 w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Key: s.snp_ref_code, p.country, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sort Method: external sort  Disk: 512kB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Hash Join  (cost=91.34..617.28 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hash Cond: (sp.person_id = p.per...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Merge Join  (cost=0.00..413....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Merge Cond: (s.snp_id = sp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-&gt;  Index Scan using idx_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-&gt;  Index Scan using idx_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-&gt;  Hash  (cost=60.04..60.04 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Buckets: 1024  Batches: 1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>-&gt;  Seq Scan on person_dim...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Total runtime: 330.149 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   GroupAggregate  (cost=1100.80..1701.68 rows=75...\n",
       "1     ->  Sort  (cost=1100.80..1119.57 rows=7511 w...\n",
       "2           Sort Key: s.snp_ref_code, p.country, p...\n",
       "3             Sort Method: external sort  Disk: 512kB\n",
       "4           ->  Hash Join  (cost=91.34..617.28 row...\n",
       "5                 Hash Cond: (sp.person_id = p.per...\n",
       "6                 ->  Merge Join  (cost=0.00..413....\n",
       "7                       Merge Cond: (s.snp_id = sp...\n",
       "8                       ->  Index Scan using idx_s...\n",
       "9                       ->  Index Scan using idx_s...\n",
       "10                ->  Hash  (cost=60.04..60.04 row...\n",
       "11                      Buckets: 1024  Batches: 1 ...\n",
       "12                      ->  Seq Scan on person_dim...\n",
       "13                          Total runtime: 330.149 ms"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT snp_ref_code, country, population_description, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, person_dimension P, snp_dim S\n",
    "WHERE P.person_key = SP.person_id AND S.snp_id=SP.snp_id\n",
    "GROUP BY country, population_description, snp_ref_code\n",
    "ORDER BY snp_ref_code\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "293 ms ± 34.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Grupo de queries 2\n",
    "- **Query que calcula para cada SNP e para cada país, as frequências alélicas encontradas para cada género. Associa esta informação à taxa de mortalidade no País (dividida por género).**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GroupAggregate  (cost=2198.40..2837.60 rows=75...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-&gt;  Sort  (cost=2198.40..2217.20 rows=7520 w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Key: s.snp_ref_code, pd.country, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sort Method: quicksort  Memory: 1009kB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Hash Join  (cost=12.71..632.74 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Hash Cond: (sp.population_id = p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Hash Join  (cost=1.81..518.4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Hash Cond: (sp.gender_id =...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-&gt;  Merge Join  (cost=0.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Merge Cond: (s.snp_i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-&gt;  Index Scan using...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>-&gt;  Index Scan using...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>-&gt;  Hash  (cost=1.36..1.36...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Buckets: 1024  Batch...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>-&gt;  Seq Scan on gend...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>-&gt;  Hash  (cost=10.40..10.40 row...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Buckets: 1024  Batches: 1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>-&gt;  Seq Scan on population...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Total runtime: 85.162 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   GroupAggregate  (cost=2198.40..2837.60 rows=75...\n",
       "1     ->  Sort  (cost=2198.40..2217.20 rows=7520 w...\n",
       "2           Sort Key: s.snp_ref_code, pd.country, ...\n",
       "3              Sort Method: quicksort  Memory: 1009kB\n",
       "4           ->  Hash Join  (cost=12.71..632.74 row...\n",
       "5                 Hash Cond: (sp.population_id = p...\n",
       "6                 ->  Hash Join  (cost=1.81..518.4...\n",
       "7                       Hash Cond: (sp.gender_id =...\n",
       "8                       ->  Merge Join  (cost=0.00...\n",
       "9                             Merge Cond: (s.snp_i...\n",
       "10                            ->  Index Scan using...\n",
       "11                            ->  Index Scan using...\n",
       "12                      ->  Hash  (cost=1.36..1.36...\n",
       "13                            Buckets: 1024  Batch...\n",
       "14                            ->  Seq Scan on gend...\n",
       "15                ->  Hash  (cost=10.40..10.40 row...\n",
       "16                      Buckets: 1024  Batches: 1 ...\n",
       "17                      ->  Seq Scan on population...\n",
       "18                           Total runtime: 85.162 ms"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL = \"\"\"\n",
    "EXPLAIN ANALYZE SELECT  snp_ref_code, PD.country, gender,  mortality_per_1000_habitants, (cast(count(case when allele_1=0 then 1 else null end)AS float)+cast(count(case when allele_2=0 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Reference_frequency,\n",
    "\t\t(cast(count(case when allele_1=1 then 1 else null end)AS float)+cast(count(case when allele_2=1 then 1 else null end)AS float))/(count(allele_1)+count(allele_2)) AS Alternative_frequency\n",
    "FROM snp_person_fact SP, gender_dimension G, snp_dim S, population_dimension PD\n",
    "WHERE G.gender_key = SP.gender_id AND S.snp_id=SP.snp_id AND PD.population_key = SP.population_id\n",
    "GROUP BY snp_ref_code, gender, gross_national_income, PD.country, mortality_per_1000_habitants\n",
    "ORDER BY snp_ref_code, country\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "155 ms ± 7.23 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Grupo de queries 3\n",
    "- **Query que calcula a frequência absoluta de um SNP de risco associado a uma determinada doença, em indíviduos de determinadas sub-populações.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sort  (cost=294.93..295.03 rows=41 width=37) (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Sort Key: (count(pd.population_description))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Method: quicksort  Memory: 25kB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-&gt;  HashAggregate  (cost=293.42..293.83 rows...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Nested Loop  (cost=241.25..293.12 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-&gt;  HashAggregate  (cost=241.25....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Hash Semi Join  (cost=...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Hash Cond: (sp.snp_i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-&gt;  Seq Scan on snp_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Filter: ((alle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Rows Removed b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>-&gt;  Hash  (cost=16.7...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Buckets: 1024 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>-&gt;  Nested Loo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>-&gt;  Inde...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>In...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Fi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>-&gt;  Bitm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Re...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>-&gt;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>-&gt;  Index Scan using idx_p_perso...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Index Cond: (person_key = ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Total runtime: 3.368 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   Sort  (cost=294.93..295.03 rows=41 width=37) (...\n",
       "1        Sort Key: (count(pd.population_description))\n",
       "2                Sort Method: quicksort  Memory: 25kB\n",
       "3     ->  HashAggregate  (cost=293.42..293.83 rows...\n",
       "4           ->  Nested Loop  (cost=241.25..293.12 ...\n",
       "5                 ->  HashAggregate  (cost=241.25....\n",
       "6                       ->  Hash Semi Join  (cost=...\n",
       "7                             Hash Cond: (sp.snp_i...\n",
       "8                             ->  Seq Scan on snp_...\n",
       "9                                   Filter: ((alle...\n",
       "10                                  Rows Removed b...\n",
       "11                            ->  Hash  (cost=16.7...\n",
       "12                                  Buckets: 1024 ...\n",
       "13                                  ->  Nested Loo...\n",
       "14                                        ->  Inde...\n",
       "15                                              In...\n",
       "16                                              Fi...\n",
       "17                                        ->  Bitm...\n",
       "18                                              Re...\n",
       "19                                              ->...\n",
       "20                                                ...\n",
       "21                ->  Index Scan using idx_p_perso...\n",
       "22                      Index Cond: (person_key = ...\n",
       "23                            Total runtime: 3.368 ms"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complex_disease = 'asthma'\n",
    "\n",
    "SQL = \"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT super_population, population_description, count(population_description) AS abs_freq \n",
    "FROM person_dimension PD \n",
    "WHERE person_key in (\n",
    "  SELECT person_id \n",
    "  FROM snp_person_fact SP\n",
    "  WHERE snp_id in (\n",
    "    SELECT snp_id \n",
    "    FROM disease_gwas_fact DG\n",
    "    WHERE disease_id in (\n",
    "      SELECT disease_key \n",
    "      FROM disease_dimension D\n",
    "      WHERE disease_name like %s\n",
    "\t  )\n",
    "    )\n",
    "    AND (SP.allele_1 = 1 or SP.allele_2 = 1)\n",
    ")\n",
    "GROUP BY super_population, population_description \n",
    "ORDER BY abs_freq desc\n",
    "\"\"\"\n",
    "\n",
    "df1  = getSQLfromQuery(SQL, (complex_disease,), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64.4 ms ± 1.51 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (complex_disease,), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Grupo de queries 4\n",
    "- **Query que calcula a frequência com que variações genéticas de risco, associadas a doenças, ocorrem em contexto \"non-coding\" e \"coding\".**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Aggregate  (cost=1185.55..1185.58 rows=1 width...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-&gt;  Hash Join  (cost=1131.68..1164.27 rows=5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hash Cond: (df.snp_id = s.snp_id)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-&gt;  Hash Join  (cost=590.14..612.10 ro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Hash Cond: (df.disease_id = d.di...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-&gt;  Seq Scan on disease_gwas_fac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>-&gt;  Hash  (cost=471.73..471.73 r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Buckets: 1024  Batches: 1 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-&gt;  Seq Scan on disease_di...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-&gt;  Hash  (cost=459.57..459.57 rows=65...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Buckets: 1024  Batches: 1  Memor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>-&gt;  Seq Scan on snp_dim s  (cost...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Total runtime: 34.605 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   Aggregate  (cost=1185.55..1185.58 rows=1 width...\n",
       "1     ->  Hash Join  (cost=1131.68..1164.27 rows=5...\n",
       "2                   Hash Cond: (df.snp_id = s.snp_id)\n",
       "3           ->  Hash Join  (cost=590.14..612.10 ro...\n",
       "4                 Hash Cond: (df.disease_id = d.di...\n",
       "5                 ->  Seq Scan on disease_gwas_fac...\n",
       "6                 ->  Hash  (cost=471.73..471.73 r...\n",
       "7                       Buckets: 1024  Batches: 1 ...\n",
       "8                       ->  Seq Scan on disease_di...\n",
       "9           ->  Hash  (cost=459.57..459.57 rows=65...\n",
       "10                Buckets: 1024  Batches: 1  Memor...\n",
       "11                ->  Seq Scan on snp_dim s  (cost...\n",
       "12                           Total runtime: 34.605 ms"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL =\"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT cast(count(case when context like '3_prime_UTR_variant' or context like '5_prime_UTR_variant' or context like 'intergenic_variant' or context like 'intron_variant' or context like 'non_coding_transcript_exon_variant' or context like 'regulatory_region_variant' or context like 'TF_binding_site_variant' then 1 else null end)AS float)/count(context) AS non_coding, \n",
    "       cast(count(case when context like 'missense_variant' or context like 'splice_donor_variant' or context like 'splice_region_variant' or context like 'stop_gained' or context like 'synonymous_variant' then 1 else null end)AS float)/count(context) AS coding\n",
    "FROM disease_dimension D , snp_dim S, disease_gwas_fact DF\n",
    "WHERE DF.disease_id = D.disease_key  and S.snp_id = DF.snp_id\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "77.7 ms ± 3.86 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)\n"
     ]
    }
   ],
   "source": [
    "%timeit getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Como podemos constatar, o uso de índices melhora de maneira inconsistente e não substancial as buscas ao DW. Isto está provavelmente relacionado com o facto de que os dados introduzidos são ainda muito poucos para justificar um aumento significativo da performance das buscas.\n",
    "### Uma escolha mais cuidada dos atributos a indexar levaria provavelmente, a aumentos de performance mais interessantes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### C.1 - Aumento de performance através do de vistas materializadas\n",
    "NOTA: Como mencionamos em cima, não se podem criar vistas materializadas na versao do postgreSQL da FCUL."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **`SELECT` correpondente À vista que agrega informação acerca de doenças, SNP de risco associado e o p_value para a sua associação com a doença.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Unique  (cost=745785.62..780668.86 rows=348832...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-&gt;  Sort  (cost=745785.62..754506.43 rows=34...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Key: df.p_value, d.disease_name, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sort Method: external merge  Disk: 176...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Nested Loop  (cost=3.37..44777.67 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-&gt;  Merge Join  (cost=3.37..697....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Merge Cond: (d.disease_key...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>-&gt;  Index Scan using idx_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-&gt;  Index Scan using idx_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-&gt;  Materialize  (cost=0.00..492...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-&gt;  Seq Scan on snp_dim s ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Total runtime: 112354.037 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   Unique  (cost=745785.62..780668.86 rows=348832...\n",
       "1     ->  Sort  (cost=745785.62..754506.43 rows=34...\n",
       "2           Sort Key: df.p_value, d.disease_name, ...\n",
       "3           Sort Method: external merge  Disk: 176...\n",
       "4           ->  Nested Loop  (cost=3.37..44777.67 ...\n",
       "5                 ->  Merge Join  (cost=3.37..697....\n",
       "6                       Merge Cond: (d.disease_key...\n",
       "7                       ->  Index Scan using idx_d...\n",
       "8                       ->  Index Scan using idx_d...\n",
       "9                 ->  Materialize  (cost=0.00..492...\n",
       "10                      ->  Seq Scan on snp_dim s ...\n",
       "11                       Total runtime: 112354.037 ms"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL =\"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT DISTINCT D.disease_name, S.snp_ref_code, DF.p_value\n",
    "FROM disease_dimension D, disease_gwas_fact DF, snp_dim S\n",
    "WHERE DF.disease_id = D.disease_key\n",
    "ORDER BY p_value\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **`SELECT` da vista materializada (que neste caso... não é materializada!).**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>QUERY PLAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Unique  (cost=745785.62..780668.86 rows=348832...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-&gt;  Sort  (cost=745785.62..754506.43 rows=34...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Sort Key: df.p_value, d.disease_name, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Sort Method: external merge  Disk: 176...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-&gt;  Nested Loop  (cost=3.37..44777.67 ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-&gt;  Merge Join  (cost=3.37..697....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Merge Cond: (d.disease_key...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>-&gt;  Index Scan using idx_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-&gt;  Index Scan using idx_d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>-&gt;  Materialize  (cost=0.00..492...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-&gt;  Seq Scan on snp_dim s ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Total runtime: 106364.285 ms</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           QUERY PLAN\n",
       "0   Unique  (cost=745785.62..780668.86 rows=348832...\n",
       "1     ->  Sort  (cost=745785.62..754506.43 rows=34...\n",
       "2           Sort Key: df.p_value, d.disease_name, ...\n",
       "3           Sort Method: external merge  Disk: 176...\n",
       "4           ->  Nested Loop  (cost=3.37..44777.67 ...\n",
       "5                 ->  Merge Join  (cost=3.37..697....\n",
       "6                       Merge Cond: (d.disease_key...\n",
       "7                       ->  Index Scan using idx_d...\n",
       "8                       ->  Index Scan using idx_d...\n",
       "9                 ->  Materialize  (cost=0.00..492...\n",
       "10                      ->  Seq Scan on snp_dim s ...\n",
       "11                       Total runtime: 106364.285 ms"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SQL =\"\"\"\n",
    "EXPLAIN ANALYZE \n",
    "SELECT *\n",
    "FROM MV_disease_pvalue\n",
    "\"\"\"\n",
    "\n",
    "df  = getSQLfromQuery(SQL, (), host=\"appserver-01.alunos.di.fc.ul.pt\",database=\"tpd020\", user=\"tpd020\", password=\"tpd020_1000genomes\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aqui não podemos testar o aumento da eficácia que a vista materializada iria conceder, pois não conseguimos criar esta estrutura no nosso DW."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### d. Discuss: With the full use of the data warehouse what are the expectation for its growth in 1, 5 and 10 years. Identify what approaches would be used to address such an increase in data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Como já foi referido nas partes anteriores do projeto, o HuGS-DW é relativamente simples em termos de actualizações no tempo. O crescimento acontecerá apenas quando novos estudos GWAS forem publicados, ou novos dados genéticos estiverem disponíveis. Neste sentido, e no ponto de vista do exercício académico, estas alterações seriam periódicas e consistiriam em adicionar novas linhas às tabelas de dimensões respetivas (com consequente aumento dos factos, para cada estudo e para cada indivíduos cujo genoma fosse sequenciado). Neste campo, será importante garantir que aquando da adição de novos dados, estes são adicionados na ordem correcta (sem reintroduzir dados previamente inseridos).\n",
    "No que diz respeito a Slow Changing Dimensions (SCDs), as únicas SCD que podemos considerar são aquelas que agregam informação demográfica. Nos actuais esquemas de negócio, não existe interesse em guardar historial dos dados, pelo que estas dimensões comportar-se-iam com SCD do tipo 1, simplesmente substituíndo os indíces demográficos por valores actualizados. Se existisse a necessidade de se referir a dados demográficos históricos, poder-se-ia considerar o uso de SCDs de tipo IV, onde se usaria uma \"mini-dimension\" com o histórico dos dados (de 5 em 5 anos) e que não obrigaria a grandes alterações ao esquema actual.\n",
    "\n",
    "Num ponto de vista teórico, se o aumento de dados genómicos e acerca de doenças acompanhasse o crescimento exponencial de dados que se observa actualmente, seríamos obrigados a implementar eficazmente vários índices e vistas materializadas, para aumentar a performance do sistema. Para além disso, e para lidar com o aumento extraordinário de dados, teríamos que tomar várias opções importantes acerca do desenho físico do DW. Nomeadamente, teríamos que ter em atenção várias questões relacionadas com o tipo de suporte físico usado para guardar os dados, estratégias de compressão para diminuir o espaço ocupado pelos dados, partição e armazenamento distribuído dos vários dados (por exemplo, separando os dados genómicos, dos dados acerca de doenças)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Do a critical assessment of this full project\n",
    "### a. For each of the 3 phases of the project discuss what went right, wrong and what would you do to make it better?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Na globalidade, acreditamos que este foi um exercícios académico que nos permitiu explorar as várias fases do processo de elaboração de um Data Warehouse. Foi, obviamente, um projecto muito exigente tendo em conta a dimensão e origem variada dos dados que nos propusemos aglomerar. Não obstante, apesar de algumas dificuldades, foi certamente um projecto bem concebido, conduzido e concretizado.\n",
    "\n",
    "No que diz respeito às várias fases do projecto, podemos concluir:\n",
    "### Fase 1 - Concepção do HuGS-DW\n",
    "Nesta fase, criou-se o esboço do que viria a ser o nosso DW. \n",
    "\n",
    "#### O positivo\n",
    "- O processo de concepção foi enriquecedor e globalmente positivo. Permitiu identificar um star-schema adequado, um conjunto de queries analíticas relevantes e um sistema de dados bem definido. O facto de que o modelo pouco teve que mudar comprova que a fase de desenho do data warehouse foi bem sucedidada. \n",
    "\n",
    "#### O menos positivo\n",
    "- Aqui o ponto mais frágil, foi uma definição mais exacta dos processos de negócio. Talvez por nos faltar a visão de \"negócio\" por detrás dos DW e dos processos analíticos procurados por certos tipos de negócios, este ponto foi abordado mais conceptualmente. Não obstante, a definação dos datamarts parece-nos adequada e, retrospectivamente, parecem-nos datamarts que poderão ser usados por Institutos e centros de investigação cientifica, departamentos e centros médicos que lidam com doenças genéticas e outras estruturas, como seguradoras e entidades de avaliação de riscos.\n",
    "\n",
    "#### O que faríamos para melhorar esta fase?\n",
    "- Para além de uma melhor definição dos processos de negócio, acreditamos que pouco havia a mudar.\n",
    "\n",
    "### Fase 2 - Criação do sistema de dados e do ETL workflow\n",
    "Esta fase foi, de longe, a mais complicada e que mais exigiu de cada um de nós. A escolha de um projecto de tamanha dimensão e envergadura, foi um claro risco. Mas entre as dificuldades, limitações e noites mal dormidas, achamos que conseguimos criar um sistema de dados que não se tem que envergonhar do objectivo com o qual foi criado. \n",
    "\n",
    "#### O positivo\n",
    "- O mais positivo desta fase, foi ter elaborado um sistema de dados muito rico, que foi buscar informação de muitas fontes de dados diferentes. Desde informação genética do 1000 genomes, NCBI, GWAS, informações acerca de doenças da ontologia DOID e dados demográficos do World Bank. Ler, filtrar e juntar dados tão diversos durante o processo de ETL, foi certamente um desafio. Mas termos conseguido fazê-lo (ainda que com algumas limitações), foi a nossa maior realização. \n",
    "\n",
    "#### O menos positivo\n",
    "- Os pontos menos positivos prendem-se com o ponto que acabamos de mencionar. O processo de ler, filtrar e carregar informação num sistema de dados é, de longe, o mais complicado e exigente. Este projecto deixou-o muito claro. Apesar de termos conseguido modelar a maior parte das dimensões, houve dimensões que, pela sua magnitude, se revelaram um desafio enorme. Nomeadamente, as dimensões Study_Dimension e SNP_Dimension. Consequentemente, a modelação e injecção de dados nas tabelas de factos também não foram ideais. Acreditamos que este foi o ponto mais negativo desta fase. Mas não temos a menor dúvida que este seria sempre, o ponto mais complicado no processo de criação de um DW com tanta informação e com tantas fontes.\n",
    "\n",
    "#### O que faríamos para melhorar esta fase?\n",
    "- Faríamos o que todos os analistas a programadores fariam com um projecto desta envergadura. Dispender muito mais tempo no processo de ETL, para garantir que os dados estavam com a melhor qualidade possível.\n",
    "\n",
    "### Fase 3 - Uso do HuGS-DW para análise e testes de eficiência\n",
    "Nesta última fase procedemos à implementação de queries analíticas para extrair conhecimento dos dados que introduzimos do DW.\n",
    "#### O positivo\n",
    "- O mais positivo foi finalmente ver o resultado das queries que podem ser efectuadas no nosso DW e ver como se pode derivar conhecimento analítico dos dados do HuGS-DW. Como um pequeno exemplo, uma das nossas primeiras queries neste projecto, acerca da variabilidade genética de diferentes populações, mostra que as populações com maior variabilidade genética são as Africanas. Obviamente, estes não são dados novos, mas reproduzem dados previamente publicados. O que é mais extraordinário, é termos tido esta informação mesmo só tendo a informação genética para 3 SNPs (de 2504 indivíduos). \n",
    "- A facilidade de aceder a informação acerca das doenças e fazer um drill-across para a variabilidade genética, também foi algo que nos satisfaz muito e tem imenso potencial no ponto de vista analítico.\n",
    "\n",
    "#### O menos positivo\n",
    "- O ponto menos positivo deriva ainda dos problemas que tivemos com a modelação das nossas dimensões e factos. Este problema foi contornado com a introdução de alguns dados (e também alguma \"dummy data\"). Visto os dados introduzidos terem sido relativamente menores que os planeados originalmente, as análises de performance ficaram de algum modo comprometidas.\n",
    "\n",
    "#### O que faríamos para melhorar esta fase?\n",
    "- Não há muito a fazer para melhorar esta fase, para além de melhorar a fase 2 e o workflow de ETL. Podemos dizer que o nosso DW está numa versão alpha, ainda muito preliminar. Não por nossa culpa, mas porque um projecto desta dimensão iria requerer um muito maior investimento de tempo e energia de cada um de nós. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. For each phase propose a grade (1-20) Justify your grade"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As notas que propomos reflectem a análise que fizemos no ponto anterior, que cremos justificar a avaliação. \n",
    "\n",
    "- Fase 1: 18\n",
    "- Fase 2: 16\n",
    "- Fase 3: 17\n",
    "\n",
    "Concluíndo, acreditamos que os objectivos e desafios que propusemos, foram atingidos e superados. A dificuldade deste projecto era inegável, assim como é o sucesso do ponto ao qual conseguimos chegar com o nosso DW."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
